// @oracle/oraclejet-webdriver/elements/oj-input-number/OjInputNumberBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjInputNumberBase = void 0;
const __1 = require("../");
class OjInputNumberBase extends __1.OjWebElement {
    getAsyncValidators() {
        return this.getProperty("asyncValidators");
    }
    getAutocomplete() {
        return this.getProperty("autocomplete");
    }
    getAutofocus() {
        return this.getProperty("autofocus");
    }
    getConverter() {
        return this.getProperty("converter");
    }
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getDisplayOptions() {
        return this.getProperty("displayOptions");
    }
    getHelp() {
        return this.getProperty("help");
    }
    getHelpHints() {
        return this.getProperty("helpHints");
    }
    getLabelEdge() {
        return this.getProperty("labelEdge");
    }
    getLabelHint() {
        return this.getProperty("labelHint");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    getMax() {
        return this.getProperty("max");
    }
    changeMessagesCustom(messagesCustom) {
        return this.setProperty("messagesCustom", messagesCustom);
    }
    getMessagesCustom() {
        return this.getProperty("messagesCustom");
    }
    getMin() {
        return this.getProperty("min");
    }
    getName() {
        return this.getProperty("name");
    }
    getPlaceholder() {
        return this.getProperty("placeholder");
    }
    getRawValue() {
        return this.getProperty("rawValue");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getRequired() {
        return this.getProperty("required");
    }
    getStep() {
        return this.getProperty("step");
    }
    getTransientValue() {
        return this.getProperty("transientValue");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUserAssistanceDensity() {
        return this.getProperty("userAssistanceDensity");
    }
    getValid() {
        return this.getProperty("valid");
    }
    getValidators() {
        return this.getProperty("validators");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
    getVirtualKeyboard() {
        return this.getProperty("virtualKeyboard");
    }
}
exports.OjInputNumberBase = OjInputNumberBase;
