// @oracle/oraclejet-webdriver/elements/oj-radioset/OjRadiosetBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjRadioset, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjRadiosetBase extends OjWebElement {
  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies if the component is disabled. If true, then all of its inputs and labels are also disabled. See the Help documentation for more information.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjRadiosetBase.DisplayOptions> {
    return this.getProperty<OjRadiosetBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjRadiosetBase.Help> {
    return this.getProperty<OjRadiosetBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjRadiosetBase.HelpHints> {
    return this.getProperty<OjRadiosetBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * Establishes a relationship between this component and another element, typically an oj-label custom element. See the Help documentation for more information.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of optionRenderer property.
   * The renderer function that renders the content of each option.
   * @return The value of optionRenderer property.
   * 
   */
  public getOptionRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("optionRenderer");
  }

  /**
   * Gets the value of options property.
   * The option items for the Radioset.
   * @return The value of options property.
   * 
   */
  public getOptions(): Promise<object|null> {
    return this.getProperty<object|null>("options");
  }

  /**
   * Gets the value of optionsKeys property.
   * Specify the key names to use in the options array. Depending on options-keys means that the signature of the data does not match what is supported by the options attribute.
   * @return The value of optionsKeys property.
   * 
   */
  public getOptionsKeys(): Promise<OjRadiosetBase.OptionsKeys> {
    return this.getProperty<OjRadiosetBase.OptionsKeys>("optionsKeys");
  }

  /**
   * Gets the value of readonly property.
   * Specifies whether the component is read-only. A read-only element cannot be modified, but user interaction is allowed. See the Help documentation for more information.
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of required property.
   * Specifies whether the component is required or optional. See the Help documentation for more information.
   * @return The value of required property.
   * 
   */
  public getRequired(): Promise<boolean> {
    return this.getProperty<boolean>("required");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjRadiosetBase.Translations|null> {
    return this.getProperty<OjRadiosetBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Sets the value of "value" property.
   * The value of the component. See the Help documentation for more information.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: any): Promise<void> {
    return this.setProperty<any>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the component. See the Help documentation for more information.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<any> {
    return this.getProperty<any>("value");
  }

}

namespace OjRadiosetBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface OptionsKeys {
    /**
     * The key name for the label.
     */
    "label": string;
    /**
     * The key name for the value.
     */
    "value": string;
  }
  export interface Translations {
    /**
     * Used for the readonly radioset when no option is selected.
     */
    "readonlyNoValue": string;
    /**
     * Provides properties to customize the summary, detail and hint text used by the implicit required validator associated to any editable component that supports the required option.
     */
    "required": OjRadiosetBase.TranslationsRequired;
  }
  export interface TranslationsRequired {
    /**
     * Hint text used by required validation error.
     */
    "hint": string;
    /**
     * Message text that describes the details of the required validation error.
     */
    "messageDetail": string;
    /**
     * Message text for summarizing a required validation error.
     */
    "messageSummary": string;
  }
}
