// @oracle/oraclejet-webdriver/elements/oj-select-single/OjSelectSingleBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjSelectSingleBase = void 0;
const __1 = require("../");
class OjSelectSingleBase extends __1.OjWebElement {
    getData() {
        return this.getProperty("data");
    }
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getDisplayOptions() {
        return this.getProperty("displayOptions");
    }
    getHelp() {
        return this.getProperty("help");
    }
    getHelpHints() {
        return this.getProperty("helpHints");
    }
    getItemText() {
        return this.getProperty("itemText");
    }
    getLabelEdge() {
        return this.getProperty("labelEdge");
    }
    getLabelHint() {
        return this.getProperty("labelHint");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    changeMessagesCustom(messagesCustom) {
        return this.setProperty("messagesCustom", messagesCustom);
    }
    getMessagesCustom() {
        return this.getProperty("messagesCustom");
    }
    getPlaceholder() {
        return this.getProperty("placeholder");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getRequired() {
        return this.getProperty("required");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUserAssistanceDensity() {
        return this.getProperty("userAssistanceDensity");
    }
    getValid() {
        return this.getProperty("valid");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
    changeValueItem(valueItem) {
        return this.setProperty("valueItem", valueItem);
    }
    getValueItem() {
        return this.getProperty("valueItem");
    }
    getVirtualKeyboard() {
        return this.getProperty("virtualKeyboard");
    }
}
exports.OjSelectSingleBase = OjSelectSingleBase;
