// @oracle/oraclejet-webdriver/elements/oj-select-single/OjSelectSingleBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjSelectSingle, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjSelectSingleBase extends OjWebElement {
  /**
   * Gets the value of data property.
   * The data for the Select.
   * @return The value of data property.
   * 
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjSelectSingleBase.DisplayOptions> {
    return this.getProperty<OjSelectSingleBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjSelectSingleBase.Help> {
    return this.getProperty<OjSelectSingleBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjSelectSingleBase.HelpHints> {
    return this.getProperty<OjSelectSingleBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of itemText property.
   * Specifies the text string to render for a data item.
   * @return The value of itemText property.
   * 
   */
  public getItemText(): Promise<string|null> {
    return this.getProperty<string|null>("itemText");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * The oj-label sets the labelledBy property programmatically on the form component.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of placeholder property.
   * A short hint that can be displayed before user selects or enters a value.
   * @return The value of placeholder property.
   * 
   */
  public getPlaceholder(): Promise<string> {
    return this.getProperty<string>("placeholder");
  }

  /**
   * Gets the value of readonly property.
   * Specifies whether a value is readonly
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of required property.
   * Specifies whether a value is required.
   * @return The value of required property.
   * 
   */
  public getRequired(): Promise<boolean> {
    return this.getProperty<boolean>("required");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjSelectSingleBase.Translations|null> {
    return this.getProperty<OjSelectSingleBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Sets the value of "value" property.
   * The value of the element.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: null|any): Promise<void> {
    return this.setProperty<null|any>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the element.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<null|any> {
    return this.getProperty<null|any>("value");
  }

  /**
   * Sets the value of "valueItem" property.
   * The current value of the element and its associated data.
   * @param valueItem The value to set for "valueItem"
   * 
   */
  public changeValueItem(valueItem: null|object): Promise<void> {
    return this.setProperty<null|object>("valueItem", valueItem);
  }

  /**
   * Gets the value of valueItem property.
   * The current value of the element and its associated data.
   * @return The value of valueItem property.
   * 
   */
  public getValueItem(): Promise<null|object> {
    return this.getProperty<null|object>("valueItem");
  }

  /**
   * Gets the value of virtualKeyboard property.
   * The type of virtual keyboard to display for entering a value on mobile browsers. See the Help documentation for more information.
   * @return The value of virtualKeyboard property.
   * 
   */
  public getVirtualKeyboard(): Promise<string> {
    return this.getProperty<string>("virtualKeyboard");
  }

}

namespace OjSelectSingleBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text. The supported attribute values are theme dependent.
     */
    "messages": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface Translations {
    /**
     * Label of the cancel button of the mobile dropdown.
     */
    "cancel": string;
    /**
     * Label of the clear value icon for screen readers.
     */
    "labelAccClearValue": string;
    /**
     * Label of the dropdown arrow for screen readers.
     */
    "labelAccOpenDropdown": string;
    /**
     * Screen reader text for the drop down when multiple results are found.
     */
    "multipleMatchesFound": string;
    /**
     * Screen reader text for the drop down when N or more results are found.
     */
    "nOrMoreMatchesFound": string;
    /**
     * Screen reader text for the drop down when no results are found.
     */
    "noMatchesFound": string;
    /**
     * First line of text in the no data template shown in the default listView.
     */
    "noResultsLine1": string;
    /**
     * Second line of text in the no data template shown in the default listView.
     */
    "noResultsLine2": string;
    /**
     * Screen reader text for the drop down when one result is found.
     */
    "oneMatchFound": string;
    /**
     * Provides properties to customize the summary, detail and hint text used by the implicit required validator associated to any editable component that supports the required option.
     */
    "required": OjSelectSingleBase.TranslationsRequired;
  }
  export interface TranslationsRequired {
    /**
     * Hint text used by required validation error.
     */
    "hint": string;
    /**
     * Message text that describes the details of the required validation error.
     */
    "messageDetail": string;
    /**
     * Message text for summarizing a required validation error.
     */
    "messageSummary": string;
  }
}
