// @oracle/oraclejet-webdriver/elements/oj-dialog/OjDialogBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjDialog, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjDialogBase extends OjWebElement {
  /**
   * Gets the value of cancelBehavior property.
   * Specifies the cancel behavior of the dialog. See the Help documentation for more information.
   * @return The value of cancelBehavior property.
   * 
   */
  public getCancelBehavior(): Promise<string> {
    return this.getProperty<string>("cancelBehavior");
  }

  /**
   * Gets the value of dialogTitle property.
   * Specify the title of the dialog. null is the default.
   * @return The value of dialogTitle property.
   * 
   */
  public getDialogTitle(): Promise<string|null> {
    return this.getProperty<string|null>("dialogTitle");
  }

  /**
   * Gets the value of dragAffordance property.
   * Specifies the drag affordance. See the Help documentation for more information.
   * @return The value of dragAffordance property.
   * 
   */
  public getDragAffordance(): Promise<string> {
    return this.getProperty<string>("dragAffordance");
  }

  /**
   * Gets the value of initialVisibility property.
   * Set the initial visibility of the dialog.
   * @return The value of initialVisibility property.
   * 
   */
  public getInitialVisibility(): Promise<string> {
    return this.getProperty<string>("initialVisibility");
  }

  /**
   * Gets the value of modality property.
   * Defines the modality of the dialog.
   * @return The value of modality property.
   * 
   */
  public getModality(): Promise<string> {
    return this.getProperty<string>("modality");
  }

  /**
   * Gets the value of position property.
   * Establishes the location that the dialog will appear relative to another element. See the Help documentation for more information.
   * @return The value of position property.
   * 
   */
  public getPosition(): Promise<OjDialogBase.Position> {
    return this.getProperty<OjDialogBase.Position>("position");
  }

  /**
   * Gets the value of resizeBehavior property.
   * Specifies the resizeBehavior of the dialog. See the Help documentation for more information.
   * @return The value of resizeBehavior property.
   * 
   */
  public getResizeBehavior(): Promise<string> {
    return this.getProperty<string>("resizeBehavior");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjDialogBase.Translations|null> {
    return this.getProperty<OjDialogBase.Translations|null>("translations");
  }

}

namespace OjDialogBase {
  export interface Position {
    /**
     * Defines which position on the target element ("of") to align the positioned element against.
     */
    "at": OjDialogBase.PositionAt;
    /**
     * Rule for alternate alignment.
     */
    "collision": string;
    /**
     * Defines which edge on the dialog to align with the target ("of") element.
     */
    "my": OjDialogBase.PositionMy;
    /**
     * Specifies which element to position the dialog against. See the Help documentation for more information.
     */
    "of": string|object;
    /**
     * Defines a point offset in pixels from the ("my") alignment.
     */
    "offset": OjDialogBase.PositionOffset;
  }
  export interface PositionAt {
    /**
     * 
     */
    "horizontal": string;
    /**
     * 
     */
    "vertical": string;
  }
  export interface PositionMy {
    /**
     * 
     */
    "horizontal": string;
    /**
     * 
     */
    "vertical": string;
  }
  export interface PositionOffset {
    /**
     * Horizontal alignment offset.
     */
    "x": number;
    /**
     * Vertical alignment offset.
     */
    "y": number;
  }
  export interface Translations {
    /**
     * Label for the dialog close button.
     */
    "labelCloseIcon": string;
  }
}
