// @oracle/oraclejet-webdriver/elements/oj-input-password/OjInputPasswordBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjInputPassword, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjInputPasswordBase extends OjWebElement {
  /**
   * Gets the value of asyncValidators property.
   * Specifies a list of asynchronous validators used by the component when performing validation. Use async-validators when you need to perform some validation work on the server. See the Help documentation for more information.
   * @return The value of asyncValidators property.
   * @deprecated Since 8.0.0. Use the validators property instead for either regular Validators or AsyncValidators.
   */
  public getAsyncValidators(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("asyncValidators");
  }

  /**
   * Gets the value of autocomplete property.
   * Specifies a component's autocomplete state. See the Help documentation for more information.
   * @return The value of autocomplete property.
   * 
   */
  public getAutocomplete(): Promise<string> {
    return this.getProperty<string>("autocomplete");
  }

  /**
   * Gets the value of autofocus property.
   * Specifies whether the component will get input focus when the page is loaded. See the Help documentation for more information.
   * @return The value of autofocus property.
   * 
   */
  public getAutofocus(): Promise<boolean> {
    return this.getProperty<boolean>("autofocus");
  }

  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjInputPasswordBase.DisplayOptions> {
    return this.getProperty<OjInputPasswordBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjInputPasswordBase.Help> {
    return this.getProperty<OjInputPasswordBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjInputPasswordBase.HelpHints> {
    return this.getProperty<OjInputPasswordBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * The oj-label sets the labelledBy property programmatically on the form component. See the Help documentation for more information.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of maskIcon property.
   * The mask icon is used to mask or unmask the visibility of the password. The password always starts out masked, but using the mask icon the user can toggle the visibility of the password so that it can be read. The mask-icon attribute is used to make the mask icon visible or hidden. If it is hidden, then the user has no way to toggle the visibility of the password.
   * @return The value of maskIcon property.
   * 
   */
  public getMaskIcon(): Promise<string> {
    return this.getProperty<string>("maskIcon");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of name property.
   * Specifies the name of the component.
   * @return The value of name property.
   * @deprecated Since 6.0.0. JET does not use form submit, so this is not needed.
   */
  public getName(): Promise<string> {
    return this.getProperty<string>("name");
  }

  /**
   * Gets the value of placeholder property.
   * The placeholder text to set on the element.
   * @return The value of placeholder property.
   * 
   */
  public getPlaceholder(): Promise<string> {
    return this.getProperty<string>("placeholder");
  }

  /**
   * Gets the value of rawValue property.
   * Read-only property used for retrieving the current value from the input field in string form. See the Help documentation for more information.
   * @return The value of rawValue property.
   * 
   */
  public getRawValue(): Promise<string> {
    return this.getProperty<string>("rawValue");
  }

  /**
   * Gets the value of readonly property.
   * Specifies whether the component is read-only. A read-only element cannot be modified, but user interaction is allowed. See the Help documentation for more information.
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of required property.
   * Specifies whether the component is required or optional. See the Help documentation for more information.
   * @return The value of required property.
   * 
   */
  public getRequired(): Promise<boolean> {
    return this.getProperty<boolean>("required");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjInputPasswordBase.Translations|null> {
    return this.getProperty<OjInputPasswordBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Gets the value of validators property.
   * Specifies a list of synchronous validators for performing validation by the element. See the Help documentation for more information.
   * @return The value of validators property.
   * 
   */
  public getValidators(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("validators");
  }

  /**
   * Sets the value of "value" property.
   * The value of the component. Value must be a string or null.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: string|null): Promise<void> {
    return this.setProperty<string|null>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the component. Value must be a string or null.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<string|null> {
    return this.getProperty<string|null>("value");
  }

}

namespace OjInputPasswordBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface Translations {
    /**
     * Label for the 'view-hide' mask visibility icon.
     */
    "accessibleHidePassword": string;
    /**
     * Message to announce to AT via aria-live when the length.max is exceeded.
     */
    "accessibleMaxLengthExceeded": string;
    /**
     * Message to annnouce to AT via aria-live for the remaining character count.
     */
    "accessibleMaxLengthRemaining": string;
    /**
     * Label for the 'view' mask visibility icon.
     */
    "accessibleShowPassword": string;
    /**
     * Provides properties to customize the message text used by the implicit regexp validator associated to the InputPassword component.
     */
    "regexp": OjInputPasswordBase.TranslationsRegexp;
    /**
     * Provides properties to customize the summary, detail and hint text used by the implicit required validator associated to any editable component that supports the required option.
     */
    "required": OjInputPasswordBase.TranslationsRequired;
  }
  export interface TranslationsRegexp {
    /**
     * Provides properties to customize the error message detail used by the implicit regexp validator associated to the InputPassword component.
     */
    "messageDetail": string;
    /**
     * Provides properties to customize the error message summary used by the implicit regexp validator associated to the InputText and TextArea components.
     */
    "messageSummary": string;
  }
  export interface TranslationsRequired {
    /**
     * Hint text used by required validation error.
     */
    "hint": string;
    /**
     * Message text that describes the details of the required validation error.
     */
    "messageDetail": string;
    /**
     * Message text for summarizing a required validation error.
     */
    "messageSummary": string;
  }
}
