// @oracle/oraclejet-webdriver/elements/oj-list-view/OjListView.ts
import { OjListViewBase } from "./OjListViewBase";
import { slotProxy, SlotProxy } from "../../";
import { WebElement } from "selenium-webdriver";

/**
 * This is the component WebElement for oj-list-view. Retrieve an instance of the class
 * using its convenience function:
 * ```javascript
 * await ojListView(driver, By.css('oj-list-view'))
 * ```
 */
export class OjListView extends OjListViewBase {
  /**
   * Retrieve a SlotProxy which represents a single listview item.
   * @param key The key within the Collection's dataset associated with the item.
   */
  async findItem<T>(itemLocator: { key: T }): Promise<SlotProxy> {
    // Call changeScrollPosition to scroll the row into view and ensure it's
    // rendere in DOM
    await this.changeScrollPosition({
      key: itemLocator.key,
      // TODO: These should be marked optional by the base class
      index: undefined,
      offsetX: undefined,
      offsetY: undefined,
      parent: undefined,
      x: undefined,
      y: undefined
    });
    const el = await this.getDriver().executeScript<WebElement>(`
    return arguments[0].getNodeBySubId({ subId: 'oj-listview-item', key: ${JSON.stringify(itemLocator.key)} })
      .querySelector('[role=gridcell]')
    `, this);
    return slotProxy(el, this);
  }
}
