// @oracle/oraclejet-webdriver/elements/oj-list-view/OjListViewBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjListViewBase = void 0;
const __1 = require("../");
class OjListViewBase extends __1.OjWebElement {
    getAs() {
        return this.getProperty("as");
    }
    changeCurrentItem(currentItem) {
        return this.setProperty("currentItem", currentItem);
    }
    getCurrentItem() {
        return this.getProperty("currentItem");
    }
    getData() {
        return this.getProperty("data");
    }
    getDisplay() {
        return this.getProperty("display");
    }
    getDnd() {
        return this.getProperty("dnd");
    }
    getDrillMode() {
        return this.getProperty("drillMode");
    }
    getFirstSelectedItem() {
        return this.getProperty("firstSelectedItem");
    }
    getGridlines() {
        return this.getProperty("gridlines");
    }
    getGroupHeaderPosition() {
        return this.getProperty("groupHeaderPosition");
    }
    getItem() {
        return this.getProperty("item");
    }
    getScrollPolicy() {
        return this.getProperty("scrollPolicy");
    }
    getScrollPolicyOptions() {
        return this.getProperty("scrollPolicyOptions");
    }
    changeScrollPosition(scrollPosition) {
        return this.setProperty("scrollPosition", scrollPosition);
    }
    getScrollPosition() {
        return this.getProperty("scrollPosition");
    }
    getScrollToKey() {
        return this.getProperty("scrollToKey");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSelectionRequired() {
        return this.getProperty("selectionRequired");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjListViewBase = OjListViewBase;
