// @oracle/oraclejet-webdriver/elements/oj-list-view/OjListViewBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjListView, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjListViewBase extends OjWebElement {
  /**
   * Gets the value of as property.
   * Specifies the alias for the current item when referenced inside the item template.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Sets the value of "currentItem" property.
   * Specifies the key of the item that should have keyboard focus. See the Help documentation for more information.
   * @param currentItem The value to set for "currentItem"
   * 
   */
  public changeCurrentItem(currentItem: any): Promise<void> {
    return this.setProperty<any>("currentItem", currentItem);
  }

  /**
   * Gets the value of currentItem property.
   * Specifies the key of the item that should have keyboard focus. See the Help documentation for more information.
   * @return The value of currentItem property.
   * 
   */
  public getCurrentItem(): Promise<any> {
    return this.getProperty<any>("currentItem");
  }

  /**
   * Gets the value of data property.
   * Specifies the data for the list. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object> {
    return this.getProperty<object>("data");
  }

  /**
   * Gets the value of display property.
   * Specifies how the items should be displayed.
   * @return The value of display property.
   * 
   */
  public getDisplay(): Promise<string> {
    return this.getProperty<string>("display");
  }

  /**
   * Gets the value of dnd property.
   * Customizes the drag and drop functionality. See the Help documentation for more information.
   * @return The value of dnd property.
   * 
   */
  public getDnd(): Promise<OjListViewBase.Dnd> {
    return this.getProperty<OjListViewBase.Dnd>("dnd");
  }

  /**
   * Gets the value of drillMode property.
   * Specifies whether expand or collapse operations are allowed.
   * @return The value of drillMode property.
   * 
   */
  public getDrillMode(): Promise<string> {
    return this.getProperty<string>("drillMode");
  }

  /**
   * Gets the value of firstSelectedItem property.
   * Specifies the key and data of the first selected item. See the Help documentation for more information.
   * @return The value of firstSelectedItem property.
   * 
   */
  public getFirstSelectedItem(): Promise<object> {
    return this.getProperty<object>("firstSelectedItem");
  }

  /**
   * Gets the value of gridlines property.
   * Customizes the functionality of how gridlines are shown in various parts of the list.
   * @return The value of gridlines property.
   * 
   */
  public getGridlines(): Promise<OjListViewBase.Gridlines> {
    return this.getProperty<OjListViewBase.Gridlines>("gridlines");
  }

  /**
   * Gets the value of groupHeaderPosition property.
   * Specifies whether group header should stick to the top as user scrolls.
   * @return The value of groupHeaderPosition property.
   * 
   */
  public getGroupHeaderPosition(): Promise<string> {
    return this.getProperty<string>("groupHeaderPosition");
  }

  /**
   * Gets the value of item property.
   * Customizes the functionality of each item on the list.
   * @return The value of item property.
   * 
   */
  public getItem(): Promise<OjListViewBase.Item> {
    return this.getProperty<OjListViewBase.Item>("item");
  }

  /**
   * Gets the value of scrollPolicy property.
   * Specifies how data are fetched as user scrolls down the list.
   * @return The value of scrollPolicy property.
   * 
   */
  public getScrollPolicy(): Promise<string|null> {
    return this.getProperty<string|null>("scrollPolicy");
  }

  /**
   * Gets the value of scrollPolicyOptions property.
   * Specifies fetch options for scrolling behaviors that trigger data fetches. See the Help documentation for more information.
   * @return The value of scrollPolicyOptions property.
   * 
   */
  public getScrollPolicyOptions(): Promise<unknown|null> {
    return this.getProperty<unknown|null>("scrollPolicyOptions");
  }

  /**
   * Sets the value of "scrollPosition" property.
   * Specifies the current scroll position of the list. See the Help documentation for more information.
   * @param scrollPosition The value to set for "scrollPosition"
   * 
   */
  public changeScrollPosition(scrollPosition: OjListViewBase.ScrollPosition): Promise<void> {
    return this.setProperty<OjListViewBase.ScrollPosition>("scrollPosition", scrollPosition);
  }

  /**
   * Gets the value of scrollPosition property.
   * Specifies the current scroll position of the list. See the Help documentation for more information.
   * @return The value of scrollPosition property.
   * 
   */
  public getScrollPosition(): Promise<OjListViewBase.ScrollPosition> {
    return this.getProperty<OjListViewBase.ScrollPosition>("scrollPosition");
  }

  /**
   * Gets the value of scrollToKey property.
   * Specifies the behavior when ListView needs to scroll to a position based on an item key.
   * @return The value of scrollToKey property.
   * 
   */
  public getScrollToKey(): Promise<string|null> {
    return this.getProperty<string|null>("scrollToKey");
  }

  /**
   * Sets the value of "selection" property.
   * Specifies the current selections in the list. See the Help documentation for more information.
   * @param selection The value to set for "selection"
   * @deprecated Since 7.0.0. Use selected attribute instead.
   */
  public changeSelection(selection: Array<any>): Promise<void> {
    return this.setProperty<Array<any>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * Specifies the current selections in the list. See the Help documentation for more information.
   * @return The value of selection property.
   * @deprecated Since 7.0.0. Use selected attribute instead.
   */
  public getSelection(): Promise<Array<any>> {
    return this.getProperty<Array<any>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the selection mode.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

  /**
   * Gets the value of selectionRequired property.
   * Specifies whether selection is required on the ListView.
   * @return The value of selectionRequired property.
   * 
   */
  public getSelectionRequired(): Promise<boolean> {
    return this.getProperty<boolean>("selectionRequired");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjListViewBase.Translations|null> {
    return this.getProperty<OjListViewBase.Translations|null>("translations");
  }

}

namespace OjListViewBase {
  export interface Dnd {
    /**
     * Customizes the drag functionality.
     */
    "drag": OjListViewBase.DndDrag;
    /**
     * Customizes the drop functionality.
     */
    "drop": OjListViewBase.DndDrop;
    /**
     * Customizes the item reordering functionality.
     */
    "reorder": OjListViewBase.DndReorder;
  }
  export interface DndDrag {
    /**
     * An object that describes drag functionality for a selected set of items. See the Help documentation for more information.
     */
    "items": OjListViewBase.DndDragItems;
  }
  export interface DndDragItems {
    /**
     * Specifies one or more MIME types to use for the dragged data in the dataTransfer object. See the Help documentation for more information.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDrop {
    /**
     * An object that describes drop functionality for a selected set of items.
     */
    "items": OjListViewBase.DndDropItems;
  }
  export interface DndDropItems {
    /**
     * Specifies one or more data types that this component can accept. See the Help documentation for more information.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndReorder {
    /**
     * Specify the item reordering functionality. See the Help documentation for more information.
     */
    "items": string;
  }
  export interface Gridlines {
    /**
     * Specifies whether the grid lines should be visible.
     */
    "item": string;
  }
  export interface Item {
    /**
     * Specifies whether the item can receive keyboard focus. See the Help documentation for more information.
     */
    "focusable": boolean|null;
    /**
     * Specifies the renderer for the item. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies whether the item can be selected. See the Help documentation for more information.
     */
    "selectable": boolean|null;
  }
  export interface ScrollPosition {
    /**
     * The zero-based index of the item. If scrollPolicy is set to 'loadMoreOnScroll' and the index is greater than maxCount set in scrollPolicyOptions, then it will scroll and fetch until the end of the list is reached and there's no more items to fetch.
     */
    "index": number;
    /**
     * The key of the item. If DataProvider is used for data and the key does not exists in the DataProvider, then the value is ignored. If DataProvider is not used then ListView will fetch and scroll until the item is found or the end of the list is reached and there's no more items to fetch.
     */
    "key": any;
    /**
     * The horizontal offset in pixels relative to the item identified by key/index.
     */
    "offsetX": number;
    /**
     * The vertical offset in pixels relative to the item identified by key/index.
     */
    "offsetY": number;
    /**
     * The key of the parent where the index is relative to. If not specified, then the root is assumed
     */
    "parent": any;
    /**
     * The horizontal position in pixels.
     */
    "x": number;
    /**
     * The vertical position in pixels.
     */
    "y": number;
  }
  export interface Translations {
    /**
     * Provides properties to customize the screen reader text when focus skips a number of items as a result of up/down arrow navigation in card layout mode.
     */
    "accessibleNavigateSkipItems": string;
    /**
     * Provides properties to customize the screen reader text when the tentative drop target is after a certain item.
     */
    "accessibleReorderAfterItem": string;
    /**
     * Provides properties to customize the screen reader text when the tentative drop target is before a certain item.
     */
    "accessibleReorderBeforeItem": string;
    /**
     * Provides properties to customize the screen reader text when the tentative drop target is inside a certain item.
     */
    "accessibleReorderInsideItem": string;
    /**
     * Provides properties to customize the screen reader touch instructional text for reordering items.
     */
    "accessibleReorderTouchInstructionText": string;
    /**
     * Provides properties to customize the characters to display in the Indexer.
     */
    "indexerCharacters": string;
    /**
     * Provides properties to customize the context menu copy label.
     */
    "labelCopy": string;
    /**
     * Provides properties to customize the context menu cut label.
     */
    "labelCut": string;
    /**
     * Provides properties to customize the context menu paste label.
     */
    "labelPaste": string;
    /**
     * Provides properties to customize the context menu paste after label.
     */
    "labelPasteAfter": string;
    /**
     * Provides properties to customize the context menu paste before label.
     */
    "labelPasteBefore": string;
    /**
     * Provides properties to customize the message text used by ListView when waiting for data.
     */
    "msgFetchingData": string;
    /**
     * Provides properties to customize the message text used by ListView when items are appended.
     */
    "msgItemsAppended": string;
    /**
     * Provides properties to customize the message text used by ListView when there are no items.
     */
    "msgNoData": string;
  }
}
