// @oracle/oraclejet-webdriver/elements/oj-popup/OjPopupBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjPopup, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjPopupBase extends OjWebElement {
  /**
   * Gets the value of autoDismiss property.
   * Specifies the auto dismissal behavior.
   * @return The value of autoDismiss property.
   * 
   */
  public getAutoDismiss(): Promise<string> {
    return this.getProperty<string>("autoDismiss");
  }

  /**
   * Gets the value of chrome property.
   * Specifies whether to use the border, shadow, and background colors from the active theme.
   * @return The value of chrome property.
   * 
   */
  public getChrome(): Promise<string> {
    return this.getProperty<string>("chrome");
  }

  /**
   * Gets the value of initialFocus property.
   * Specifies whether the popup steals focus to its content when initially opened.
   * @return The value of initialFocus property.
   * 
   */
  public getInitialFocus(): Promise<string> {
    return this.getProperty<string>("initialFocus");
  }

  /**
   * Gets the value of modality property.
   * Specifies whether the popup should block user input to the page.
   * @return The value of modality property.
   * 
   */
  public getModality(): Promise<string> {
    return this.getProperty<string>("modality");
  }

  /**
   * Gets the value of position property.
   * Specifies the position of a popup when launched. See the Help documentation for more information.
   * @return The value of position property.
   * 
   */
  public getPosition(): Promise<OjPopupBase.Position> {
    return this.getProperty<OjPopupBase.Position>("position");
  }

  /**
   * Gets the value of tail property.
   * Specifies whether to display a decoration pointing from the popup to the launching element. See the Help documentation for more information.
   * @return The value of tail property.
   * 
   */
  public getTail(): Promise<string> {
    return this.getProperty<string>("tail");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjPopupBase.Translations|null> {
    return this.getProperty<OjPopupBase.Translations|null>("translations");
  }

}

namespace OjPopupBase {
  export interface Position {
    /**
     * Defines which position on the target element ("of") to align the positioned element against.
     */
    "at": OjPopupBase.PositionAt;
    /**
     * Rule for alternate alignment.
     */
    "collision": string;
    /**
     * Defines which edge on the popup to align with the target ("of") element.
     */
    "my": OjPopupBase.PositionMy;
    /**
     * Which element to position the popup against. See the Help documentation for more information.
     */
    "of": string|object;
    /**
     * Defines a point offset in pixels from the ("my") alignment.
     */
    "offset": OjPopupBase.PositionOffset;
  }
  export interface PositionAt {
    /**
     * Defines the horizontal alignment of what the popup is aligned to.
     */
    "horizontal": string;
    /**
     * Defines the vertical alignment of what the popup is aligned to.
     */
    "vertical": string;
  }
  export interface PositionMy {
    /**
     * Defines the horizontal alignment of the popup.
     */
    "horizontal": string;
    /**
     * Defines the vertical alignment of the popup.
     */
    "vertical": string;
  }
  export interface PositionOffset {
    /**
     * Horizontal alignment offset.
     */
    "x": number;
    /**
     * Vertical alignment offset.
     */
    "y": number;
  }
  export interface Translations {
    /**
     * Label assigned to a hidden anchor tag used for accessibility navigation on platforms supporting VoiceOver. The link is injected as a sibling to the context area of an open popup and used for ojPopup#close activation.
     */
    "ariaCloseSkipLink": string;
    /**
     * Label assigned to a hidden anchor tag used for accessibility navigation on platforms supporting VoiceOver. The link is injected as a sibling to the launcher of an open popup. The launcher is passed as the first argument to ojPopup#open. Activation of the link will move focus to the content of the associated popup.
     */
    "ariaFocusSkipLink": string;
    /**
     * Message announced to a WAI-ARIA live region on desktop platforms when a popup opens and steals initial focus.
     */
    "ariaLiveRegionInitialFocusFirstFocusable": string;
    /**
     * Message announced to a WAI-ARIA live region on platforms supporting VoiceOver when a popup opens and steals initial focus.
     */
    "ariaLiveRegionInitialFocusFirstFocusableTouch": string;
    /**
     * Message announced to a WAI-ARIA live region on desktop platforms when a popup opens but doesn't change focus.
     */
    "ariaLiveRegionInitialFocusNone": string;
    /**
     * Message announced to a WAI-ARIA live region on platforms supporting VoiceOver when a popup opens but doesn't change focus.
     */
    "ariaLiveRegionInitialFocusNoneTouch": string;
  }
}
