// @oracle/oraclejet-webdriver/elements/oj-menu/OjMenuBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjMenu, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjMenuBase extends OjWebElement {
  /**
   * Gets the value of disabled property.
   * Disables the menu if set to true.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of openOptions property.
   * Specifies settings for launching a menu. See the Help documentation for more information.
   * @return The value of openOptions property.
   * 
   */
  public getOpenOptions(): Promise<OjMenuBase.OpenOptions> {
    return this.getProperty<OjMenuBase.OpenOptions>("openOptions");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjMenuBase.Translations|null> {
    return this.getProperty<OjMenuBase.Translations|null>("translations");
  }

}

namespace OjMenuBase {
  export interface OpenOptions {
    /**
     * Specifies whether the menu displays as a drop down or as a sheet. See the Help documentation for more information.
     */
    "display": string;
    /**
     * Specifies focus behavior when the menu is initially opened.
     */
    "initialFocus": string;
    /**
     * Specifies the DOM node that launches this menu. See the Help documentation for more information.
     */
    "launcher": string;
    /**
     * Specifies the position of a drop down menu when launched. See the Help documentation for more information.
     */
    "position": OjMenuBase.OpenOptionsPosition;
  }
  export interface OpenOptionsPosition {
    /**
     * Defines which position on the target element ("of") to align the positioned element against.
     */
    "at": OjMenuBase.OpenOptionsPositionAt;
    /**
     * Rule for alternate alignment.
     */
    "collision": string;
    /**
     * Defines which edge on the menu to align with the target ("of") element.
     */
    "my": OjMenuBase.OpenOptionsPositionMy;
    /**
     * Which element to position the menu against. See the Help documentation for more information.
     */
    "of": string|object;
    /**
     * Defines a point offset in pixels from the ("my") alignment.
     */
    "offset": OjMenuBase.OpenOptionsPositionOffset;
  }
  export interface OpenOptionsPositionAt {
    /**
     * Defines the horizontal alignment of what the menu is aligned to. See the Help documentation for more information.
     */
    "horizontal": string;
    /**
     * Defines the vertical alignment of what the menu is aligned to. See the Help documentation for more information.
     */
    "vertical": string;
  }
  export interface OpenOptionsPositionMy {
    /**
     * Defines the horizontal alignment of the menu.
     */
    "horizontal": string;
    /**
     * Defines the vertical alignment of the menu.
     */
    "vertical": string;
  }
  export interface OpenOptionsPositionOffset {
    /**
     * Horizontal alignment offset.
     */
    "x": number;
    /**
     * Vertical alignment offset.
     */
    "y": number;
  }
  export interface Translations {
    /**
     * Label assigned to a hidden anchor tag used for accessibility navigation on platforms supporting VoiceOver. The link is injected before the first menu item. It is used for establishing VO cursor focus. Activation of the link will move focus to the first menu item.
     */
    "ariaFocusSkipLink": string;
    /**
     * Label for the "Cancel" menu item.
     */
    "labelCancel": string;
  }
}
