// @oracle/oraclejet-webdriver/elements/oj-input-search/OjInputSearchBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjInputSearch, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjInputSearchBase extends OjWebElement {
  /**
   * Gets the value of suggestions property.
   * Data for the InputSearch suggestions.
   * @return The value of suggestions property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getSuggestions(): Promise<object|null> {
    return this.getProperty<object|null>("suggestions");
  }

  /**
   * Gets the value of suggestionItemText property.
   * Specifies the text string to render for a suggestion.
   * @return The value of suggestionItemText property.
   * 
   */
  public getSuggestionItemText(): Promise<null> {
    return this.getProperty<null>("suggestionItemText");
  }

  /**
   * Gets the value of placeholder property.
   * A short hint that can be displayed before user selects or enters a value.
   * @return The value of placeholder property.
   * 
   */
  public getPlaceholder(): Promise<string> {
    return this.getProperty<string>("placeholder");
  }

  /**
   * Gets the value of rawValue property.
   * Read-only property used for retrieving the current value from the input field in string form.
   * @return The value of rawValue property.
   * 
   */
  public getRawValue(): Promise<string|null> {
    return this.getProperty<string|null>("rawValue");
  }

  /**
   * Sets the value of "value" property.
   * The value of the element.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: string|null): Promise<void> {
    return this.setProperty<string|null>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the element.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<string|null> {
    return this.getProperty<string|null>("value");
  }

}
