// @oracle/oraclejet-webdriver/elements/oj-status-meter-gauge/OjStatusMeterGaugeBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjStatusMeterGaugeBase = void 0;
const __1 = require("../");
class OjStatusMeterGaugeBase extends __1.OjWebElement {
    getAngleExtent() {
        return this.getProperty("angleExtent");
    }
    getAnimationDuration() {
        return this.getProperty("animationDuration");
    }
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getBorderColor() {
        return this.getProperty("borderColor");
    }
    getBorderRadius() {
        return this.getProperty("borderRadius");
    }
    getCenter() {
        return this.getProperty("center");
    }
    getColor() {
        return this.getProperty("color");
    }
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getIndicatorSize() {
        return this.getProperty("indicatorSize");
    }
    getInnerRadius() {
        return this.getProperty("innerRadius");
    }
    getLabel() {
        return this.getProperty("label");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    getMax() {
        return this.getProperty("max");
    }
    getMetricLabel() {
        return this.getProperty("metricLabel");
    }
    getMin() {
        return this.getProperty("min");
    }
    getOrientation() {
        return this.getProperty("orientation");
    }
    getPlotArea() {
        return this.getProperty("plotArea");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getReferenceLines() {
        return this.getProperty("referenceLines");
    }
    getStartAngle() {
        return this.getProperty("startAngle");
    }
    getStep() {
        return this.getProperty("step");
    }
    getSvgClassName() {
        return this.getProperty("svgClassName");
    }
    getSvgStyle() {
        return this.getProperty("svgStyle");
    }
    getThresholdDisplay() {
        return this.getProperty("thresholdDisplay");
    }
    getThresholds() {
        return this.getProperty("thresholds");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTransientValue() {
        return this.getProperty("transientValue");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
    getVisualEffects() {
        return this.getProperty("visualEffects");
    }
}
exports.OjStatusMeterGaugeBase = OjStatusMeterGaugeBase;
