// @oracle/oraclejet-webdriver/elements/oj-status-meter-gauge/OjStatusMeterGaugeBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjStatusMeterGauge, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjStatusMeterGaugeBase extends OjWebElement {
  /**
   * Gets the value of angleExtent property.
   * Specifies the angle extent of a gauge with circular orientation. Value should be provided in degrees.
   * @return The value of angleExtent property.
   * 
   */
  public getAngleExtent(): Promise<number> {
    return this.getProperty<number>("angleExtent");
  }

  /**
   * Gets the value of animationDuration property.
   * The duration of the animations in milliseconds.
   * @return The value of animationDuration property.
   * 
   */
  public getAnimationDuration(): Promise<number> {
    return this.getProperty<number>("animationDuration");
  }

  /**
   * Gets the value of animationOnDataChange property.
   * Defines the animation that is applied on data changes.
   * @return The value of animationOnDataChange property.
   * 
   */
  public getAnimationOnDataChange(): Promise<string> {
    return this.getProperty<string>("animationOnDataChange");
  }

  /**
   * Gets the value of animationOnDisplay property.
   * Defines the animation that is shown on initial display.
   * @return The value of animationOnDisplay property.
   * 
   */
  public getAnimationOnDisplay(): Promise<string> {
    return this.getProperty<string>("animationOnDisplay");
  }

  /**
   * Gets the value of borderColor property.
   * The border color of the gauge. Only applies when thresholds are not defined. The default value varies based on theme.
   * @return The value of borderColor property.
   * 
   */
  public getBorderColor(): Promise<string> {
    return this.getProperty<string>("borderColor");
  }

  /**
   * Gets the value of borderRadius property.
   * Specifies the border radius of the indicator and plot area. See the Help documentation for more information.
   * @return The value of borderRadius property.
   * 
   */
  public getBorderRadius(): Promise<string> {
    return this.getProperty<string>("borderRadius");
  }

  /**
   * Gets the value of center property.
   * An object defining the center content of a status meter with circular orientation.
   * @return The value of center property.
   * 
   */
  public getCenter(): Promise<OjStatusMeterGaugeBase.Center> {
    return this.getProperty<OjStatusMeterGaugeBase.Center>("center");
  }

  /**
   * Gets the value of color property.
   * The color of the gauge. Only applies when thresholds are not defined. The default value varies based on theme.
   * @return The value of color property.
   * 
   */
  public getColor(): Promise<string> {
    return this.getProperty<string>("color");
  }

  /**
   * Gets the value of describedBy property.
   * Specifies a relationship between this component and another element.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string|null> {
    return this.getProperty<string|null>("describedBy");
  }

  /**
   * Gets the value of indicatorSize property.
   * Defines the ratio of relative thickness of the indicator to the plot area.
   * @return The value of indicatorSize property.
   * 
   */
  public getIndicatorSize(): Promise<number> {
    return this.getProperty<number>("indicatorSize");
  }

  /**
   * Gets the value of innerRadius property.
   * Specifies the inner radius of a gauge with circular orientation, defined by the distance from the center of the gauge to the innermost edge of the indicator and plot area. Valid values are a percent or ratio from 0 to 1.
   * @return The value of innerRadius property.
   * 
   */
  public getInnerRadius(): Promise<number> {
    return this.getProperty<number>("innerRadius");
  }

  /**
   * Gets the value of label property.
   * An object defining the label.
   * @return The value of label property.
   * 
   */
  public getLabel(): Promise<OjStatusMeterGaugeBase.Label> {
    return this.getProperty<OjStatusMeterGaugeBase.Label>("label");
  }

  /**
   * Gets the value of labelledBy property.
   * Establishes a relationship between this component and another element, typically an oj-label custom element. See the Help documentation for more information.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of max property.
   * The maximum value of the gauge.
   * @return The value of max property.
   * 
   */
  public getMax(): Promise<number> {
    return this.getProperty<number>("max");
  }

  /**
   * Gets the value of metricLabel property.
   * An object defining the value label.
   * @return The value of metricLabel property.
   * 
   */
  public getMetricLabel(): Promise<OjStatusMeterGaugeBase.MetricLabel> {
    return this.getProperty<OjStatusMeterGaugeBase.MetricLabel>("metricLabel");
  }

  /**
   * Gets the value of min property.
   * The minimum value of the gauge.
   * @return The value of min property.
   * 
   */
  public getMin(): Promise<number> {
    return this.getProperty<number>("min");
  }

  /**
   * Gets the value of orientation property.
   * Defines the type of status meter to be rendered.
   * @return The value of orientation property.
   * 
   */
  public getOrientation(): Promise<string> {
    return this.getProperty<string>("orientation");
  }

  /**
   * Gets the value of plotArea property.
   * Plot Area for Status Meter Gauge
   * @return The value of plotArea property.
   * 
   */
  public getPlotArea(): Promise<OjStatusMeterGaugeBase.PlotArea> {
    return this.getProperty<OjStatusMeterGaugeBase.PlotArea>("plotArea");
  }

  /**
   * Gets the value of readonly property.
   * Defines whether the value of the gauge can be changed by the end user.
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of referenceLines property.
   * An array of objects specifying the reference lines for the gauge.
   * @return The value of referenceLines property.
   * 
   */
  public getReferenceLines(): Promise<Array<OjStatusMeterGaugeBase.ReferenceLines>> {
    return this.getProperty<Array<OjStatusMeterGaugeBase.ReferenceLines>>("referenceLines");
  }

  /**
   * Gets the value of startAngle property.
   * Specifies the start angle of a gauge with circular orientation. Value should be provided in degrees.
   * @return The value of startAngle property.
   * 
   */
  public getStartAngle(): Promise<number> {
    return this.getProperty<number>("startAngle");
  }

  /**
   * Gets the value of step property.
   * Specifies the increment by which values can be changed by the end user. See the Help documentation for more information.
   * @return The value of step property.
   * 
   */
  public getStep(): Promise<number|null> {
    return this.getProperty<number|null>("step");
  }

  /**
   * Gets the value of svgClassName property.
   * The CSS style class to apply to the gauge indicator. See the Help documentation for more information.
   * @return The value of svgClassName property.
   * 
   */
  public getSvgClassName(): Promise<string> {
    return this.getProperty<string>("svgClassName");
  }

  /**
   * Gets the value of svgStyle property.
   * The inline style to apply to the gauge indicator. See the Help documentation for more information.
   * @return The value of svgStyle property.
   * 
   */
  public getSvgStyle(): Promise<object> {
    return this.getProperty<object>("svgStyle");
  }

  /**
   * Gets the value of thresholdDisplay property.
   * Controls whether the current threshold is displayed on the indicator, in the plotArea, or if all the thresholds are displayed in the plot area.
   * @return The value of thresholdDisplay property.
   * 
   */
  public getThresholdDisplay(): Promise<string> {
    return this.getProperty<string>("thresholdDisplay");
  }

  /**
   * Gets the value of thresholds property.
   * An array of objects specifying the gauge thresholds.
   * @return The value of thresholds property.
   * 
   */
  public getThresholds(): Promise<Array<OjStatusMeterGaugeBase.Thresholds>> {
    return this.getProperty<Array<OjStatusMeterGaugeBase.Thresholds>>("thresholds");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjStatusMeterGaugeBase.Tooltip> {
    return this.getProperty<OjStatusMeterGaugeBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of transientValue property.
   * Read-only property used for retrieving the transient value from the component. See the Help documentation for more information.
   * @return The value of transientValue property.
   * 
   */
  public getTransientValue(): Promise<number|null> {
    return this.getProperty<number|null>("transientValue");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjStatusMeterGaugeBase.Translations|null> {
    return this.getProperty<OjStatusMeterGaugeBase.Translations|null>("translations");
  }

  /**
   * Sets the value of "value" property.
   * The metric value.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: number|null): Promise<void> {
    return this.setProperty<number|null>("value", value);
  }

  /**
   * Gets the value of value property.
   * The metric value.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<number|null> {
    return this.getProperty<number|null>("value");
  }

  /**
   * Gets the value of visualEffects property.
   * Defines whether visual effects such as overlays are applied to the gauge.
   * @return The value of visualEffects property.
   * 
   */
  public getVisualEffects(): Promise<string> {
    return this.getProperty<string>("visualEffects");
  }

}

namespace OjStatusMeterGaugeBase {
  export interface Center {
  }
  export interface Label {
    /**
     * Specifies the label position for horizontal and vertical gauges. See the Help documentation for more information.
     */
    "position": string;
    /**
     * The CSS style object to apply to the label. The following style properties are supported: color, cursor, fontFamily, fontSize, fontStyle, fontWeight, textDecoration.
     */
    "style": object;
    /**
     * The text for the label.
     */
    "text": string;
  }
  export interface MetricLabel {
    /**
     * The converter used to format the labels. See the Help documentation for more information.
     */
    "converter": object;
    /**
     * Specifies the metric label position for horizontal and vertical gauges. See the Help documentation for more information.
     */
    "position": string;
    /**
     * Defines if the label is rendered. If set to auto, the label is rendered if the orientation is circular.
     */
    "rendered": string;
    /**
     * The scaling behavior of the labels. When using a converter, scaling should be set to none, as the formatted result may not be compatible with the scaling suffixes.
     */
    "scaling": string;
    /**
     * The CSS style object to apply to the label.
     */
    "style": object;
    /**
     * The text for the label. If specified, text will overwrite the numeric value that is displayed by default. See the Help documentation for more information.
     */
    "text": string;
    /**
     * Defines whether the label is a number or a percentage of the total value.
     */
    "textType": string;
  }
  export interface PlotArea {
    /**
     * The border color of the plot area.
     */
    "borderColor": string;
    /**
     * Specifies the border radius of the plot area shape. See the Help documentation for more information.
     */
    "borderRadius": string;
    /**
     * The color of the plot area. Only applies when useThresholdFillColor is off.
     */
    "color": string;
    /**
     * Specifies whether to render the plot area. See the Help documentation for more information.
     */
    "rendered": string;
    /**
     * The CSS style class to apply to the plot area. See the Help documentation for more information.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the plot area. See the Help documentation for more information.
     */
    "svgStyle": object;
  }
  export interface ReferenceLines {
    /**
     * The color of the reference line.
     */
    "color": string;
    /**
     * The line style of the reference line.
     */
    "lineStyle": string;
    /**
     * The width of the reference line.
     */
    "lineWidth": number;
    /**
     * The value of the reference line.
     */
    "value": number;
  }
  export interface Thresholds {
    /**
     * The border color of the threshold.
     */
    "borderColor": string;
    /**
     * The color of the threshold.
     */
    "color": string;
    /**
     * The upper bound of the threshold. This value is ignored for the final threshold, which uses the maximum value of the gauge.
     */
    "max": number;
    /**
     * Specific description for the threshold and overwrites the shortDesc specified on gauge. This is used for accessibility and also for customizing the tooltip text.
     */
    "shortDesc": string;
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
  }
}
