// @oracle/oraclejet-webdriver/elements/oj-file-picker/OjFilePickerBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjFilePicker, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjFilePickerBase extends OjWebElement {
  /**
   * Gets the value of accept property.
   * An array of strings of allowed MIME types or file extensions that can be uploaded. If not specified, accept all file types
   * @return The value of accept property.
   * 
   */
  public getAccept(): Promise<Array<string>|null> {
    return this.getProperty<Array<string>|null>("accept");
  }

  /**
   * Gets the value of capture property.
   * Specifies the preferred facing mode for the device's media capture mechanism.
   * @return The value of capture property.
   * 
   */
  public getCapture(): Promise<string|null> {
    return this.getProperty<string|null>("capture");
  }

  /**
   * Gets the value of disabled property.
   * Disables the filepicker if set to true
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of primaryText property.
   * The primary text for the default file picker.
   * @return The value of primaryText property.
   * 
   */
  public getPrimaryText(): Promise<string|null> {
    return this.getProperty<string|null>("primaryText");
  }

  /**
   * Gets the value of secondaryText property.
   * The secondary text for the default file picker.
   * @return The value of secondaryText property.
   * 
   */
  public getSecondaryText(): Promise<string|null> {
    return this.getProperty<string|null>("secondaryText");
  }

  /**
   * Gets the value of selectOn property.
   * The type of event to select the files.
   * @return The value of selectOn property.
   * 
   */
  public getSelectOn(): Promise<string> {
    return this.getProperty<string>("selectOn");
  }

  /**
   * Gets the value of selectionMode property.
   * Whether to allow single or multiple file selection.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

}
