// @oracle/oraclejet-webdriver/elements/oj-buttonset-many/OjButtonsetManyBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjButtonsetMany, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjButtonsetManyBase extends OjWebElement {
  /**
   * Gets the value of chroming property.
   * Indicates in what states the buttonset has chrome (background and border).
   * @return The value of chroming property.
   * 
   */
  public getChroming(): Promise<string> {
    return this.getProperty<string>("chroming");
  }

  /**
   * Gets the value of describedBy property.
   * buttonset's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies that the buttonset element should be disabled.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of display property.
   * Specifies whether the buttonset displays label and icons, or just icons.
   * @return The value of display property.
   * 
   */
  public getDisplay(): Promise<string> {
    return this.getProperty<string>("display");
  }

  /**
   * Gets the value of focusManagement property.
   * Should be set to none when the buttonset is placed within a toolbar.
   * @return The value of focusManagement property.
   * 
   */
  public getFocusManagement(): Promise<string> {
    return this.getProperty<string>("focusManagement");
  }

  /**
   * Gets the value of labelledBy property.
   * Establishes a relationship between this component and another element, typically an oj-label custom element. See the Help documentation for more information.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<object|null> {
    return this.getProperty<object|null>("translations");
  }

  /**
   * Sets the value of "value" property.
   * Indicates which oj-options in the buttonset are selected.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: Array<any>|null): Promise<void> {
    return this.setProperty<Array<any>|null>("value", value);
  }

  /**
   * Gets the value of value property.
   * Indicates which oj-options in the buttonset are selected.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<Array<any>|null> {
    return this.getProperty<Array<any>|null>("value");
  }

}
