// @oracle/oraclejet-webdriver/elements/oj-collapsible/OjCollapsibleBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjCollapsible, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjCollapsibleBase extends OjWebElement {
  /**
   * Gets the value of disabled property.
   * Disables the collapsible if set to true.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of expandArea property.
   * Where in the header to click to toggle disclosure.
   * @return The value of expandArea property.
   * 
   */
  public getExpandArea(): Promise<string> {
    return this.getProperty<string>("expandArea");
  }

  /**
   * Sets the value of "expanded" property.
   * Specifies if the content is expanded.
   * @param expanded The value to set for "expanded"
   * 
   */
  public changeExpanded(expanded: boolean): Promise<void> {
    return this.setProperty<boolean>("expanded", expanded);
  }

  /**
   * Gets the value of expanded property.
   * Specifies if the content is expanded.
   * @return The value of expanded property.
   * 
   */
  public getExpanded(): Promise<boolean> {
    return this.getProperty<boolean>("expanded");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<object|null> {
    return this.getProperty<object|null>("translations");
  }

}
