// @oracle/oraclejet-webdriver/elements/oj-date-time-picker/OjDateTimePickerBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjDateTimePickerBase = void 0;
const __1 = require("../");
class OjDateTimePickerBase extends __1.OjWebElement {
    getAsyncValidators() {
        return this.getProperty("asyncValidators");
    }
    getAutofocus() {
        return this.getProperty("autofocus");
    }
    getConverter() {
        return this.getProperty("converter");
    }
    getDatePicker() {
        return this.getProperty("datePicker");
    }
    getDayFormatter() {
        return this.getProperty("dayFormatter");
    }
    getDayMetaData() {
        return this.getProperty("dayMetaData");
    }
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getDisplayOptions() {
        return this.getProperty("displayOptions");
    }
    getHelp() {
        return this.getProperty("help");
    }
    getHelpHints() {
        return this.getProperty("helpHints");
    }
    getKeyboardEdit() {
        return this.getProperty("keyboardEdit");
    }
    getLabelEdge() {
        return this.getProperty("labelEdge");
    }
    getLabelHint() {
        return this.getProperty("labelHint");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    getMax() {
        return this.getProperty("max");
    }
    changeMessagesCustom(messagesCustom) {
        return this.setProperty("messagesCustom", messagesCustom);
    }
    getMessagesCustom() {
        return this.getProperty("messagesCustom");
    }
    getMin() {
        return this.getProperty("min");
    }
    getName() {
        return this.getProperty("name");
    }
    getPickerAttributes() {
        return this.getProperty("pickerAttributes");
    }
    getPlaceholder() {
        return this.getProperty("placeholder");
    }
    getRawValue() {
        return this.getProperty("rawValue");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getRenderMode() {
        return this.getProperty("renderMode");
    }
    getRequired() {
        return this.getProperty("required");
    }
    getTimePicker() {
        return this.getProperty("timePicker");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUserAssistanceDensity() {
        return this.getProperty("userAssistanceDensity");
    }
    getValid() {
        return this.getProperty("valid");
    }
    getValidators() {
        return this.getProperty("validators");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
}
exports.OjDateTimePickerBase = OjDateTimePickerBase;
