// @oracle/oraclejet-webdriver/elements/oj-slider/OjSliderBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjSlider, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjSliderBase extends OjWebElement {
  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjSliderBase.DisplayOptions> {
    return this.getProperty<OjSliderBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjSliderBase.Help> {
    return this.getProperty<OjSliderBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjSliderBase.HelpHints> {
    return this.getProperty<OjSliderBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * The oj-label sets the labelledBy property programmatically on the form component.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of max property.
   * The maximum value of the slider. See the Help documentation for more information.
   * @return The value of max property.
   * 
   */
  public getMax(): Promise<number> {
    return this.getProperty<number>("max");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of min property.
   * The minimum value of the slider. See the Help documentation for more information.
   * @return The value of min property.
   * 
   */
  public getMin(): Promise<number> {
    return this.getProperty<number>("min");
  }

  /**
   * Gets the value of orientation property.
   * Specifies the orientation of the slider.
   * @return The value of orientation property.
   * 
   */
  public getOrientation(): Promise<string> {
    return this.getProperty<string>("orientation");
  }

  /**
   * Gets the value of step property.
   * Specifies the amount to increase or decrease the value when moving in step increments. See the Help documentation for more information.
   * @return The value of step property.
   * 
   */
  public getStep(): Promise<number> {
    return this.getProperty<number>("step");
  }

  /**
   * Gets the value of transientValue property.
   * Read-only property used for retrieving the transient value from the component. See the Help documentation for more information.
   * @return The value of transientValue property.
   * 
   */
  public getTransientValue(): Promise<number> {
    return this.getProperty<number>("transientValue");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjSliderBase.Translations|null> {
    return this.getProperty<OjSliderBase.Translations|null>("translations");
  }

  /**
   * Gets the value of type property.
   * The slider type specifies how the slider value is represented.
   * @return The value of type property.
   * 
   */
  public getType(): Promise<string> {
    return this.getProperty<string>("type");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Sets the value of "value" property.
   * The numerical value of the slider.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: number): Promise<void> {
    return this.setProperty<number>("value", value);
  }

  /**
   * Gets the value of value property.
   * The numerical value of the slider.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<number> {
    return this.getProperty<number>("value");
  }

}

namespace OjSliderBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface Translations {
    /**
     * Invalid step value.
     */
    "invalidStep": string;
    /**
     * Max is greater than min.
     */
    "maxMin": string;
    /**
     * Missing value exception.
     */
    "noValue": string;
    /**
     * The option is not a number.
     */
    "optionNum": string;
    /**
     * Value outside min to max range.
     */
    "valueRange": string;
  }
}
