// @oracle/oraclejet-webdriver/elements/oj-sunburst/OjSunburstBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjSunburstBase = void 0;
const __1 = require("../");
class OjSunburstBase extends __1.OjWebElement {
    getAnimationDuration() {
        return this.getProperty("animationDuration");
    }
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAnimationUpdateColor() {
        return this.getProperty("animationUpdateColor");
    }
    getAs() {
        return this.getProperty("as");
    }
    getColorLabel() {
        return this.getProperty("colorLabel");
    }
    getData() {
        return this.getProperty("data");
    }
    getDisplayLevels() {
        return this.getProperty("displayLevels");
    }
    getDrilling() {
        return this.getProperty("drilling");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    getHighlightMode() {
        return this.getProperty("highlightMode");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getHoverBehaviorDelay() {
        return this.getProperty("hoverBehaviorDelay");
    }
    getNodeDefaults() {
        return this.getProperty("nodeDefaults");
    }
    getNodes() {
        return this.getProperty("nodes");
    }
    getRootNode() {
        return this.getProperty("rootNode");
    }
    getRootNodeContent() {
        return this.getProperty("rootNodeContent");
    }
    getRotation() {
        return this.getProperty("rotation");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSizeLabel() {
        return this.getProperty("sizeLabel");
    }
    getSorting() {
        return this.getProperty("sorting");
    }
    changeStartAngle(startAngle) {
        return this.setProperty("startAngle", startAngle);
    }
    getStartAngle() {
        return this.getProperty("startAngle");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTouchResponse() {
        return this.getProperty("touchResponse");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjSunburstBase = OjSunburstBase;
