// @oracle/oraclejet-webdriver/elements/oj-sunburst/OjSunburstBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjSunburst, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjSunburstBase extends OjWebElement {
  /**
   * Gets the value of animationDuration property.
   * Specifies the animation duration in milliseconds. For data change animations with multiple stages, the value specifies the duration of each stage.
   * @return The value of animationDuration property.
   * 
   */
  public getAnimationDuration(): Promise<number> {
    return this.getProperty<number>("animationDuration");
  }

  /**
   * Gets the value of animationOnDataChange property.
   * Specifies the animation that is applied on data changes.
   * @return The value of animationOnDataChange property.
   * 
   */
  public getAnimationOnDataChange(): Promise<string> {
    return this.getProperty<string>("animationOnDataChange");
  }

  /**
   * Gets the value of animationOnDisplay property.
   * Specifies the animation that is shown on initial display.
   * @return The value of animationOnDisplay property.
   * 
   */
  public getAnimationOnDisplay(): Promise<string> {
    return this.getProperty<string>("animationOnDisplay");
  }

  /**
   * Gets the value of animationUpdateColor property.
   * Specifies the color displayed during a data change animation when a node is updated.
   * @return The value of animationUpdateColor property.
   * 
   */
  public getAnimationUpdateColor(): Promise<string> {
    return this.getProperty<string>("animationUpdateColor");
  }

  /**
   * Gets the value of as property.
   * An alias for the '$current' context variable passed to slot content for the nodeTemplate slot.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Gets the value of colorLabel property.
   * Specifies the label displayed in the legend describing the color metric of the sunburst.
   * @return The value of colorLabel property.
   * 
   */
  public getColorLabel(): Promise<string> {
    return this.getProperty<string>("colorLabel");
  }

  /**
   * Gets the value of data property.
   * Specifies the DataProvider for the sunburst. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of displayLevels property.
   * The number of levels of nodes to display. By default all nodes are displayed.
   * @return The value of displayLevels property.
   * 
   */
  public getDisplayLevels(): Promise<number> {
    return this.getProperty<number>("displayLevels");
  }

  /**
   * Gets the value of drilling property.
   * Specifies whether drilling is enabled. Drillable nodes will show a pointer cursor on hover and fire ojBeforeDrill and ojDrill events on click (double click if selection is enabled). See the Help documentation for more information.
   * @return The value of drilling property.
   * 
   */
  public getDrilling(): Promise<string> {
    return this.getProperty<string>("drilling");
  }

  /**
   * Sets the value of "hiddenCategories" property.
   * An array of category strings used for filtering. Nodes with any category matching an item in this array will be filtered.
   * @param hiddenCategories The value to set for "hiddenCategories"
   * 
   */
  public changeHiddenCategories(hiddenCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("hiddenCategories", hiddenCategories);
  }

  /**
   * Gets the value of hiddenCategories property.
   * An array of category strings used for filtering. Nodes with any category matching an item in this array will be filtered.
   * @return The value of hiddenCategories property.
   * 
   */
  public getHiddenCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("hiddenCategories");
  }

  /**
   * Gets the value of highlightMatch property.
   * The matching condition for the highlightedCategories property. See the Help documentation for more information.
   * @return The value of highlightMatch property.
   * 
   */
  public getHighlightMatch(): Promise<string> {
    return this.getProperty<string>("highlightMatch");
  }

  /**
   * Gets the value of highlightMode property.
   * Specifies whether the highlight-mode should be by category or descendants.
   * @return The value of highlightMode property.
   * 
   */
  public getHighlightMode(): Promise<string> {
    return this.getProperty<string>("highlightMode");
  }

  /**
   * Sets the value of "highlightedCategories" property.
   * An array of category strings used for highlighting. Nodes matching categories in this array will be highlighted.
   * @param highlightedCategories The value to set for "highlightedCategories"
   * 
   */
  public changeHighlightedCategories(highlightedCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("highlightedCategories", highlightedCategories);
  }

  /**
   * Gets the value of highlightedCategories property.
   * An array of category strings used for highlighting. Nodes matching categories in this array will be highlighted.
   * @return The value of highlightedCategories property.
   * 
   */
  public getHighlightedCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("highlightedCategories");
  }

  /**
   * Gets the value of hoverBehavior property.
   * Defines the behavior applied when hovering over the nodes.
   * @return The value of hoverBehavior property.
   * 
   */
  public getHoverBehavior(): Promise<string> {
    return this.getProperty<string>("hoverBehavior");
  }

  /**
   * Gets the value of hoverBehaviorDelay property.
   * Specifies initial hover delay in milliseconds for highlighting nodes.
   * @return The value of hoverBehaviorDelay property.
   * 
   */
  public getHoverBehaviorDelay(): Promise<number> {
    return this.getProperty<number>("hoverBehaviorDelay");
  }

  /**
   * Gets the value of nodeDefaults property.
   * An object defining default properties for the nodes. See the Help documentation for more information.
   * @return The value of nodeDefaults property.
   * 
   */
  public getNodeDefaults(): Promise<OjSunburstBase.NodeDefaults> {
    return this.getProperty<OjSunburstBase.NodeDefaults>("nodeDefaults");
  }

  /**
   * Gets the value of nodes property.
   * An array of objects defining the data for the nodes. Also accepts a Promise for deferred data rendering.
   * @return The value of nodes property.
   * 
   */
  public getNodes(): Promise<Array<OjSunburstBase.Nodes>|null|null> {
    return this.getProperty<Array<OjSunburstBase.Nodes>|null|null>("nodes");
  }

  /**
   * Gets the value of rootNode property.
   * The id of the root node. When specified, only the root node and children of the root will be displayed.
   * @return The value of rootNode property.
   * 
   */
  public getRootNode(): Promise<any> {
    return this.getProperty<any>("rootNode");
  }

  /**
   * Gets the value of rootNodeContent property.
   * An object defining custom root node content for the sunburst.
   * @return The value of rootNodeContent property.
   * 
   */
  public getRootNodeContent(): Promise<OjSunburstBase.RootNodeContent> {
    return this.getProperty<OjSunburstBase.RootNodeContent>("rootNodeContent");
  }

  /**
   * Gets the value of rotation property.
   * Specifies whether client side rotation is enabled.
   * @return The value of rotation property.
   * 
   */
  public getRotation(): Promise<string> {
    return this.getProperty<string>("rotation");
  }

  /**
   * Sets the value of "selection" property.
   * An array containing the ids of the initially selected nodes.
   * @param selection The value to set for "selection"
   * 
   */
  public changeSelection(selection: Array<any>): Promise<void> {
    return this.setProperty<Array<any>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * An array containing the ids of the initially selected nodes.
   * @return The value of selection property.
   * 
   */
  public getSelection(): Promise<Array<any>> {
    return this.getProperty<Array<any>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the selection behavior on the sunburst. See the Help documentation for more information.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

  /**
   * Gets the value of sizeLabel property.
   * Specifies the label used in the legend describing the size metric of the sunburst.
   * @return The value of sizeLabel property.
   * 
   */
  public getSizeLabel(): Promise<string> {
    return this.getProperty<string>("sizeLabel");
  }

  /**
   * Gets the value of sorting property.
   * Specifies whether the nodes are sorted by size. Nodes with the same parent are sorted in order of descending size.
   * @return The value of sorting property.
   * 
   */
  public getSorting(): Promise<string> {
    return this.getProperty<string>("sorting");
  }

  /**
   * Sets the value of "startAngle" property.
   * Specifies the starting angle of the sunburst. Valid values are numbers between 0 and 360.
   * @param startAngle The value to set for "startAngle"
   * 
   */
  public changeStartAngle(startAngle: number): Promise<void> {
    return this.setProperty<number>("startAngle", startAngle);
  }

  /**
   * Gets the value of startAngle property.
   * Specifies the starting angle of the sunburst. Valid values are numbers between 0 and 360.
   * @return The value of startAngle property.
   * 
   */
  public getStartAngle(): Promise<number> {
    return this.getProperty<number>("startAngle");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjSunburstBase.Tooltip> {
    return this.getProperty<OjSunburstBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of touchResponse property.
   * Specifies configuration options for touch and hold delays on mobile devices. See the Help documentation for more information.
   * @return The value of touchResponse property.
   * 
   */
  public getTouchResponse(): Promise<string> {
    return this.getProperty<string>("touchResponse");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjSunburstBase.Translations|null> {
    return this.getProperty<OjSunburstBase.Translations|null>("translations");
  }

}

namespace OjSunburstBase {
  export interface NodeDefaults {
    /**
     * The default border color of the nodes.
     */
    "borderColor": string;
    /**
     * The default border width of the nodes.
     */
    "borderWidth": number;
    /**
     * The color of the node hover feedback.
     */
    "hoverColor": string;
    /**
     * The label display behavior for the nodes. See the Help documentation for more information.
     */
    "labelDisplay": string;
    /**
     * The horizontal alignment for labels displayed within the node. Only applies to rotated text.
     */
    "labelHalign": string;
    /**
     * The minimum number of visible characters needed to render a truncated label. See the Help documentation for more information.
     */
    "labelMinLength": number;
    /**
     * The CSS style object defining the style of the label.
     */
    "labelStyle": object;
    /**
     * The inner color of the node selection feedback.
     */
    "selectedInnerColor": string;
    /**
     * The outer color of the node selection feedback.
     */
    "selectedOuterColor": string;
    /**
     * Specifies whether to display the expand/collapse button on hover.
     */
    "showDisclosure": string;
  }
  export interface Nodes {
    /**
     * The border color of the node.
     */
    "borderColor": string;
    /**
     * The border width of the node.
     */
    "borderWidth": number;
    /**
     * An optional array of category strings corresponding to this data item. This enables highlighting and filtering of individual data items through interactions with the legend and other visualization elements. The categories array of each node is required to be a superset of the categories array of its parent node. If not specified, the ids of the node and its ancestors will be used.
     */
    "categories": Array<string>;
    /**
     * The fill color of the node.
     */
    "color": string;
    /**
     * Specifies whether drilling is enabled for the node. Drillable nodes will show a pointer cursor on hover and fire an ojDrill event on click (double click if selection is enabled). To enable drilling for all nodes at once, use the drilling attribute in the top level.
     */
    "drilling": string;
    /**
     * The id of the node. For the DataProvider case, the key for the node will be used as the id.
     */
    "id": any;
    /**
     * The label for this node.
     */
    "label": string;
    /**
     * The label display behavior for the nodes. More labels are generally displayed when using rotation, with the trade off of readability. When auto is used, rotated or horizontal labels will be used based on the client browser and platform.
     */
    "labelDisplay": string;
    /**
     * The horizontal alignment for labels displayed within the node. Only applies to rotated text.
     */
    "labelHalign": string;
    /**
     * The CSS style object defining the style of the label. The following style properties are supported: color, cursor, fontFamily, fontSize, fontStyle, fontWeight, textDecoration. The default value comes from the CSS and varies based on theme. The default value comes from the CSS and varies based on theme.
     */
    "labelStyle": object;
    /**
     * An array of objects with properties for the child nodes.
     */
    "nodes": Array<object>;
    /**
     * The pattern used to fill the node.
     */
    "pattern": string;
    /**
     * The radius of the node relative to the other nodes.
     */
    "radius": number;
    /**
     * Specifies whether or not the node will be selectable.
     */
    "selectable": string;
    /**
     * The description of this element. Will be lazily created if a function is used. This is used for accessibility and also for customizing the tooltip text.
     */
    "shortDesc": string|null;
    /**
     * Specifies whether to display the expand/collapse button on hover for a specific node. If the button is clicked, the expanded attribute is updated with the new array of node ids.
     */
    "showDisclosure": string;
    /**
     * The CSS style class to apply to the node. The style class and inline style will override any other styling specified through the properties. For tooltip interactivity, it's recommended to also pass a representative color to the node color attribute.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the node. The style class and inline style will override any other styling specified through the properties. For tooltip interactivity, it's recommended to also pass a representative color to the node color attribute. Only SVG CSS style properties are supported. The default value comes from the CSS and varies based on theme.
     */
    "svgStyle": object;
    /**
     * The value of the node, which determines the relative size of the node.
     */
    "value": number;
  }
  export interface RootNodeContent {
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used for the color label.
     */
    "labelColor": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used for the size label.
     */
    "labelSize": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
    /**
     * Used for the collapse tooltip.
     */
    "tooltipCollapse": string;
    /**
     * Used for the expand tooltip.
     */
    "tooltipExpand": string;
  }
}
