// @oracle/oraclejet-webdriver/elements/oj-thematic-map/OjThematicMapBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjThematicMapBase = void 0;
const __1 = require("../");
class OjThematicMapBase extends __1.OjWebElement {
    getAnimationDuration() {
        return this.getProperty("animationDuration");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAreaData() {
        return this.getProperty("areaData");
    }
    getAreas() {
        return this.getProperty("areas");
    }
    getAs() {
        return this.getProperty("as");
    }
    getFocusRenderer() {
        return this.getProperty("focusRenderer");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getHoverRenderer() {
        return this.getProperty("hoverRenderer");
    }
    getInitialZooming() {
        return this.getProperty("initialZooming");
    }
    getIsolatedItem() {
        return this.getProperty("isolatedItem");
    }
    getLabelDisplay() {
        return this.getProperty("labelDisplay");
    }
    getLabelType() {
        return this.getProperty("labelType");
    }
    getLinkData() {
        return this.getProperty("linkData");
    }
    getLinks() {
        return this.getProperty("links");
    }
    getMapProvider() {
        return this.getProperty("mapProvider");
    }
    getMarkerData() {
        return this.getProperty("markerData");
    }
    getMarkerZoomBehavior() {
        return this.getProperty("markerZoomBehavior");
    }
    getMarkers() {
        return this.getProperty("markers");
    }
    getMaxZoom() {
        return this.getProperty("maxZoom");
    }
    getPanning() {
        return this.getProperty("panning");
    }
    getRenderer() {
        return this.getProperty("renderer");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSelectionRenderer() {
        return this.getProperty("selectionRenderer");
    }
    getStyleDefaults() {
        return this.getProperty("styleDefaults");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTooltipDisplay() {
        return this.getProperty("tooltipDisplay");
    }
    getTouchResponse() {
        return this.getProperty("touchResponse");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getZooming() {
        return this.getProperty("zooming");
    }
}
exports.OjThematicMapBase = OjThematicMapBase;
