// @oracle/oraclejet-webdriver/elements/oj-tree-view/OjTreeView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTreeView = void 0;
const OjTreeViewBase_1 = require("./OjTreeViewBase");
const __1 = require("../../");
const selenium_webdriver_1 = require("selenium-webdriver");
class OjTreeView extends OjTreeViewBase_1.OjTreeViewBase {
    async findItem(itemLocator) {
        let item;
        if (itemLocator.key) {
            await this.whenBusyContextReady();
            item = await this.getDriver().executeAsyncScript(`
      var element = arguments[0];
      var key = arguments[1];
      var done = arguments[2];
      require([\'ojs/ojcore\'], function(oj) {
        const items = element.querySelectorAll('.oj-treeview-item');
        let item;
        for (var i = 0; i < items.length; i++) {
          if (oj.KeyUtils.equals(items[i]['oj-item-metadata'].key, key)) {
            item = items[i];
            break;
          }
        }
        if (item) {
         done(item.querySelector('.oj-treeview-item-content'));
        };
      });`, this, itemLocator.key);
        }
        else if (itemLocator.id) {
            await this.whenBusyContextReady();
            item = await this.getDriver().executeScript(`
      const item = document.getElementById(arguments[0]);
      if (item) {
        return item.querySelector('.oj-treeview-item-content');
      }`, itemLocator.id);
        }
        if (item) {
            return __1.slotProxy(item, this);
        }
        throw Error('No element found for given locator');
    }
    async changeExpanded(expanded) {
        await this.whenReady();
        await this.getDriver().executeScript(`
        var ele = arguments[0];
        var expanded  = arguments[1];
        require([\'ojs/ojkeyset\'], function(keySet) {
          var expandedKeySet = new keySet.ExpandedKeySet(expanded);
          ele.expanded = expandedKeySet;
        });`, this, expanded);
        await this.whenBusyContextReady();
    }
    async doExpand(itemLocator) {
        let itemContent = await this.findItem(itemLocator);
        let isExpanded = await this._isExpanded(itemContent);
        if (!isExpanded) {
            const ins = await this._getDisclosureIcon(itemContent);
            return ins.click();
        }
    }
    async doCollapse(itemLocator) {
        let itemContent = await this.findItem(itemLocator);
        let isExpanded = await this._isExpanded(itemContent);
        if (isExpanded) {
            const ins = await this._getDisclosureIcon(itemContent);
            return ins.click();
        }
    }
    async _isExpanded(itemContent) {
        const ins = await this._getDisclosureIcon(itemContent);
        const driver = await ins.getDriver();
        return driver.executeScript("return arguments[0].parentNode.parentNode.classList.contains('oj-expanded')", ins);
    }
    async _getDisclosureIcon(itemContent) {
        return itemContent.findElement(selenium_webdriver_1.By.css('INS'));
    }
}
exports.OjTreeView = OjTreeView;
