// @oracle/oraclejet-webdriver/elements/oj-tree-view/OjTreeViewBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjTreeView, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjTreeViewBase extends OjWebElement {
  /**
   * Gets the value of currentItem property.
   * Read-only property used for retrieving the key of the item that currently has focus.
   * @return The value of currentItem property.
   * 
   */
  public getCurrentItem(): Promise<any> {
    return this.getProperty<any>("currentItem");
  }

  /**
   * Gets the value of data property.
   * Specifies the data for the tree. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object> {
    return this.getProperty<object>("data");
  }

  /**
   * Gets the value of dnd property.
   * Customizes the drag and drop functionality. See the Help documentation for more information.
   * @return The value of dnd property.
   * 
   */
  public getDnd(): Promise<OjTreeViewBase.Dnd> {
    return this.getProperty<OjTreeViewBase.Dnd>("dnd");
  }

  /**
   * Gets the value of item property.
   * Customizes the functionality of each item in the tree.
   * @return The value of item property.
   * 
   */
  public getItem(): Promise<OjTreeViewBase.Item> {
    return this.getProperty<OjTreeViewBase.Item>("item");
  }

  /**
   * Gets the value of scrollPolicyOptions property.
   * Specifies fetch options for scrolling behaviors that trigger data fetches. See the Help documentation for more information.
   * @return The value of scrollPolicyOptions property.
   * 
   */
  public getScrollPolicyOptions(): Promise<unknown|null> {
    return this.getProperty<unknown|null>("scrollPolicyOptions");
  }

  /**
   * Sets the value of "selection" property.
   * Specifies the current selections in the tree. An empty array indicates nothing is selected.
   * @param selection The value to set for "selection"
   * @deprecated Since 8.0.0. Use selected attribute instead.
   */
  public changeSelection(selection: Array<any>): Promise<void> {
    return this.setProperty<Array<any>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * Specifies the current selections in the tree. An empty array indicates nothing is selected.
   * @return The value of selection property.
   * @deprecated Since 8.0.0. Use selected attribute instead.
   */
  public getSelection(): Promise<Array<any>> {
    return this.getProperty<Array<any>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the selection mode.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjTreeViewBase.Translations|null> {
    return this.getProperty<OjTreeViewBase.Translations|null>("translations");
  }

}

namespace OjTreeViewBase {
  export interface Dnd {
    /**
     * Customizes the drag functionality.
     */
    "drag": OjTreeViewBase.DndDrag;
    /**
     * Customizes the drop functionality.
     */
    "drop": OjTreeViewBase.DndDrop;
  }
  export interface DndDrag {
    /**
     * If this object is specified, TreeView will initiate drag operation when the user drags on an item.
     */
    "items": OjTreeViewBase.DndDragItems;
  }
  export interface DndDragItems {
    /**
     * The MIME types to use for the dragged data in the dataTransfer object. This can be a string if there is only one type, or an array of strings if multiple types are needed.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDrop {
    /**
     * An object that specifies callback functions to handle dropping items
     */
    "items": OjTreeViewBase.DndDropItems;
  }
  export interface DndDropItems {
    /**
     * A data type or an array of data types this component can accept.
     */
    "dataTypes": string|Array<string>;
  }
  export interface Item {
  }
  export interface Translations {
    /**
     * Aria Label for each treeview row selector
     */
    "treeViewSelectorAria": string;
  }
}
