// @oracle/oraclejet-webdriver/elements/oj-button/OjButtonBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjButton, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjButtonBase extends OjWebElement {
  /**
   * Gets the value of disabled property.
   * Specifies that the button element should be disabled.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of display property.
   * Specifies whether the button displays label and icons, or just icons.
   * @return The value of display property.
   * 
   */
  public getDisplay(): Promise<string> {
    return this.getProperty<string>("display");
  }

  /**
   * Gets the value of label property.
   * Text to show in the button. The default is from the label in the DOM.
   * @return The value of label property.
   * 
   */
  public getLabel(): Promise<string> {
    return this.getProperty<string>("label");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<object|null> {
    return this.getProperty<object|null>("translations");
  }

  /**
   * Gets the value of chroming property.
   * Indicates in what states the button has chrome (background and border).
   * @return The value of chroming property.
   * 
   */
  public getChroming(): Promise<string> {
    return this.getProperty<string>("chroming");
  }

}
