// @oracle/oraclejet-webdriver/elements/oj-waterfall-layout/OjWaterfallLayoutBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjWaterfallLayout, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjWaterfallLayoutBase extends OjWebElement {
  /**
   * Gets the value of data property.
   * Specifies the data for the component. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of scrollPolicy property.
   * Specifies how data are fetched as user scrolls towards the bottom of the grid.
   * @return The value of scrollPolicy property.
   * 
   */
  public getScrollPolicy(): Promise<string> {
    return this.getProperty<string>("scrollPolicy");
  }

  /**
   * Gets the value of scrollPolicyOptions property.
   * Specifies fetch options for scrolling behaviors that trigger data fetches. See the Help documentation for more information.
   * @return The value of scrollPolicyOptions property.
   * 
   */
  public getScrollPolicyOptions(): Promise<OjWaterfallLayoutBase.ScrollPolicyOptions> {
    return this.getProperty<OjWaterfallLayoutBase.ScrollPolicyOptions>("scrollPolicyOptions");
  }

  /**
   * Sets the value of "scrollPosition" property.
   * Specifies the current scroll position of the WaterfallLayout. See the Help documentation for more information.
   * @param scrollPosition The value to set for "scrollPosition"
   * 
   */
  public changeScrollPosition(scrollPosition: OjWaterfallLayoutBase.ScrollPosition): Promise<void> {
    return this.setProperty<OjWaterfallLayoutBase.ScrollPosition>("scrollPosition", scrollPosition);
  }

  /**
   * Gets the value of scrollPosition property.
   * Specifies the current scroll position of the WaterfallLayout. See the Help documentation for more information.
   * @return The value of scrollPosition property.
   * 
   */
  public getScrollPosition(): Promise<OjWaterfallLayoutBase.ScrollPosition> {
    return this.getProperty<OjWaterfallLayoutBase.ScrollPosition>("scrollPosition");
  }

}

namespace OjWaterfallLayoutBase {
  export interface ScrollPolicyOptions {
    /**
     * The number of items to fetch in each block.
     */
    "fetchSize": number;
    /**
     * The maximum total number of items to fetch.
     */
    "maxCount": number;
    /**
     * The element or a CSS selector string to an element which WaterfallLayout uses to determine the scroll position as well as the maximum scroll position.
     */
    "scroller": string|null;
  }
  export interface ScrollPosition {
    /**
     * The vertical position in pixels.
     */
    "y": number;
    /**
     * The key of the item. If DataProvider is used for data and the key does not exist in the DataProvider or if the item has not been fetched yet, then the value is ignored.
     */
    "key": any;
    /**
     * The vertical offset in pixels relative to the item identified by key.
     */
    "offsetY": number;
  }
}
