// @oracle/oraclejet-webdriver/elements/oj-progress/OjProgressBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjProgress, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjProgressBase extends OjWebElement {
  /**
   * Gets the value of max property.
   * The maximum allowed value.
   * @return The value of max property.
   * 
   */
  public getMax(): Promise<number> {
    return this.getProperty<number>("max");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjProgressBase.Translations|null> {
    return this.getProperty<OjProgressBase.Translations|null>("translations");
  }

  /**
   * Gets the value of type property.
   * The shape of the Progress.
   * @return The value of type property.
   * 
   */
  public getType(): Promise<string> {
    return this.getProperty<string>("type");
  }

  /**
   * Sets the value of "value" property.
   * The value of the Progress.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: number): Promise<void> {
    return this.setProperty<number>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the Progress.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<number> {
    return this.getProperty<number>("value");
  }

}

namespace OjProgressBase {
  export interface Translations {
    /**
     * Accessible text read by a screen reader while an indeterminate oj-progress is in progress.
     */
    "ariaIndeterminateProgressText": string;
  }
}
