// @oracle/oraclejet-webdriver/lib/driver-override.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const selenium_webdriver_1 = require("selenium-webdriver");
const OjWebElement_1 = require("../elements/OjWebElement");
const stale_element_1 = require("./stale-element");
function register(tagName, ctor) {
    WebElementRegistry.register(tagName, ctor);
}
exports.register = register;
let lastIsJetPage = false;
class WebElementRegistry {
    static register(tagName, ctor) {
        WebElementRegistry.tagToCtorMap[tagName] = ctor;
    }
    static get(tagName) {
        return WebElementRegistry.tagToCtorMap[tagName];
    }
}
WebElementRegistry.tagToCtorMap = {};
selenium_webdriver_1.WebDriver.prototype.findElement = wrapFindElement(selenium_webdriver_1.WebDriver.prototype.findElement);
selenium_webdriver_1.WebDriver.prototype.findElements = wrapFindElements(selenium_webdriver_1.WebDriver.prototype.findElements);
selenium_webdriver_1.WebElement.prototype.findElement = wrapFindElement(selenium_webdriver_1.WebElement.prototype.findElement);
selenium_webdriver_1.WebElement.prototype.findElements = wrapFindElements(selenium_webdriver_1.WebElement.prototype.findElements);
selenium_webdriver_1.WebDriver.prototype.executeScript = wrapExecuteScript(selenium_webdriver_1.WebDriver.prototype.executeScript);
selenium_webdriver_1.WebDriver.prototype.executeAsyncScript = wrapExecuteScript(selenium_webdriver_1.WebDriver.prototype.executeAsyncScript);
function wrapFindElement(origFunc) {
    return function findElement(locator) {
        const isDriver = this instanceof selenium_webdriver_1.WebDriver;
        const driver = isDriver ? this : this.getDriver();
        return new selenium_webdriver_1.WebElementPromise(driver, new Promise(async (resolve, reject) => {
            try {
                const isJet = isDriver && await isJetPage(this) || lastIsJetPage;
                const el = await origFunc.call(this, locator);
                const parentEl = !isDriver && this;
                try {
                    resolve(isJet ? await createTestAdapter(el, locator, parentEl) : el);
                }
                catch (ex) {
                    resolve(findElement.call(this, locator));
                }
            }
            catch (ex) {
                reject(ex);
            }
        }));
    };
}
function wrapFindElements(origFunc) {
    return async function findElements(locator) {
        const isDriver = this instanceof selenium_webdriver_1.WebDriver;
        const isJet = isDriver && await isJetPage(this) || lastIsJetPage;
        const elRef = !isDriver && this;
        const matches = await origFunc.call(this, locator);
        let weMatches = [];
        if (isJet) {
            for (let el, i = 0, len = matches.length; i < len; i++) {
                el = matches[i];
                try {
                    const ce = await createTestAdapter(el, locator, elRef, i);
                    weMatches.push(ce);
                }
                catch (ex) {
                    return findElements.call(this, locator);
                }
            }
        }
        else {
            weMatches = matches;
        }
        return weMatches;
    };
}
function wrapExecuteScript(origFunc) {
    return async function (...args) {
        var _a;
        try {
            const ret = await origFunc.apply(this, args);
            return ret;
        }
        catch (ex) {
            if (ex instanceof selenium_webdriver_1.error.StaleElementReferenceError) {
                const newArgs = [];
                const locators = [];
                for (let i = 0, len = args.length; i < len; i++) {
                    let arg = args[i];
                    if (arg instanceof OjWebElement_1.OjWebElement) {
                        arg = await stale_element_1.requeryElement(arg);
                        locators.push((_a = arg._locators) === null || _a === void 0 ? void 0 : _a.elLocator);
                    }
                    newArgs.push(arg);
                }
                console.warn(`Trapped stale element executing script on ${locators}`);
                return origFunc.apply(this, newArgs);
            }
            throw ex;
        }
    };
}
async function isJetPage(driver) {
    try {
        lastIsJetPage = await driver.executeAsyncScript(`
    const done = arguments[0];
    require("ojs/ojcontext").getPageContext().getBusyContext()
    .whenReady().then(done)
    .catch(ex => done(true))
    `);
    }
    catch (ex) {
        if (ex instanceof selenium_webdriver_1.error.TimeoutError || ex instanceof selenium_webdriver_1.error.ScriptTimeoutError) {
            lastIsJetPage = true;
            throw ex;
        }
        lastIsJetPage = false;
    }
    return lastIsJetPage;
}
async function createTestAdapter(el, elLocator, parentEl, elementsIndex) {
    const tagName = await el.getTagName();
    const ctor = WebElementRegistry.get(tagName) || OjWebElement_1.OjWebElement;
    return new ctor(el, {
        elLocator,
        elementsIndex,
        parentEl,
        tagName,
    });
}
