// @oracle/oraclejet-webdriver/lib/driver-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DriverManager = void 0;
const selenium_webdriver_1 = require("selenium-webdriver");
const defaultConfigName = "";
const defaultCapabilities = new selenium_webdriver_1.Capabilities({
    autoWebview: true,
    browserName: "chrome",
});
const defaultTimeouts = {
    implicit: 0,
    pageLoad: 30000,
    script: 30000,
};
const configs = {};
let currentInstance = {
    configName: null,
    driver: null,
};
const quitDelay = 500;
let builder;
let releaseTimer = setTimeout(() => true, 0);
let quitPromise = Promise.resolve();
class DriverManager {
    static setBuilder(b) {
        if (builder) {
            throw Error("DriverManager Builder instance has already been set, and cannot be set again");
        }
        builder = b;
    }
    static registerConfig(config, name) {
        configs[name || defaultConfigName] = config;
    }
    static getDriver(configName) {
        clearTimeout(releaseTimer);
        configName = configName || defaultConfigName;
        if (configName === currentInstance.configName) {
            return Promise.resolve(currentInstance.driver);
        }
        quitCurrentDriver();
        return createDriver(configName);
    }
    static async getCurrentDriver() {
        return DriverManager.getDriver(currentInstance.configName);
    }
    static releaseDriver(driver) {
        clearTimeout(releaseTimer);
        releaseTimer = setTimeout(quitCurrentDriver, quitDelay);
    }
}
exports.DriverManager = DriverManager;
DriverManager.registerConfig({ capabilities: defaultCapabilities, timeouts: defaultTimeouts });
function createDriver(configName) {
    const config = configs[configName];
    if (!config) {
        throw Error(`No driver configuration exists for "${configName}"`);
    }
    const caps = new selenium_webdriver_1.Capabilities(configs[defaultConfigName].capabilities);
    if (config.capabilities) {
        caps.merge(config.capabilities);
    }
    const timeouts = Object.assign({}, configs[defaultConfigName].timeouts, config.timeouts);
    if (!builder) {
        builder = new selenium_webdriver_1.Builder();
    }
    return quitPromise.then(async () => {
        const driver = builder.withCapabilities(caps).build();
        currentInstance = { configName, driver };
        await driver.manage().setTimeouts(timeouts);
        return driver;
    });
}
function quitCurrentDriver() {
    const driver = currentInstance.driver;
    if (driver) {
        currentInstance = {
            configName: null,
            driver: null,
        };
        quitPromise = driver.quit()
            .catch(ex => console.warn(ex));
    }
}
