// @oracle/oraclejet-webdriver/lib/ScreenshotManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.set = exports.get = exports.create = void 0;
const fs = require("fs");
const path = require("path");
class NoOpScreenShotManager {
    beforeTest(suiteName, testName) { }
    afterTest(passed) { }
    takeAndSaveScreenShot(screenshotTaker, name) { return null; }
}
class TakeScreenshotManager {
    constructor(screenshotRoot) {
        this._screenshotLocation = "";
        var mochawesomeReportDir = screenshotRoot;
        try {
            if (!fs.existsSync(mochawesomeReportDir))
                fs.mkdirSync(mochawesomeReportDir);
        }
        catch (err) {
            console.error(err);
            console.error(err.stack);
        }
        this._screenshotsDir = mochawesomeReportDir + path.sep + "screenshots";
        try {
            if (!fs.existsSync(this._screenshotsDir))
                fs.mkdirSync(this._screenshotsDir);
        }
        catch (err) {
            console.error(err);
            console.error(err.stack);
        }
    }
    replaceCharacters(name) {
        var output = name.replace(/[^a-z0-9]/ig, '_');
        return output;
    }
    beforeTest(suiteName, testName) {
        var suiteN = this.replaceCharacters(suiteName);
        var testN = this.replaceCharacters(testName);
        var suiteF = this._screenshotsDir + path.sep + suiteN;
        var testF = suiteF + path.sep + testN;
        this._screenshotLocation = testF;
        try {
            if (!fs.existsSync(suiteF))
                fs.mkdirSync(suiteF);
        }
        catch (err) {
            console.error(err);
            console.error(err.stack);
        }
        try {
            if (!fs.existsSync(testF))
                fs.mkdirSync(testF);
        }
        catch (err) {
            console.error(err);
            console.error(err.stack);
        }
    }
    afterTest(passed) {
        if (passed) {
            try {
                var FOLDER_PATH = this._screenshotLocation;
                if (fs.existsSync(FOLDER_PATH)) {
                    var files = fs.readdirSync(FOLDER_PATH);
                    files.forEach(element => {
                        fs.unlinkSync(FOLDER_PATH + path.sep + element);
                    });
                    fs.rmdirSync(FOLDER_PATH);
                }
            }
            catch (err) {
                console.error(err);
                console.error(err.stack);
            }
        }
    }
    async takeAndSaveScreenShot(screenshotTaker, name) {
        var counter = 1;
        var FOLDER_PATH = this._screenshotLocation;
        if (fs.existsSync(FOLDER_PATH)) {
            var files = fs.readdirSync(FOLDER_PATH);
            files.forEach(element => {
                if (element.includes(name))
                    counter++;
            });
        }
        try {
            const screenshotImg = await screenshotTaker.takeScreenshot();
            fs.writeFileSync(this._screenshotLocation + path.sep + name + counter + ".jpg", screenshotImg, 'base64');
        }
        catch (err) {
            console.error(err);
            console.error(err.stack);
        }
        return null;
    }
}
var currentInstance = new NoOpScreenShotManager();
function create(screenshotRoot) {
    if (screenshotRoot) {
        return new TakeScreenshotManager(screenshotRoot);
    }
    else {
        return new NoOpScreenShotManager();
    }
}
exports.create = create;
function get() { return currentInstance; }
exports.get = get;
function set(screenshotManager) { currentInstance = screenshotManager; }
exports.set = set;
