// @oracle/oraclejet-webdriver/lib/slot-proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.slotProxy = void 0;
const selenium_webdriver_1 = require("selenium-webdriver");
const elements_1 = require("../elements");
function slotProxy(el, parent) {
    if (!el) {
        throw Error('No root element given for SlotProxy');
    }
    return new SlotProxyImpl(el, parent);
}
exports.slotProxy = slotProxy;
class SlotProxyImpl {
    constructor(el, parent) {
        this._root = el;
        this._parent = parent;
    }
    getDriver() {
        return this._root.getDriver();
    }
    findElement(locator) {
        return this._root.findElement(locator);
    }
    findElements(locator) {
        return this._root.findElements(locator);
    }
    async doContextualAction(action) {
        if (!this._parent) {
            throw Error('Slot Proxy must be initialized with a parent to do a contextual action');
        }
        this._touchSupported = await this.isTouchScreenSupported();
        await this.scrollIntoView();
        const doesSwipeActionExist = await this.doesSwipeActionExist();
        if (doesSwipeActionExist) {
            const ojSwipeAction = await this.getSwipeAction();
            return ojSwipeAction.doAction(action);
        }
        const menu = await this.locateMenu();
        if (menu) {
            await this.openMenu(menu);
            const menuOption = await this.findMenuOptionByValue(action, menu);
            return this.doMenuAction(menuOption, menu);
        }
        throw Error('No Context menu configured for this host component.');
    }
    async scrollIntoView() {
        if (this._touchSupported) {
            const touchActions = new selenium_webdriver_1.TouchSequence(this.getDriver());
            const rootLocation = await this._root.getLocation();
            return touchActions.tapAndHold({ x: 0, y: 0 }).move(rootLocation).release(rootLocation).perform();
        }
        else {
            const actions = this.getDriver().actions();
            if (actions.mouseMove) {
                return actions.mouseMove(this._root).perform();
            }
            else if (actions.move) {
                return actions.move({ origin: this._root }).perform();
            }
        }
    }
    async isTouchScreenSupported() {
        if (!this._touchSupportCalled) {
            this._touchSupported = await this.getDriver().executeAsyncScript(`
     const done  = arguments[0];
     require([\`ojs/ojdomutils\`], function(DomUtils) {
        done(DomUtils.isTouchSupported());
     })`);
            this._touchSupportCalled = true;
        }
        return this._touchSupported;
    }
    async doesSwipeActionExist() {
        const swipeActions = await this._root.findElements(selenium_webdriver_1.By.css('OJ-SWIPE-ACTIONS'));
        return swipeActions.length > 0;
    }
    async getSwipeAction() {
        return elements_1.ojSwipeActions(this._root, selenium_webdriver_1.By.css('OJ-SWIPE-ACTIONS'));
    }
    async locateMenu() {
        const doesClosedMenuExist = await this.doesClosedMenuExist();
        if (doesClosedMenuExist) {
            return elements_1.ojMenu(this._parent, selenium_webdriver_1.By.css('oj-menu[slot=contextMenu]'));
        }
        else {
            const doesOpenMenuExist = await this.doesOpenMenuExist();
            if (doesOpenMenuExist) {
                const menuId = await this.getOpenMenuId();
                const menu = await elements_1.ojMenu(this.getDriver(), selenium_webdriver_1.By.id(menuId));
                await this.closeMenu(menu);
                return menu;
            }
        }
    }
    async closeMenu(menu) {
        return this.getDriver().executeScript(`
     const menu = arguments[0];
     menu.close();
  `, menu);
    }
    async doesClosedMenuExist() {
        const menus = await this._parent.findElements(selenium_webdriver_1.By.css('oj-menu[slot=contextMenu]'));
        return menus.length > 0;
    }
    async getOpenMenuId() {
        const surrogate = await this._parent.findElement(selenium_webdriver_1.By.css('OJ-SURROGATE'));
        const surrogateId = await surrogate.getAttribute('id');
        const lastIndex = surrogateId.lastIndexOf("_layer_surrogate");
        return surrogateId.substring(0, lastIndex);
    }
    async doesOpenMenuExist() {
        const surrogates = await this._parent.findElements(selenium_webdriver_1.By.css('OJ-SURROGATE'));
        return surrogates.length > 0;
    }
    async openMenu(menu) {
        await this.giveRootFocus();
        return this.getDriver().executeScript(`
      const menu = arguments[0];
      const launcher = arguments[1];
      const clickEvent = new CustomEvent('click');
      launcher.dispatchEvent(clickEvent);
      menu.open(clickEvent, { launcher: launcher })
    `, menu, this._root);
    }
    async giveRootFocus() {
        if (this._touchSupported) {
            const touchActions = new selenium_webdriver_1.TouchSequence(this.getDriver());
            await touchActions.tap(this._root).perform();
        }
        else {
            await this.getDriver().executeScript(`
      const root = arguments[0];
      root.click();
      `, this._root);
        }
    }
    async findMenuOptionByValue(targetValue, menu) {
        const options = await menu.findElements(selenium_webdriver_1.By.css('oj-option'));
        for (let i = 0; i < options.length; i++) {
            const value = await options[i].getAttribute('value');
            if (value === targetValue) {
                return options[i];
            }
        }
        throw Error('No option associated with value: ' + targetValue);
    }
    async doMenuAction(item, menu) {
        const a = await item.findElement(selenium_webdriver_1.By.css('a'));
        const aId = await a.getAttribute('id');
        let activeAId = await menu.getAttribute("aria-activedescendant");
        while (activeAId !== aId) {
            await menu.sendKeys(selenium_webdriver_1.Key.DOWN);
            activeAId = await menu.getAttribute("aria-activedescendant");
        }
        await item.click();
        await menu.whenBusyContextReady();
    }
}
