// @oracle/oraclejet-webdriver/elements/oj-n-box/OjNBoxBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjNBoxBase = void 0;
const __1 = require("../");
class OjNBoxBase extends __1.OjWebElement {
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAs() {
        return this.getProperty("as");
    }
    getCellContent() {
        return this.getProperty("cellContent");
    }
    getCellMaximize() {
        return this.getProperty("cellMaximize");
    }
    getCells() {
        return this.getProperty("cells");
    }
    getColumns() {
        return this.getProperty("columns");
    }
    getColumnsTitle() {
        return this.getProperty("columnsTitle");
    }
    getCountLabel() {
        return this.getProperty("countLabel");
    }
    getData() {
        return this.getProperty("data");
    }
    getGroupAttributes() {
        return this.getProperty("groupAttributes");
    }
    getGroupBehavior() {
        return this.getProperty("groupBehavior");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getLabelTruncation() {
        return this.getProperty("labelTruncation");
    }
    changeMaximizedColumn(maximizedColumn) {
        return this.setProperty("maximizedColumn", maximizedColumn);
    }
    getMaximizedColumn() {
        return this.getProperty("maximizedColumn");
    }
    changeMaximizedRow(maximizedRow) {
        return this.setProperty("maximizedRow", maximizedRow);
    }
    getMaximizedRow() {
        return this.getProperty("maximizedRow");
    }
    getNodes() {
        return this.getProperty("nodes");
    }
    getOtherColor() {
        return this.getProperty("otherColor");
    }
    getOtherThreshold() {
        return this.getProperty("otherThreshold");
    }
    getRows() {
        return this.getProperty("rows");
    }
    getRowsTitle() {
        return this.getProperty("rowsTitle");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getStyleDefaults() {
        return this.getProperty("styleDefaults");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTouchResponse() {
        return this.getProperty("touchResponse");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjNBoxBase = OjNBoxBase;
