// @oracle/oraclejet-webdriver/elements/oj-color-palette/OjColorPaletteBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjColorPalette, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjColorPaletteBase extends OjWebElement {
  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjColorPaletteBase.DisplayOptions> {
    return this.getProperty<OjColorPaletteBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjColorPaletteBase.Help> {
    return this.getProperty<OjColorPaletteBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjColorPaletteBase.HelpHints> {
    return this.getProperty<OjColorPaletteBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelDisplay property.
   * Specifies whether a text label accompanies the color swatch.
   * @return The value of labelDisplay property.
   * 
   */
  public getLabelDisplay(): Promise<string> {
    return this.getProperty<string>("labelDisplay");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * Used to establish a relationship between this element and another element.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string> {
    return this.getProperty<string>("labelledBy");
  }

  /**
   * Gets the value of layout property.
   * Specifies the layout of the color swatches.
   * @return The value of layout property.
   * 
   */
  public getLayout(): Promise<string> {
    return this.getProperty<string>("layout");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of palette property.
   * Specifies an array of objects defining the palette's color set.
   * @return The value of palette property.
   * 
   */
  public getPalette(): Promise<Array<OjColorPaletteBase.Palette>> {
    return this.getProperty<Array<OjColorPaletteBase.Palette>>("palette");
  }

  /**
   * Gets the value of swatchSize property.
   * Specifies the swatch size.
   * @return The value of swatchSize property.
   * 
   */
  public getSwatchSize(): Promise<string> {
    return this.getProperty<string>("swatchSize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjColorPaletteBase.Translations|null> {
    return this.getProperty<OjColorPaletteBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Sets the value of "value" property.
   * The current value of the palette element.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: object): Promise<void> {
    return this.setProperty<object>("value", value);
  }

  /**
   * Gets the value of value property.
   * The current value of the palette element.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<object> {
    return this.getProperty<object>("value");
  }

}

namespace OjColorPaletteBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface Palette {
    /**
     * the color definition
     */
    "color": object;
    /**
     * optional descriptive string (refer to attribute label-display). If omitted, label defaults to the color's hex string format.
     */
    "label": string;
  }
  export interface Translations {
    /**
     * Label for swatch with no color.
     */
    "labelNone": string;
  }
}
