// @oracle/oraclejet-webdriver/elements/oj-label/OjLabelBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjLabel, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjLabelBase extends OjWebElement {
  /**
   * Gets the value of for property.
   * Specifies the form element associated with this label. See the Help documentation for more information.
   * @return The value of for property.
   * 
   */
  public getFor(): Promise<string|null> {
    return this.getProperty<string|null>("for");
  }

  /**
   * Gets the value of help property.
   * Help information associated with this label.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjLabelBase.Help|null> {
    return this.getProperty<OjLabelBase.Help|null>("help");
  }

  /**
   * Gets the value of labelId property.
   * Specifies the id to set on the internal label element, if required. See the Help documentation for more information.
   * @return The value of labelId property.
   * 
   */
  public getLabelId(): Promise<string|null> {
    return this.getProperty<string|null>("labelId");
  }

  /**
   * Gets the value of showRequired property.
   * Specifies whether the label should render an icon indicating that the associated form field requires a value. See the Help documentation for more information.
   * @return The value of showRequired property.
   * 
   */
  public getShowRequired(): Promise<boolean> {
    return this.getProperty<boolean>("showRequired");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjLabelBase.Translations|null> {
    return this.getProperty<OjLabelBase.Translations|null>("translations");
  }

}

namespace OjLabelBase {
  export interface Help {
    /**
     * help definition text
     */
    "definition": string|null;
    /**
     * help source url
     */
    "source": string|null;
  }
  export interface Translations {
    /**
     * Used for the default help icon tooltip.. Most likely the application developer would use the oj-label's help.definition property to specify the help definition text per oj-label element.
     */
    "tooltipHelp": string;
    /**
     * Used for the default required icon tooltip.
     */
    "tooltipRequired": string;
  }
}
