// @oracle/oraclejet-webdriver/elements/oj-color-spectrum/OjColorSpectrumBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjColorSpectrumBase = void 0;
const __1 = require("../");
class OjColorSpectrumBase extends __1.OjWebElement {
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getDisplayOptions() {
        return this.getProperty("displayOptions");
    }
    getHelp() {
        return this.getProperty("help");
    }
    getHelpHints() {
        return this.getProperty("helpHints");
    }
    getLabelEdge() {
        return this.getProperty("labelEdge");
    }
    getLabelHint() {
        return this.getProperty("labelHint");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    changeMessagesCustom(messagesCustom) {
        return this.setProperty("messagesCustom", messagesCustom);
    }
    getMessagesCustom() {
        return this.getProperty("messagesCustom");
    }
    getTransientValue() {
        return this.getProperty("transientValue");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUserAssistanceDensity() {
        return this.getProperty("userAssistanceDensity");
    }
    getValid() {
        return this.getProperty("valid");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
}
exports.OjColorSpectrumBase = OjColorSpectrumBase;
