// @oracle/oraclejet-webdriver/elements/oj-color-spectrum/OjColorSpectrumBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjColorSpectrum, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjColorSpectrumBase extends OjWebElement {
  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjColorSpectrumBase.DisplayOptions> {
    return this.getProperty<OjColorSpectrumBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjColorSpectrumBase.Help> {
    return this.getProperty<OjColorSpectrumBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjColorSpectrumBase.HelpHints> {
    return this.getProperty<OjColorSpectrumBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * Used to establish a relationship between this element and another element.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string> {
    return this.getProperty<string>("labelledBy");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of transientValue property.
   * Retrieves the transient value of the component.
   * @return The value of transientValue property.
   * 
   */
  public getTransientValue(): Promise<object> {
    return this.getProperty<object>("transientValue");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjColorSpectrumBase.Translations|null> {
    return this.getProperty<OjColorSpectrumBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Sets the value of "value" property.
   * Specifies the value of the element representing the current color.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: object): Promise<void> {
    return this.setProperty<object>("value", value);
  }

  /**
   * Gets the value of value property.
   * Specifies the value of the element representing the current color.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<object> {
    return this.getProperty<object>("value");
  }

}

namespace OjColorSpectrumBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface Translations {
    /**
     * Label for Hue slider tooltip.
     */
    "labelHue": string;
    /**
     * Label for Opacity slider tooltip.
     */
    "labelOpacity": string;
    /**
     * Label for spectrum thumb tooltip.
     */
    "labelSatLum": string;
    /**
     * Label for spectrum thumb description.
     */
    "labelThumbDesc": string;
  }
}
