// @oracle/oraclejet-webdriver/elements/oj-conveyor-belt/OjConveyorBeltBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjConveyorBelt, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjConveyorBeltBase extends OjWebElement {
  /**
   * Sets the value of "arrowVisibility" property.
   * Specifies visibility of overflow arrow buttons.
   * @param arrowVisibility The value to set for "arrowVisibility"
   * 
   */
  public changeArrowVisibility(arrowVisibility: string): Promise<void> {
    return this.setProperty<string>("arrowVisibility", arrowVisibility);
  }

  /**
   * Gets the value of arrowVisibility property.
   * Specifies visibility of overflow arrow buttons.
   * @return The value of arrowVisibility property.
   * 
   */
  public getArrowVisibility(): Promise<string> {
    return this.getProperty<string>("arrowVisibility");
  }

  /**
   * Sets the value of "contentParent" property.
   * Specify the selector of the descendant DOM element in the conveyorBelt that directly contains the items to scroll among.
   * @param contentParent The value to set for "contentParent"
   * 
   */
  public changeContentParent(contentParent: string): Promise<void> {
    return this.setProperty<string>("contentParent", contentParent);
  }

  /**
   * Gets the value of contentParent property.
   * Specify the selector of the descendant DOM element in the conveyorBelt that directly contains the items to scroll among.
   * @return The value of contentParent property.
   * 
   */
  public getContentParent(): Promise<string> {
    return this.getProperty<string>("contentParent");
  }

  /**
   * Sets the value of "orientation" property.
   * Specifies the orientation of the conveyorBelt.
   * @param orientation The value to set for "orientation"
   * 
   */
  public changeOrientation(orientation: string): Promise<void> {
    return this.setProperty<string>("orientation", orientation);
  }

  /**
   * Gets the value of orientation property.
   * Specifies the orientation of the conveyorBelt.
   * @return The value of orientation property.
   * 
   */
  public getOrientation(): Promise<string> {
    return this.getProperty<string>("orientation");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjConveyorBeltBase.Translations|null> {
    return this.getProperty<OjConveyorBeltBase.Translations|null>("translations");
  }

}

namespace OjConveyorBeltBase {
  export interface Translations {
    /**
     * Tooltip for the next overflow content arrow.
     */
    "tipArrowNext": string;
    /**
     * Tooltip for the previous overflow content arrow.
     */
    "tipArrowPrevious": string;
  }
}
