// @oracle/oraclejet-webdriver/elements/oj-input-date/OjInputDateBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjInputDate, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjInputDateBase extends OjWebElement {
  /**
   * Gets the value of asyncValidators property.
   * Specifies a list of asynchronous validators used by the component when performing validation. Use async-validators when you need to perform some validation work on the server. See the Help documentation for more information.
   * @return The value of asyncValidators property.
   * @deprecated Since 8.0.0. Use the validators property instead for either regular Validators or AsyncValidators.
   */
  public getAsyncValidators(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("asyncValidators");
  }

  /**
   * Gets the value of autocomplete property.
   * Specifies a component's autocomplete state. See the Help documentation for more information.
   * @return The value of autocomplete property.
   * 
   */
  public getAutocomplete(): Promise<string> {
    return this.getProperty<string>("autocomplete");
  }

  /**
   * Gets the value of autofocus property.
   * Specifies whether the component will get input focus when the page is loaded. See the Help documentation for more information.
   * @return The value of autofocus property.
   * 
   */
  public getAutofocus(): Promise<boolean> {
    return this.getProperty<boolean>("autofocus");
  }

  /**
   * Gets the value of converter property.
   * An object that converts the value. See the Help documentation for more information.
   * @return The value of converter property.
   * 
   */
  public getConverter(): Promise<object> {
    return this.getProperty<object>("converter");
  }

  /**
   * Gets the value of datePicker property.
   * An object whose properties describe the appearance and behavior of the date picker. See the Help documentation for more information.
   * @return The value of datePicker property.
   * 
   */
  public getDatePicker(): Promise<OjInputDateBase.DatePicker> {
    return this.getProperty<OjInputDateBase.DatePicker>("datePicker");
  }

  /**
   * Gets the value of dayFormatter property.
   * Specifies a callback function used when rendering the day. See the Help documentation for more information.
   * @return The value of dayFormatter property.
   * 
   */
  public getDayFormatter(): Promise<null> {
    return this.getProperty<null>("dayFormatter");
  }

  /**
   * Gets the value of dayMetaData property.
   * Specifies additional information to be used when rendering the day. See the Help documentation for more information.
   * @return The value of dayMetaData property.
   * 
   */
  public getDayMetaData(): Promise<object> {
    return this.getProperty<object>("dayMetaData");
  }

  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjInputDateBase.DisplayOptions> {
    return this.getProperty<OjInputDateBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjInputDateBase.Help> {
    return this.getProperty<OjInputDateBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjInputDateBase.HelpHints> {
    return this.getProperty<OjInputDateBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of keyboardEdit property.
   * Determines if keyboard entry of the text is allowed. When disabled, the picker must be used to select the date. See the Help documentation for more information.
   * @return The value of keyboardEdit property.
   * 
   */
  public getKeyboardEdit(): Promise<string> {
    return this.getProperty<string>("keyboardEdit");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * The oj-label sets the labelledBy property programmatically on the form component. See the Help documentation for more information.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of max property.
   * The maximum selectable date, in ISO string format. When set to null, there is no maximum. min and max must be in the same ISO string format as value (local, zulu, or offset).
   * @return The value of max property.
   * 
   */
  public getMax(): Promise<string|null> {
    return this.getProperty<string|null>("max");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of min property.
   * The minimum selectable date, in ISO string format. When set to null, there is no minimum. min and max must be in the same ISO string format as value (local, zulu, or offset).
   * @return The value of min property.
   * 
   */
  public getMin(): Promise<string|null> {
    return this.getProperty<string|null>("min");
  }

  /**
   * Gets the value of name property.
   * Specifies the name of the component.
   * @return The value of name property.
   * @deprecated Since 6.0.0. JET does not use form submit, so this is not needed.
   */
  public getName(): Promise<string> {
    return this.getProperty<string>("name");
  }

  /**
   * Gets the value of pickerAttributes property.
   * Specifies attributes to be set on the picker DOM element when it is launched. See the Help documentation for more information.
   * @return The value of pickerAttributes property.
   * 
   */
  public getPickerAttributes(): Promise<OjInputDateBase.PickerAttributes> {
    return this.getProperty<OjInputDateBase.PickerAttributes>("pickerAttributes");
  }

  /**
   * Gets the value of placeholder property.
   * The placeholder text to set on the element.
   * @return The value of placeholder property.
   * 
   */
  public getPlaceholder(): Promise<string> {
    return this.getProperty<string>("placeholder");
  }

  /**
   * Gets the value of rawValue property.
   * Read-only property used for retrieving the current value from the input field in string form. See the Help documentation for more information.
   * @return The value of rawValue property.
   * 
   */
  public getRawValue(): Promise<string> {
    return this.getProperty<string>("rawValue");
  }

  /**
   * Gets the value of readonly property.
   * Specifies whether the component is read-only. A read-only element cannot be modified, but user interaction is allowed. See the Help documentation for more information.
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of renderMode property.
   * Specifies whether to render the date picker in JET, or as a native picker control. See the Help documentation for more information.
   * @return The value of renderMode property.
   * @deprecated Since 8.0.0. The "native" mode rendering is deprecated because JET is promoting a consistent Oracle UX over native look and feel in Redwood. Since this property takes only two values the property itself is deprecated. The theme variable "$inputDateTimeRenderModeOptionDefault" is also deprecated for the same reason.
   */
  public getRenderMode(): Promise<string> {
    return this.getProperty<string>("renderMode");
  }

  /**
   * Gets the value of required property.
   * Specifies whether the component is required or optional. See the Help documentation for more information.
   * @return The value of required property.
   * 
   */
  public getRequired(): Promise<boolean> {
    return this.getProperty<boolean>("required");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjInputDateBase.Translations|null> {
    return this.getProperty<OjInputDateBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Gets the value of validators property.
   * A list of validators used by the element, along with any implicit component validators, when performing validation. See the Help documentation for more information.
   * @return The value of validators property.
   * 
   */
  public getValidators(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("validators");
  }

  /**
   * Sets the value of "value" property.
   * The value of the input date element, which must be an ISOString. See the Help documentation for more information.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: string): Promise<void> {
    return this.setProperty<string>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the input date element, which must be an ISOString. See the Help documentation for more information.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<string> {
    return this.getProperty<string>("value");
  }

}

namespace OjInputDateBase {
  export interface DatePicker {
    /**
     * Specifies whether the month should be rendered as a button to allow selection, instead of as text.
     */
    "changeMonth": string;
    /**
     * Specifies whether the year should be rendered as a button to allow selection, instead of as text.
     */
    "changeYear": string;
    /**
     * Specifies the position in multiple months at which to show the current month (starting at 0).
     */
    "currentMonthPos": number;
    /**
     * Specifies the behavior of days outside the current viewing month.
     */
    "daysOutsideMonth": string;
    /**
     * Specifies what content is shown within the footer of the calendar.
     */
    "footerLayout": string;
    /**
     * Specifies the number of months to show at once. See the Help documentation for more information.
     */
    "numberOfMonths": number;
    /**
     * Specifies when the date picker should be shown.
     */
    "showOn": string;
    /**
     * Specifies the number of months to step backwards and forwards for the Ctrl+Alt+Page Up and Ctrl+Alt+Page Down keystrokes.
     */
    "stepBigMonths": number;
    /**
     * Specifies how the prev and next keys will step backwards and forwards through the months. See the Help documentation for more information.
     */
    "stepMonths": string|number;
    /**
     * Specifies whether the week of the year will be shown.
     */
    "weekDisplay": string;
    /**
     * Specifies the range of years displayed in the year drop-down. See the Help documentation for more information.
     */
    "yearRange": string;
  }
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface PickerAttributes {
    /**
     * 
     */
    "class": string;
    /**
     * 
     */
    "style": string;
  }
  export interface Translations {
    /**
     * Message to announce to AT via aria-live when the length.max is exceeded.
     */
    "accessibleMaxLengthExceeded": string;
    /**
     * Message to annnouce to AT via aria-live for the remaining character count.
     */
    "accessibleMaxLengthRemaining": string;
    /**
     * The text to display for the current day link.
     */
    "currentText": string;
    /**
     * Provides properties to customize the hint and message text used by the implicit date restriction validator associated to the InputDateTime and InputDate components.
     */
    "dateRestriction": OjInputDateBase.TranslationsDateRestriction;
    /**
     * Provides properties to customize the hint and message text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components.
     */
    "dateTimeRange": OjInputDateBase.TranslationsDateTimeRange;
    /**
     * The text to display for the next month link.
     */
    "nextText": string;
    /**
     * The text to display for the previous month link.
     */
    "prevText": string;
    /**
     * Provides properties to customize the message text used by the implicit regexp validator associated to the InputText and TextArea components.
     */
    "regexp": OjInputDateBase.TranslationsRegexp;
    /**
     * Provides properties to customize the summary, detail and hint text used by the implicit required validator associated to any editable component that supports the required option.
     */
    "required": OjInputDateBase.TranslationsRequired;
    /**
     * Tooltip text for the calendar icon.
     */
    "tooltipCalendar": string;
    /**
     * Tooltip text for the calendar icon when the component is disabled.
     */
    "tooltipCalendarDisabled": string;
    /**
     * Tooltip text for the calendar + time icon.
     */
    "tooltipCalendarTime": string;
    /**
     * Tooltip text for the calendar + time icon when the component is disabled.
     */
    "tooltipCalendarTimeDisabled": string;
    /**
     * The text to display for the week of the year column heading.
     */
    "weekHeader": string;
  }
  export interface TranslationsDateRestriction {
    /**
     * Hint text used by the implicit date restriction validator associated to the InputDateTime and InputDate components.
     */
    "hint": string;
    /**
     * Message detail for the implicit date restriction validator associated to the InputDateTime and InputDate components.
     */
    "messageDetail": string;
    /**
     * Message summary for the implicit date restriction validator associated to the InputDateTime and InputDate components.
     */
    "messageSummary": string;
  }
  export interface TranslationsDateTimeRange {
    /**
     * Provides properties to customize the hint text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components.
     */
    "hint": OjInputDateBase.TranslationsDateTimeRangeHint;
    /**
     * Provides properties to customize the error message text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components.
     */
    "messageDetail": OjInputDateBase.TranslationsDateTimeRangeMessageDetail;
    /**
     * Provides properties to customize the error message summary text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components.
     */
    "messageSummary": OjInputDateBase.TranslationsDateTimeRangeMessageSummary;
  }
  export interface TranslationsDateTimeRangeHint {
    /**
     * Hint text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components. hint.inRange is shown when both min and max are set, and is used to tell the user the allowed number range.
     */
    "inRange": string;
    /**
     * Hint text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components. hint.max is shown when max is set and min is not set, and is used to tell the user the allowed maximum.
     */
    "max": string;
    /**
     * Hint text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components. hint.min is shown when min is set and max is not set, and is used to tell the user the allowed minimum.
     */
    "min": string;
  }
  export interface TranslationsDateTimeRangeMessageDetail {
    /**
     * Error message text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components. messageDetail.rangeOverflow is shown when max is set, and the value is greater than the maximum.
     */
    "rangeOverflow": string;
    /**
     * Error message text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components. messageDetail.rangeUnderflow is shown when min is set, and the value is less than the minimum.
     */
    "rangeUnderflow": string;
  }
  export interface TranslationsDateTimeRangeMessageSummary {
    /**
     * Error message summary text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components. messageSummary.rangeOverflow is shown when max is set, and the value is greater than the maximum.
     */
    "rangeOverflow": string;
    /**
     * Error message summary text used by the implicit datetime range validator associated to the InputDateTime, InputDate, and InputTime components. messageSummary.rangeUnderflow is shown when min is set, and the value is less than the minimum.
     */
    "rangeUnderflow": string;
  }
  export interface TranslationsRegexp {
    /**
     * Provides properties to customize the error message detail used by the implicit regexp validator associated to the InputText and TextArea components.
     */
    "messageDetail": string;
    /**
     * Provides properties to customize the error message summary used by the implicit regexp validator associated to the InputText and TextArea components.
     */
    "messageSummary": string;
  }
  export interface TranslationsRequired {
    /**
     * Hint text used by required validation error.
     */
    "hint": string;
    /**
     * Message text that describes the details of the required validation error.
     */
    "messageDetail": string;
    /**
     * Message text for summarizing a required validation error.
     */
    "messageSummary": string;
  }
}
