// @oracle/oraclejet-webdriver/elements/oj-combobox-many/OjComboboxManyBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjComboboxMany, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjComboboxManyBase extends OjWebElement {
  /**
   * Gets the value of asyncValidators property.
   * Specifies a list of asynchronous validators used by the component when performing validation. Use async-validators when you need to perform some validation work on the server. See the Help documentation for more information.
   * @return The value of asyncValidators property.
   * @deprecated Since 8.0.0. Use the validators property instead for either regular Validators or AsyncValidators.
   */
  public getAsyncValidators(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("asyncValidators");
  }

  /**
   * Gets the value of converter property.
   * An object that converts the value. See the Help documentation for more information.
   * @return The value of converter property.
   * 
   */
  public getConverter(): Promise<object|null> {
    return this.getProperty<object|null>("converter");
  }

  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjComboboxManyBase.DisplayOptions> {
    return this.getProperty<OjComboboxManyBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjComboboxManyBase.Help> {
    return this.getProperty<OjComboboxManyBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjComboboxManyBase.HelpHints> {
    return this.getProperty<OjComboboxManyBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * The oj-label sets the labelledBy property programmatically on the form component.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of maximumResultCount property.
   * The maximum number of results displayed in the dropdown.
   * @return The value of maximumResultCount property.
   * 
   */
  public getMaximumResultCount(): Promise<number> {
    return this.getProperty<number>("maximumResultCount");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of minLength property.
   * The minimum number of characters a user must type before search filtering is performed.
   * @return The value of minLength property.
   * 
   */
  public getMinLength(): Promise<number> {
    return this.getProperty<number>("minLength");
  }

  /**
   * Gets the value of optionRenderer property.
   * The renderer function that renders the content of each option.
   * @return The value of optionRenderer property.
   * 
   */
  public getOptionRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("optionRenderer");
  }

  /**
   * Gets the value of options property.
   * The option items for the Combobox.
   * @return The value of options property.
   * 
   */
  public getOptions(): Promise<Array<object>|object|null> {
    return this.getProperty<Array<object>|object|null>("options");
  }

  /**
   * Gets the value of optionsKeys property.
   * Specify the key names to use in the options array. Depending on options-keys means that the signature of the data does not match what is supported by the options attribute.
   * @return The value of optionsKeys property.
   * 
   */
  public getOptionsKeys(): Promise<OjComboboxManyBase.OptionsKeys> {
    return this.getProperty<OjComboboxManyBase.OptionsKeys>("optionsKeys");
  }

  /**
   * Gets the value of pickerAttributes property.
   * The style attributes for the drop down.
   * @return The value of pickerAttributes property.
   * 
   */
  public getPickerAttributes(): Promise<OjComboboxManyBase.PickerAttributes> {
    return this.getProperty<OjComboboxManyBase.PickerAttributes>("pickerAttributes");
  }

  /**
   * Gets the value of placeholder property.
   * A short hint that can be displayed before user selects or enters a value.
   * @return The value of placeholder property.
   * 
   */
  public getPlaceholder(): Promise<string|null> {
    return this.getProperty<string|null>("placeholder");
  }

  /**
   * Gets the value of rawValue property.
   * The currently displayed text retrieved from the input field.
   * @return The value of rawValue property.
   * 
   */
  public getRawValue(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("rawValue");
  }

  /**
   * Gets the value of readonly property.
   * Specifies whether the component is read-only. A read-only element cannot be modified, but user interaction is allowed. See the Help documentation for more information.
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of required property.
   * Specifies whether a value is required.
   * @return The value of required property.
   * 
   */
  public getRequired(): Promise<boolean> {
    return this.getProperty<boolean>("required");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjComboboxManyBase.Translations|null> {
    return this.getProperty<OjComboboxManyBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Gets the value of validators property.
   * Specifies a list of synchronous validators for performing validation by the element. See the Help documentation for more information.
   * @return The value of validators property.
   * 
   */
  public getValidators(): Promise<Array<unknown>> {
    return this.getProperty<Array<unknown>>("validators");
  }

  /**
   * Sets the value of "value" property.
   * The value of the element.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: Array<any>|null): Promise<void> {
    return this.setProperty<Array<any>|null>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the element.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<Array<any>|null> {
    return this.getProperty<Array<any>|null>("value");
  }

  /**
   * Sets the value of "valueOptions" property.
   * The current values of the element and their associated display labels.
   * @param valueOptions The value to set for "valueOptions"
   * 
   */
  public changeValueOptions(valueOptions: null|Array<OjComboboxManyBase.ValueOptions>): Promise<void> {
    return this.setProperty<null|Array<OjComboboxManyBase.ValueOptions>>("valueOptions", valueOptions);
  }

  /**
   * Gets the value of valueOptions property.
   * The current values of the element and their associated display labels.
   * @return The value of valueOptions property.
   * 
   */
  public getValueOptions(): Promise<null|Array<OjComboboxManyBase.ValueOptions>> {
    return this.getProperty<null|Array<OjComboboxManyBase.ValueOptions>>("valueOptions");
  }

}

namespace OjComboboxManyBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface OptionsKeys {
    /**
     * The object for the child keys. It is ignored when using a TreeDataProvider.
     */
    "childKeys": OjComboboxManyBase.OptionsKeysChildKeys;
    /**
     * The key name for the children. It is ignored when using a TreeDataProvider.
     */
    "children": string;
    /**
     * The key name for the label.
     */
    "label": string;
    /**
     * The key name for the value.
     */
    "value": string;
  }
  export interface OptionsKeysChildKeys {
    /**
     * The object for the child keys.
     */
    "childKeys": OjComboboxManyBase.OptionsKeysChildKeysChildKeys;
    /**
     * The key name for the children.
     */
    "children": string;
    /**
     * The key name for the label.
     */
    "label": string;
    /**
     * The key name for the value.
     */
    "value": string;
  }
  export interface OptionsKeysChildKeysChildKeys {
    /**
     * The object for the child keys.
     */
    "childKeys": object;
    /**
     * The key name for the children.
     */
    "children": string;
    /**
     * The key name for the label.
     */
    "label": string;
    /**
     * The key name for the value.
     */
    "value": string;
  }
  export interface PickerAttributes {
    /**
     * The css class to append to the picker.
     */
    "class": string;
    /**
     * The css style to append to the picker.
     */
    "style": string;
  }
  export interface Translations {
    /**
     * Text at the top of drop down when not all results are fetched and users need to filter further.
     */
    "filterFurther": string;
    /**
     * Text for the drop down when more than one options are found.
     */
    "moreMatchesFound": string;
    /**
     * No options found text for drop down.
     */
    "noMatchesFound": string;
    /**
     * Text for the drop down when all options are selected
     */
    "noMoreResults": string;
    /**
     * Text for the drop down when one option is found.
     */
    "oneMatchesFound": string;
    /**
     * Provides properties to customize the summary, detail and hint text used by the implicit required validator associated to any editable component that supports the required option.
     */
    "required": OjComboboxManyBase.TranslationsRequired;
  }
  export interface TranslationsRequired {
    /**
     * Hint text used by required validation error.
     */
    "hint": string;
    /**
     * Message text that describes the details of the required validation error.
     */
    "messageDetail": string;
    /**
     * Message text for summarizing a required validation error.
     */
    "messageSummary": string;
  }
  export interface ValueOptions {
    /**
     * display label of value above. If missing, String(value) is used.
     */
    "label": string;
    /**
     * the current value of JET Combobox
     */
    "value": any;
  }
}
