// @oracle/oraclejet-webdriver/elements/oj-rating-gauge/OjRatingGaugeBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjRatingGaugeBase = void 0;
const __1 = require("../");
class OjRatingGaugeBase extends __1.OjWebElement {
    changeChanged(changed) {
        return this.setProperty("changed", changed);
    }
    getChanged() {
        return this.getProperty("changed");
    }
    getChangedState() {
        return this.getProperty("changedState");
    }
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getHoverState() {
        return this.getProperty("hoverState");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    getMax() {
        return this.getProperty("max");
    }
    getMin() {
        return this.getProperty("min");
    }
    getOrientation() {
        return this.getProperty("orientation");
    }
    getPreserveAspectRatio() {
        return this.getProperty("preserveAspectRatio");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getSelectedState() {
        return this.getProperty("selectedState");
    }
    getSizeProperty() {
        return this.getProperty("size");
    }
    getStep() {
        return this.getProperty("step");
    }
    getThresholds() {
        return this.getProperty("thresholds");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTransientValue() {
        return this.getProperty("transientValue");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUnselectedState() {
        return this.getProperty("unselectedState");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
    getVisualEffects() {
        return this.getProperty("visualEffects");
    }
}
exports.OjRatingGaugeBase = OjRatingGaugeBase;
