// @oracle/oraclejet-webdriver/elements/oj-rating-gauge/OjRatingGaugeBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjRatingGauge, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjRatingGaugeBase extends OjWebElement {
  /**
   * Sets the value of "changed" property.
   * Whether there has been a value entered by the user.
   * @param changed The value to set for "changed"
   * 
   */
  public changeChanged(changed: boolean): Promise<void> {
    return this.setProperty<boolean>("changed", changed);
  }

  /**
   * Gets the value of changed property.
   * Whether there has been a value entered by the user.
   * @return The value of changed property.
   * 
   */
  public getChanged(): Promise<boolean> {
    return this.getProperty<boolean>("changed");
  }

  /**
   * Gets the value of changedState property.
   * The changed shape for the gauge. Displayed after the user has set a value, or when the changed attribute of the data object is set to true.
   * @return The value of changedState property.
   * 
   */
  public getChangedState(): Promise<OjRatingGaugeBase.ChangedState> {
    return this.getProperty<OjRatingGaugeBase.ChangedState>("changedState");
  }

  /**
   * Gets the value of describedBy property.
   * Specifies a relationship between this component and another element.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string|null> {
    return this.getProperty<string|null>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Defines whether the gauge is disabled or not. User interaction is prevented if set to true. Visual indication will not be present when custom image source is used. Other state specific svg-class-name and svg-style will be overridden by disabled default style.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of hoverState property.
   * The shape that displays on hover.
   * @return The value of hoverState property.
   * 
   */
  public getHoverState(): Promise<OjRatingGaugeBase.HoverState> {
    return this.getProperty<OjRatingGaugeBase.HoverState>("hoverState");
  }

  /**
   * Gets the value of labelledBy property.
   * Establishes a relationship between this component and another element, typically an oj-label custom element. See the Help documentation for more information.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of max property.
   * Integer value specifying the maximum value of the gauge, which determines the number of shapes or images that are displayed.
   * @return The value of max property.
   * 
   */
  public getMax(): Promise<number> {
    return this.getProperty<number>("max");
  }

  /**
   * Gets the value of min property.
   * The minimum value that can be set on the gauge by the end user. Does not affect the value set on the gauge by API.
   * @return The value of min property.
   * 
   */
  public getMin(): Promise<number> {
    return this.getProperty<number>("min");
  }

  /**
   * Gets the value of orientation property.
   * Defines the type of rating gauge to be rendered.
   * @return The value of orientation property.
   * @deprecated Since 9.0.0. Usage of vertical rating gauges is not recommended.
   */
  public getOrientation(): Promise<string> {
    return this.getProperty<string>("orientation");
  }

  /**
   * Gets the value of preserveAspectRatio property.
   * Specifies whether the images provided should show up at their defined aspect ratios. See the Help documentation for more information.
   * @return The value of preserveAspectRatio property.
   * 
   */
  public getPreserveAspectRatio(): Promise<string> {
    return this.getProperty<string>("preserveAspectRatio");
  }

  /**
   * Gets the value of readonly property.
   * Defines whether the value of the gauge can be changed by the end user.
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of selectedState property.
   * The selected shape for the gauge.
   * @return The value of selectedState property.
   * 
   */
  public getSelectedState(): Promise<OjRatingGaugeBase.SelectedState> {
    return this.getProperty<OjRatingGaugeBase.SelectedState>("selectedState");
  }

  /**
   * Gets the value of size property.
   * Specifies the size of the rating gauge item.
   * @return The value of size property.
   * 
   */
  public getSizeProperty(): Promise<string> {
    return this.getProperty<string>("size");
  }

  /**
   * Gets the value of step property.
   * Specifies the increment by which values can be specified by the end user.
   * @return The value of step property.
   * 
   */
  public getStep(): Promise<number> {
    return this.getProperty<number>("step");
  }

  /**
   * Gets the value of thresholds property.
   * An array of objects specifying the gauge thresholds.
   * @return The value of thresholds property.
   * 
   */
  public getThresholds(): Promise<Array<OjRatingGaugeBase.Thresholds>> {
    return this.getProperty<Array<OjRatingGaugeBase.Thresholds>>("thresholds");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjRatingGaugeBase.Tooltip> {
    return this.getProperty<OjRatingGaugeBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of transientValue property.
   * Read-only property used for retrieving the transient value from the component. See the Help documentation for more information.
   * @return The value of transientValue property.
   * 
   */
  public getTransientValue(): Promise<number|null> {
    return this.getProperty<number|null>("transientValue");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjRatingGaugeBase.Translations|null> {
    return this.getProperty<OjRatingGaugeBase.Translations|null>("translations");
  }

  /**
   * Gets the value of unselectedState property.
   * The unselected shape for the gauge.
   * @return The value of unselectedState property.
   * 
   */
  public getUnselectedState(): Promise<OjRatingGaugeBase.UnselectedState> {
    return this.getProperty<OjRatingGaugeBase.UnselectedState>("unselectedState");
  }

  /**
   * Sets the value of "value" property.
   * The value set on the gauge.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: number|null): Promise<void> {
    return this.setProperty<number|null>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value set on the gauge.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<number|null> {
    return this.getProperty<number|null>("value");
  }

  /**
   * Gets the value of visualEffects property.
   * Defines whether visual effects such as overlays are applied to the gauge.
   * @return The value of visualEffects property.
   * 
   */
  public getVisualEffects(): Promise<string> {
    return this.getProperty<string>("visualEffects");
  }

}

namespace OjRatingGaugeBase {
  export interface ChangedState {
    /**
     * The border color for changed state. Does not apply if a custom image is specified. The default value comes from the CSS and varies based on theme.
     */
    "borderColor": string;
    /**
     * The color for changed state. Does not apply if a custom image is specified. See the Help documentation for more information.
     */
    "color": string;
    /**
     * The shape to be used. Can take the name of a built-in shape or the SVG path commands for a custom shape. Does not apply if a custom image is specified.
     */
    "shape": string;
    /**
     * The URI of the custom image. If specified, it takes precedence over shape. See the Help documentation for more information.
     */
    "source": string;
    /**
     * The CSS style class to apply to the changed state. See the Help documentation for more information.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the changed state. See the Help documentation for more information.
     */
    "svgStyle": object;
  }
  export interface HoverState {
    /**
     * The border color for hover state. Does not apply if a custom image is specified. The default value comes from the CSS and varies based on theme.
     */
    "borderColor": string;
    /**
     * The color for hover state. Does not apply if a custom image is specified. See the Help documentation for more information.
     */
    "color": string;
    /**
     * The shape to be used. Can take the name of a built-in shape or the SVG path commands for a custom shape. Does not apply if a custom image is specified.
     */
    "shape": string;
    /**
     * The URI of the custom image. If specified, it takes precedence over shape. See the Help documentation for more information.
     */
    "source": string;
    /**
     * The CSS style class to apply to the hover state. See the Help documentation for more information.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the hover state. See the Help documentation for more information.
     */
    "svgStyle": object;
  }
  export interface SelectedState {
    /**
     * The border color for selected state. Does not apply if a custom image is specified. The default value comes from the CSS and varies based on theme.
     */
    "borderColor": string;
    /**
     * The color for selected state. Does not apply if a custom image is specified. See the Help documentation for more information.
     */
    "color": string;
    /**
     * The shape to be used. Can take the name of a built-in shape or the SVG path commands for a custom shape. Does not apply if a custom image is specified.
     */
    "shape": string;
    /**
     * The URI of the custom image. If specified, it takes precedence over shape. See the Help documentation for more information.
     */
    "source": string;
    /**
     * The CSS style class to apply to the selected state. See the Help documentation for more information.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the selected state. See the Help documentation for more information.
     */
    "svgStyle": object;
  }
  export interface Thresholds {
    /**
     * The border color of the threshold.
     */
    "borderColor": string;
    /**
     * The color of the threshold.
     */
    "color": string;
    /**
     * The upper bound of the threshold. This value is ignored for the final threshold, which uses the maximum value of the gauge.
     */
    "max": number;
    /**
     * Specific description for the threshold and overwrites the shortDesc specified on gauge. This is used for accessibility and also for customizing the tooltip text.
     */
    "shortDesc": string;
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
  }
  export interface UnselectedState {
    /**
     * The border color for unselected state. Does not apply if a custom image is specified. The default value comes from the CSS and varies based on theme.
     */
    "borderColor": string;
    /**
     * The color for unselected state. Does not apply if a custom image is specified. See the Help documentation for more information.
     */
    "color": string;
    /**
     * The shape to be used. Can take the name of a built-in shape or the SVG path commands for a custom shape. Does not apply if a custom image is specified.
     */
    "shape": string;
    /**
     * The URI of the custom image. If specified, it takes precedence over shape. See the Help documentation for more information.
     */
    "source": string;
    /**
     * The CSS style class to apply to the unselected state. See the Help documentation for more information.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the unselected state. See the Help documentation for more information.
     */
    "svgStyle": object;
  }
}
