// @oracle/oraclejet-webdriver/elements/oj-tab-bar/OjTabBarBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTabBarBase = void 0;
const __1 = require("../");
class OjTabBarBase extends __1.OjWebElement {
    getAs() {
        return this.getProperty("as");
    }
    changeCurrentItem(currentItem) {
        return this.setProperty("currentItem", currentItem);
    }
    getCurrentItem() {
        return this.getProperty("currentItem");
    }
    getData() {
        return this.getProperty("data");
    }
    getDisplay() {
        return this.getProperty("display");
    }
    getEdge() {
        return this.getProperty("edge");
    }
    getItem() {
        return this.getProperty("item");
    }
    getLayout() {
        return this.getProperty("layout");
    }
    getOverflow() {
        return this.getProperty("overflow");
    }
    getReorderable() {
        return this.getProperty("reorderable");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getTruncation() {
        return this.getProperty("truncation");
    }
}
exports.OjTabBarBase = OjTabBarBase;
