// @oracle/oraclejet-webdriver/elements/oj-tab-bar/OjTabBarBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjTabBar, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjTabBarBase extends OjWebElement {
  /**
   * Gets the value of as property.
   * Specifies the alias for the current item when referenced inside the item template.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Sets the value of "currentItem" property.
   * Specifies the key of the item that should have keyboard focus. See the Help documentation for more information.
   * @param currentItem The value to set for "currentItem"
   * 
   */
  public changeCurrentItem(currentItem: any): Promise<void> {
    return this.setProperty<any>("currentItem", currentItem);
  }

  /**
   * Gets the value of currentItem property.
   * Specifies the key of the item that should have keyboard focus. See the Help documentation for more information.
   * @return The value of currentItem property.
   * 
   */
  public getCurrentItem(): Promise<any> {
    return this.getProperty<any>("currentItem");
  }

  /**
   * Gets the value of data property.
   * Specifies the data provider for the Tab Bar. See the Help documentation for more information.
   * @return The value of data property.
   * 
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of display property.
   * Specifies what needs to be displayed. See the Help documentation for more information.
   * @return The value of display property.
   * 
   */
  public getDisplay(): Promise<string> {
    return this.getProperty<string>("display");
  }

  /**
   * Gets the value of edge property.
   * Specifies the edge position of the Tab Bar.
   * @return The value of edge property.
   * 
   */
  public getEdge(): Promise<string> {
    return this.getProperty<string>("edge");
  }

  /**
   * Gets the value of item property.
   * Customize the functionality of each tab on Tab Bar.
   * @return The value of item property.
   * 
   */
  public getItem(): Promise<OjTabBarBase.Item> {
    return this.getProperty<OjTabBarBase.Item>("item");
  }

  /**
   * Gets the value of layout property.
   * Specifies whether to stretch the tab bar items to occupy available space or to condense items.
   * @return The value of layout property.
   * 
   */
  public getLayout(): Promise<string> {
    return this.getProperty<string>("layout");
  }

  /**
   * Gets the value of overflow property.
   * Specifies overflow behaviour for the Tab Bar.
   * @return The value of overflow property.
   * 
   */
  public getOverflow(): Promise<string> {
    return this.getProperty<string>("overflow");
  }

  /**
   * Gets the value of reorderable property.
   * Specifies whether tabs can be reordered.
   * @return The value of reorderable property.
   * 
   */
  public getReorderable(): Promise<string> {
    return this.getProperty<string>("reorderable");
  }

  /**
   * Sets the value of "selection" property.
   * Specifies the key of the selected item. See the Help documentation for more information.
   * @param selection The value to set for "selection"
   * 
   */
  public changeSelection(selection: any): Promise<void> {
    return this.setProperty<any>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * Specifies the key of the selected item. See the Help documentation for more information.
   * @return The value of selection property.
   * 
   */
  public getSelection(): Promise<any> {
    return this.getProperty<any>("selection");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjTabBarBase.Translations|null> {
    return this.getProperty<OjTabBarBase.Translations|null>("translations");
  }

  /**
   * Gets the value of truncation property.
   * Specifies whether truncation needs to be applied.
   * @return The value of truncation property.
   * 
   */
  public getTruncation(): Promise<string> {
    return this.getProperty<string>("truncation");
  }

}

namespace OjTabBarBase {
  export interface Item {
    /**
     * Specifies the renderer for the tab item. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies whether the tab can be selected. See the Help documentation for more information.
     */
    "selectable": null|boolean;
  }
  export interface Translations {
    /**
     * Provides properties to customize the screen reader text when the tentative drop target is after a certain item.
     */
    "accessibleReorderAfterItem": string;
    /**
     * Provides properties to customize the screen reader text when the tentative drop target is before a certain item.
     */
    "accessibleReorderBeforeItem": string;
    /**
     * Provides properties to customize the screen reader touch instructional text for reordering items.
     */
    "accessibleReorderTouchInstructionText": string;
    /**
     * Provides properties to customize the context menu cut label.
     */
    "labelCut": string;
    /**
     * Provides properties to customize the context menu paste after label.
     */
    "labelPasteAfter": string;
    /**
     * Provides properties to customize the context menu paste before label.
     */
    "labelPasteBefore": string;
    /**
     * Provides properties to customize the context menu remove label.
     */
    "labelRemove": string;
    /**
     * Provides properties to customize the message text used by Tab Bar when waiting for data.
     */
    "msgFetchingData": string;
    /**
     * Provides properties to customize the message text used by Tab Bar when there are no items.
     */
    "msgNoData": string;
    /**
     * Label for overflow menu button.
     */
    "overflowItemLabel": string;
    /**
     * Text cue for the removable tab, used as the aria-label for the close icon.
     */
    "removeCueText": string;
    /**
     * Provides text to read to screen reader when an item is selected.
     */
    "selectedLabel": string;
  }
}
