// @oracle/oraclejet-webdriver/elements/oj-message/OjMessageBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjMessage, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjMessageBase extends OjWebElement {
  /**
   * Gets the value of displayOptions property.
   * Specifies the display options for contents of the message.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjMessageBase.DisplayOptions> {
    return this.getProperty<OjMessageBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of message property.
   * Structured data specifying essential information to display a message on the UI.
   * @return The value of message property.
   * 
   */
  public getMessage(): Promise<OjMessageBase.Message> {
    return this.getProperty<OjMessageBase.Message>("message");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjMessageBase.Translations> {
    return this.getProperty<OjMessageBase.Translations>("translations");
  }

}

namespace OjMessageBase {
  export interface DisplayOptions {
    /**
     * Specifies the display option for message category text in this message.
     */
    "category": string;
  }
  export interface Message {
    /**
     * Specifies the duration in milliseconds that the message will be shown before it closes automatically. See the Help documentation for more information.
     */
    "autoTimeout": number;
    /**
     * Specifies message category text which is shown next to the message icon. See the Help documentation for more information.
     */
    "category": string;
    /**
     * Specifies the UI affordance provided to end users to be able to close the message. See the Help documentation for more information.
     */
    "closeAffordance": string;
    /**
     * Specifies detail text for the message.
     */
    "detail": string;
    /**
     * Specifies the URL for the custom image to be used as an icon representing the message. See the Help documentation for more information.
     */
    "icon": string;
    /**
     * Specifies the severity of message.
     */
    "severity": string;
    /**
     * Specifies the sound to be played when a message is opened. This is needed for accessibility. See the Help documentation for more information.
     */
    "sound": string;
    /**
     * Specifies summary text for the message.
     */
    "summary": string;
    /**
     * Specifies a timestamp for the message to be displayed in the message header. See the Help documentation for more information.
     */
    "timestamp": string;
  }
  export interface Translations {
    /**
     * Category text to be displayed for the messages of different severity levels.
     */
    "categories": OjMessageBase.TranslationsCategories;
    /**
     * Label for the message close button.
     */
    "labelCloseIcon": string;
  }
  export interface TranslationsCategories {
    /**
     * Category text for the 'confirmation' severity level.
     */
    "confirmation": string;
    /**
     * Category text for the 'error' severity level.
     */
    "error": string;
    /**
     * Category text for the 'info' severity level.
     */
    "info": string;
    /**
     * Category text for the 'warning' severity level.
     */
    "warning": string;
  }
}
