// @oracle/oraclejet-webdriver/elements/oj-checkboxset/OjCheckboxsetBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjCheckboxsetBase = void 0;
const __1 = require("../");
class OjCheckboxsetBase extends __1.OjWebElement {
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getDisplayOptions() {
        return this.getProperty("displayOptions");
    }
    getHelp() {
        return this.getProperty("help");
    }
    getHelpHints() {
        return this.getProperty("helpHints");
    }
    getLabelEdge() {
        return this.getProperty("labelEdge");
    }
    getLabelHint() {
        return this.getProperty("labelHint");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    changeMessagesCustom(messagesCustom) {
        return this.setProperty("messagesCustom", messagesCustom);
    }
    getMessagesCustom() {
        return this.getProperty("messagesCustom");
    }
    getOptionRenderer() {
        return this.getProperty("optionRenderer");
    }
    getOptions() {
        return this.getProperty("options");
    }
    getOptionsKeys() {
        return this.getProperty("optionsKeys");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getRequired() {
        return this.getProperty("required");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUserAssistanceDensity() {
        return this.getProperty("userAssistanceDensity");
    }
    getValid() {
        return this.getProperty("valid");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
}
exports.OjCheckboxsetBase = OjCheckboxsetBase;
