// @oracle/oraclejet-webdriver/elements/oj-treemap/OjTreemapBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTreemapBase = void 0;
const __1 = require("../");
class OjTreemapBase extends __1.OjWebElement {
    getAnimationDuration() {
        return this.getProperty("animationDuration");
    }
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAnimationUpdateColor() {
        return this.getProperty("animationUpdateColor");
    }
    getAs() {
        return this.getProperty("as");
    }
    getColorLabel() {
        return this.getProperty("colorLabel");
    }
    getData() {
        return this.getProperty("data");
    }
    getDisplayLevels() {
        return this.getProperty("displayLevels");
    }
    getDrilling() {
        return this.getProperty("drilling");
    }
    getGroupGaps() {
        return this.getProperty("groupGaps");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    getHighlightMode() {
        return this.getProperty("highlightMode");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getHoverBehaviorDelay() {
        return this.getProperty("hoverBehaviorDelay");
    }
    changeIsolatedNode(isolatedNode) {
        return this.setProperty("isolatedNode", isolatedNode);
    }
    getIsolatedNode() {
        return this.getProperty("isolatedNode");
    }
    getLayout() {
        return this.getProperty("layout");
    }
    getNodeContent() {
        return this.getProperty("nodeContent");
    }
    getNodeDefaults() {
        return this.getProperty("nodeDefaults");
    }
    getNodeSeparators() {
        return this.getProperty("nodeSeparators");
    }
    getNodes() {
        return this.getProperty("nodes");
    }
    getRootNode() {
        return this.getProperty("rootNode");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSizeLabel() {
        return this.getProperty("sizeLabel");
    }
    getSorting() {
        return this.getProperty("sorting");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTouchResponse() {
        return this.getProperty("touchResponse");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjTreemapBase = OjTreemapBase;
