// @oracle/oraclejet-webdriver/elements/oj-train/OjTrainBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjTrain, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjTrainBase extends OjWebElement {
  /**
   * Sets the value of "selectedStep" property.
   * Specifies the id of the current selected step.
   * @param selectedStep The value to set for "selectedStep"
   * 
   */
  public changeSelectedStep(selectedStep: string): Promise<void> {
    return this.setProperty<string>("selectedStep", selectedStep);
  }

  /**
   * Gets the value of selectedStep property.
   * Specifies the id of the current selected step.
   * @return The value of selectedStep property.
   * 
   */
  public getSelectedStep(): Promise<string> {
    return this.getProperty<string>("selectedStep");
  }

  /**
   * Gets the value of steps property.
   * The array of step objects.
   * @return The value of steps property.
   * 
   */
  public getSteps(): Promise<Array<OjTrainBase.Steps>> {
    return this.getProperty<Array<OjTrainBase.Steps>>("steps");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<object|null> {
    return this.getProperty<object|null>("translations");
  }

}

namespace OjTrainBase {
  export interface Steps {
    /**
     * indicates whether the step is disabled
     */
    "disabled": boolean;
    /**
     * id of the step
     */
    "id": string;
    /**
     * label of the step
     */
    "label": string;
    /**
     * the type of message icon displayed on the step, null removes the message icon
     */
    "messageType": string|null;
    /**
     * indicates whether the step has been visited
     */
    "visited": boolean;
  }
}
