// @oracle/oraclejet-webdriver/elements/oj-tag-cloud/OjTagCloudBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjTagCloud, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjTagCloudBase extends OjWebElement {
  /**
   * Gets the value of animationOnDataChange property.
   * Specifies the animation that is applied on data changes.
   * @return The value of animationOnDataChange property.
   * 
   */
  public getAnimationOnDataChange(): Promise<string> {
    return this.getProperty<string>("animationOnDataChange");
  }

  /**
   * Gets the value of animationOnDisplay property.
   * Specifies the animation that is shown on initial display.
   * @return The value of animationOnDisplay property.
   * 
   */
  public getAnimationOnDisplay(): Promise<string> {
    return this.getProperty<string>("animationOnDisplay");
  }

  /**
   * Gets the value of as property.
   * An alias for the '$current' context variable passed to slot content for the itemTemplate slot.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Gets the value of data property.
   * Specifies the DataProvider for the tag cloud. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Sets the value of "hiddenCategories" property.
   * An array of category strings used for category filtering. Data items with a category in hiddenCategories will be filtered.
   * @param hiddenCategories The value to set for "hiddenCategories"
   * 
   */
  public changeHiddenCategories(hiddenCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("hiddenCategories", hiddenCategories);
  }

  /**
   * Gets the value of hiddenCategories property.
   * An array of category strings used for category filtering. Data items with a category in hiddenCategories will be filtered.
   * @return The value of hiddenCategories property.
   * 
   */
  public getHiddenCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("hiddenCategories");
  }

  /**
   * Gets the value of highlightMatch property.
   * The matching condition for the highlightedCategories property. See the Help documentation for more information.
   * @return The value of highlightMatch property.
   * 
   */
  public getHighlightMatch(): Promise<string> {
    return this.getProperty<string>("highlightMatch");
  }

  /**
   * Sets the value of "highlightedCategories" property.
   * An array of category strings used for highlighting. Data items matching categories in this array will be highlighted.
   * @param highlightedCategories The value to set for "highlightedCategories"
   * 
   */
  public changeHighlightedCategories(highlightedCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("highlightedCategories", highlightedCategories);
  }

  /**
   * Gets the value of highlightedCategories property.
   * An array of category strings used for highlighting. Data items matching categories in this array will be highlighted.
   * @return The value of highlightedCategories property.
   * 
   */
  public getHighlightedCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("highlightedCategories");
  }

  /**
   * Gets the value of hoverBehavior property.
   * Defines the behavior applied when hovering over data items.
   * @return The value of hoverBehavior property.
   * 
   */
  public getHoverBehavior(): Promise<string> {
    return this.getProperty<string>("hoverBehavior");
  }

  /**
   * Gets the value of items property.
   * An array of objects defining the data items for the tag cloud items. Also accepts a Promise for deferred data rendering.
   * @return The value of items property.
   * 
   */
  public getItems(): Promise<Array<OjTagCloudBase.Items>|null|null> {
    return this.getProperty<Array<OjTagCloudBase.Items>|null|null>("items");
  }

  /**
   * Gets the value of layout property.
   * The layout to use for tag display.
   * @return The value of layout property.
   * 
   */
  public getLayout(): Promise<string> {
    return this.getProperty<string>("layout");
  }

  /**
   * Sets the value of "selection" property.
   * An array containing the ids of the initially selected data items.
   * @param selection The value to set for "selection"
   * 
   */
  public changeSelection(selection: Array<any>): Promise<void> {
    return this.setProperty<Array<any>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * An array containing the ids of the initially selected data items.
   * @return The value of selection property.
   * 
   */
  public getSelection(): Promise<Array<any>> {
    return this.getProperty<Array<any>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the selection mode.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

  /**
   * Gets the value of styleDefaults property.
   * An object defining the style defaults for this tag cloud.
   * @return The value of styleDefaults property.
   * 
   */
  public getStyleDefaults(): Promise<OjTagCloudBase.StyleDefaults> {
    return this.getProperty<OjTagCloudBase.StyleDefaults>("styleDefaults");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjTagCloudBase.Tooltip> {
    return this.getProperty<OjTagCloudBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of touchResponse property.
   * Specifies configuration options for touch and hold delays on mobile devices. See the Help documentation for more information.
   * @return The value of touchResponse property.
   * 
   */
  public getTouchResponse(): Promise<string> {
    return this.getProperty<string>("touchResponse");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjTagCloudBase.Translations|null> {
    return this.getProperty<OjTagCloudBase.Translations|null>("translations");
  }

}

namespace OjTagCloudBase {
  export interface Items {
    /**
     * An array of category strings corresponding to the tag cloud items. This allows highlighting and filtering of items.
     */
    "categories": Array<string>;
    /**
     * The color of the text. Will be overridden by any color defined in the style option. The default value comes from the CSS and varies based on theme.
     */
    "color": string;
    /**
     * The item id should be set by the application if the DataProvider is not being used. The row key will be used as id in the DataProvider case.
     */
    "id": any;
    /**
     * The text of the item.
     */
    "label": string;
    /**
     * The description of the item. This is used for customizing the tooltip text.
     */
    "shortDesc": string|null;
    /**
     * The CSS style class defining the style of the item text.
     */
    "svgClassName": string;
    /**
     * The CSS style object defining the style of the item text. Only SVG CSS style properties are supported. The default value comes from the CSS and varies based on theme.
     */
    "svgStyle": object;
    /**
     * The url this item references.
     */
    "url": string;
    /**
     * The value of this item which will be used to scale its font-size within the tag cloud.
     */
    "value": number;
  }
  export interface StyleDefaults {
    /**
     * The duration of the animations in milliseconds.
     */
    "animationDuration": number;
    /**
     * Specifies initial hover delay in milliseconds for highlighting data items.
     */
    "hoverBehaviorDelay": number;
    /**
     * The CSS style object defining the style of the items.
     */
    "svgStyle": object;
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
  }
}
