// @oracle/oraclejet-webdriver/elements/oj-tag-cloud/OjTagCloudBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTagCloudBase = void 0;
const __1 = require("../");
class OjTagCloudBase extends __1.OjWebElement {
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAs() {
        return this.getProperty("as");
    }
    getData() {
        return this.getProperty("data");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getItems() {
        return this.getProperty("items");
    }
    getLayout() {
        return this.getProperty("layout");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getStyleDefaults() {
        return this.getProperty("styleDefaults");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTouchResponse() {
        return this.getProperty("touchResponse");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjTagCloudBase = OjTagCloudBase;
