// @oracle/oraclejet-webdriver/elements/oj-select-one/OjSelectOneBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjSelectOne, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjSelectOneBase extends OjWebElement {
  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjSelectOneBase.DisplayOptions> {
    return this.getProperty<OjSelectOneBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjSelectOneBase.Help> {
    return this.getProperty<OjSelectOneBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjSelectOneBase.HelpHints> {
    return this.getProperty<OjSelectOneBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * The oj-label sets the labelledBy property programmatically on the form component.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of maximumResultCount property.
   * The maximum number of results displayed in the dropdown.
   * @return The value of maximumResultCount property.
   * 
   */
  public getMaximumResultCount(): Promise<number> {
    return this.getProperty<number>("maximumResultCount");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of minimumResultsForSearch property.
   * The threshold for showing the search box in the dropdown.
   * @return The value of minimumResultsForSearch property.
   * 
   */
  public getMinimumResultsForSearch(): Promise<number> {
    return this.getProperty<number>("minimumResultsForSearch");
  }

  /**
   * Gets the value of optionRenderer property.
   * The renderer function that renders the content of each option.
   * @return The value of optionRenderer property.
   * 
   */
  public getOptionRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("optionRenderer");
  }

  /**
   * Gets the value of options property.
   * The option items for the Select.
   * @return The value of options property.
   * 
   */
  public getOptions(): Promise<Array<object>|object|null> {
    return this.getProperty<Array<object>|object|null>("options");
  }

  /**
   * Gets the value of optionsKeys property.
   * Specify the key names to use in the options array. Depending on options-keys means that the signature of the data does not match what is supported by the options attribute.
   * @return The value of optionsKeys property.
   * 
   */
  public getOptionsKeys(): Promise<OjSelectOneBase.OptionsKeys> {
    return this.getProperty<OjSelectOneBase.OptionsKeys>("optionsKeys");
  }

  /**
   * Gets the value of pickerAttributes property.
   * The style attributes for the drop down.
   * @return The value of pickerAttributes property.
   * 
   */
  public getPickerAttributes(): Promise<OjSelectOneBase.PickerAttributes> {
    return this.getProperty<OjSelectOneBase.PickerAttributes>("pickerAttributes");
  }

  /**
   * Gets the value of placeholder property.
   * A short hint that can be displayed before user selects a value.
   * @return The value of placeholder property.
   * 
   */
  public getPlaceholder(): Promise<string|null> {
    return this.getProperty<string|null>("placeholder");
  }

  /**
   * Gets the value of readonly property.
   * Specifies whether a value is readonly
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of renderMode property.
   * Specifies whether to render select in JET or as a HTML Select tag.
   * @return The value of renderMode property.
   * @deprecated Since 8.0.0. The "native" mode rendering is deprecated because JET is promoting a consistent Oracle UX over native look and feel in Redwood. Since this property takes only two values the property itself is deprecated. The theme variable "$selectRenderModeOptionDefault" is also deprecated for the same reason.
   */
  public getRenderMode(): Promise<string> {
    return this.getProperty<string>("renderMode");
  }

  /**
   * Gets the value of required property.
   * Specifies whether a value is required.
   * @return The value of required property.
   * 
   */
  public getRequired(): Promise<boolean> {
    return this.getProperty<boolean>("required");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjSelectOneBase.Translations|null> {
    return this.getProperty<OjSelectOneBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Sets the value of "value" property.
   * The value of the element.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: any): Promise<void> {
    return this.setProperty<any>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the element.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<any> {
    return this.getProperty<any>("value");
  }

  /**
   * Sets the value of "valueOption" property.
   * The current value of the element and its associated display label.
   * @param valueOption The value to set for "valueOption"
   * 
   */
  public changeValueOption(valueOption: null|OjSelectOneBase.ValueOption): Promise<void> {
    return this.setProperty<null|OjSelectOneBase.ValueOption>("valueOption", valueOption);
  }

  /**
   * Gets the value of valueOption property.
   * The current value of the element and its associated display label.
   * @return The value of valueOption property.
   * 
   */
  public getValueOption(): Promise<null|OjSelectOneBase.ValueOption> {
    return this.getProperty<null|OjSelectOneBase.ValueOption>("valueOption");
  }

}

namespace OjSelectOneBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface OptionsKeys {
    /**
     * The object for the child keys. It is ignored when using a TreeDataProvider.
     */
    "childKeys": OjSelectOneBase.OptionsKeysChildKeys;
    /**
     * The key name for the children. It is ignored when using a TreeDataProvider.
     */
    "children": string;
    /**
     * The key name for the label.
     */
    "label": string;
    /**
     * The key name for the value.
     */
    "value": string;
  }
  export interface OptionsKeysChildKeys {
    /**
     * The object for the child keys.
     */
    "childKeys": object;
    /**
     * The key name for the children.
     */
    "children": string;
    /**
     * The key name for the label.
     */
    "label": string;
    /**
     * The key name for the value.
     */
    "value": string;
  }
  export interface PickerAttributes {
    /**
     * The css class to append to the picker.
     */
    "class": string;
    /**
     * The css style to append to the picker.
     */
    "style": string;
  }
  export interface Translations {
    /**
     * Text at the top of drop down when not all results are fetched and users need to filter further.
     */
    "filterFurther": string;
    /**
     * Text for the drop down when more than one options are found.
     */
    "moreMatchesFound": string;
    /**
     * Text for the drop down when no options are found.
     */
    "noMatchesFound": string;
    /**
     * Text for the drop down when one option is found.
     */
    "oneMatchesFound": string;
    /**
     * Provides properties to customize the summary, detail and hint text used by the implicit required validator associated to any editable component that supports the required option.
     */
    "required": OjSelectOneBase.TranslationsRequired;
    /**
     * Tooltip text for the search box
     */
    "searchField": string;
  }
  export interface TranslationsRequired {
    /**
     * Hint text used by required validation error.
     */
    "hint": string;
    /**
     * Message text that describes the details of the required validation error.
     */
    "messageDetail": string;
    /**
     * Message text for summarizing a required validation error.
     */
    "messageSummary": string;
  }
  export interface ValueOption {
    /**
     * display label of value above. If missing, String(value) is used.
     */
    "label": string;
    /**
     * current value of JET Select
     */
    "value": any;
  }
}
