// @oracle/oraclejet-webdriver/elements/oj-select-one/OjSelectOneBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjSelectOneBase = void 0;
const __1 = require("../");
class OjSelectOneBase extends __1.OjWebElement {
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getDisplayOptions() {
        return this.getProperty("displayOptions");
    }
    getHelp() {
        return this.getProperty("help");
    }
    getHelpHints() {
        return this.getProperty("helpHints");
    }
    getLabelEdge() {
        return this.getProperty("labelEdge");
    }
    getLabelHint() {
        return this.getProperty("labelHint");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    getMaximumResultCount() {
        return this.getProperty("maximumResultCount");
    }
    changeMessagesCustom(messagesCustom) {
        return this.setProperty("messagesCustom", messagesCustom);
    }
    getMessagesCustom() {
        return this.getProperty("messagesCustom");
    }
    getMinimumResultsForSearch() {
        return this.getProperty("minimumResultsForSearch");
    }
    getOptionRenderer() {
        return this.getProperty("optionRenderer");
    }
    getOptions() {
        return this.getProperty("options");
    }
    getOptionsKeys() {
        return this.getProperty("optionsKeys");
    }
    getPickerAttributes() {
        return this.getProperty("pickerAttributes");
    }
    getPlaceholder() {
        return this.getProperty("placeholder");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getRenderMode() {
        return this.getProperty("renderMode");
    }
    getRequired() {
        return this.getProperty("required");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUserAssistanceDensity() {
        return this.getProperty("userAssistanceDensity");
    }
    getValid() {
        return this.getProperty("valid");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
    changeValueOption(valueOption) {
        return this.setProperty("valueOption", valueOption);
    }
    getValueOption() {
        return this.getProperty("valueOption");
    }
}
exports.OjSelectOneBase = OjSelectOneBase;
