// @oracle/oraclejet-webdriver/elements/oj-timeline/OjTimelineBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjTimeline, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjTimelineBase extends OjWebElement {
  /**
   * Gets the value of animationOnDataChange property.
   * Specifies the animation that is applied on data changes.
   * @return The value of animationOnDataChange property.
   * 
   */
  public getAnimationOnDataChange(): Promise<string> {
    return this.getProperty<string>("animationOnDataChange");
  }

  /**
   * Gets the value of animationOnDisplay property.
   * Specifies the animation that is shown on initial display.
   * @return The value of animationOnDisplay property.
   * 
   */
  public getAnimationOnDisplay(): Promise<string> {
    return this.getProperty<string>("animationOnDisplay");
  }

  /**
   * Gets the value of data property.
   * Specifies the data provider for the timeline. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object> {
    return this.getProperty<object>("data");
  }

  /**
   * Gets the value of end property.
   * The end time of the timeline. See the Help documentation for more information.
   * @return The value of end property.
   * 
   */
  public getEnd(): Promise<string> {
    return this.getProperty<string>("end");
  }

  /**
   * Gets the value of majorAxis property.
   * An object defining the optional timeline major axis.
   * @return The value of majorAxis property.
   * 
   */
  public getMajorAxis(): Promise<OjTimelineBase.MajorAxis> {
    return this.getProperty<OjTimelineBase.MajorAxis>("majorAxis");
  }

  /**
   * Gets the value of minorAxis property.
   * An object defining the timeline minor axis.
   * @return The value of minorAxis property.
   * 
   */
  public getMinorAxis(): Promise<OjTimelineBase.MinorAxis> {
    return this.getProperty<OjTimelineBase.MinorAxis>("minorAxis");
  }

  /**
   * Gets the value of orientation property.
   * The orientation of the element.
   * @return The value of orientation property.
   * 
   */
  public getOrientation(): Promise<string> {
    return this.getProperty<string>("orientation");
  }

  /**
   * Gets the value of overview property.
   * An object defining the optional timeline overview.
   * @return The value of overview property.
   * 
   */
  public getOverview(): Promise<OjTimelineBase.Overview> {
    return this.getProperty<OjTimelineBase.Overview>("overview");
  }

  /**
   * Gets the value of referenceObjects property.
   * The array of reference objects associated with the timeline. Currently only the first reference object in the array is supported. See the Help documentation for more information.
   * @return The value of referenceObjects property.
   * 
   */
  public getReferenceObjects(): Promise<Array<OjTimelineBase.ReferenceObjects>> {
    return this.getProperty<Array<OjTimelineBase.ReferenceObjects>>("referenceObjects");
  }

  /**
   * Sets the value of "selection" property.
   * An array of strings containing the ids of the initially selected items.
   * @param selection The value to set for "selection"
   * 
   */
  public changeSelection(selection: Array<any>): Promise<void> {
    return this.setProperty<Array<any>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * An array of strings containing the ids of the initially selected items.
   * @return The value of selection property.
   * 
   */
  public getSelection(): Promise<Array<any>> {
    return this.getProperty<Array<any>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the selection mode.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

  /**
   * Gets the value of series property.
   * An array of objects defining each timeline series. Also accepts a Promise for deferred data rendering.
   * @return The value of series property.
   * 
   */
  public getSeries(): Promise<Array<OjTimelineBase.Series>|null> {
    return this.getProperty<Array<OjTimelineBase.Series>|null>("series");
  }

  /**
   * Gets the value of start property.
   * The start time of the timeline. See the Help documentation for more information.
   * @return The value of start property.
   * 
   */
  public getStart(): Promise<string> {
    return this.getProperty<string>("start");
  }

  /**
   * Gets the value of styleDefaults property.
   * An object defining the default styling for this timeline.
   * @return The value of styleDefaults property.
   * 
   */
  public getStyleDefaults(): Promise<OjTimelineBase.StyleDefaults> {
    return this.getProperty<OjTimelineBase.StyleDefaults>("styleDefaults");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjTimelineBase.Tooltip> {
    return this.getProperty<OjTimelineBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjTimelineBase.Translations|null> {
    return this.getProperty<OjTimelineBase.Translations|null>("translations");
  }

  /**
   * Gets the value of valueFormats property.
   * An object specifying value formatting and tooltip behavior, whose keys generally correspond to item properties.
   * @return The value of valueFormats property.
   * 
   */
  public getValueFormats(): Promise<OjTimelineBase.ValueFormats> {
    return this.getProperty<OjTimelineBase.ValueFormats>("valueFormats");
  }

  /**
   * Gets the value of viewportEnd property.
   * The end time of the timeline's viewport. See the Help documentation for more information.
   * @return The value of viewportEnd property.
   * 
   */
  public getViewportEnd(): Promise<string> {
    return this.getProperty<string>("viewportEnd");
  }

  /**
   * Gets the value of viewportStart property.
   * The start time of the timeline's viewport. See the Help documentation for more information.
   * @return The value of viewportStart property.
   * 
   */
  public getViewportStart(): Promise<string> {
    return this.getProperty<string>("viewportStart");
  }

}

namespace OjTimelineBase {
  export interface MajorAxis {
    /**
     * An object that converts the labels of the major axis for all 'scale' values'. See the Help documentation for more information.
     */
    "converter": OjTimelineBase.MajorAxisConverter;
    /**
     * Specifies the time scale used for the major axis.
     */
    "scale": string;
    /**
     * The CSS style object defining any additional styling of the major axis.
     */
    "svgStyle": object;
  }
  export interface MajorAxisConverter {
    /**
     * A converter (an instance that duck types Converter) used to format the 'days' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "days": object;
    /**
     * The default converter (an object literal or instance that duck types Converter) to use for all 'scale' values that do not otherwise have a converter object provided. See also DateTimeConverter.
     */
    "default": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'hours' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "hours": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'minutes' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "minutes": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'months' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "months": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'quarters' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "quarters": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'seconds' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "seconds": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'weeks' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "weeks": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'years' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "years": object;
  }
  export interface MinorAxis {
    /**
     * An object that converts the labels of the minor axis for all 'scale' values. See the Help documentation for more information.
     */
    "converter": OjTimelineBase.MinorAxisConverter;
    /**
     * Specifies the time scale used for the minor axis.
     */
    "scale": string;
    /**
     * The CSS style object defining any additional styling of the minor axis.
     */
    "svgStyle": object;
  }
  export interface MinorAxisConverter {
    /**
     * A converter (an instance that duck types Converter) used to format the 'days' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "days": object;
    /**
     * The default converter (an object literal or instance that duck types Converter) to use for all 'scale' values that do not otherwise have a converter object provided. See also DateTimeConverter.
     */
    "default": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'hours' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "hours": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'minutes' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "minutes": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'months' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "months": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'quarters' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "quarters": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'seconds' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "seconds": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'weeks' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "weeks": object;
    /**
     * A converter (an instance that duck types Converter) used to format the 'years' scale. If not specified, the default converter will be used for this scale. See also DateTimeConverter.
     */
    "years": object;
  }
  export interface Overview {
    /**
     * Specifies whether the overview scrollbar is rendered.
     */
    "rendered": string;
    /**
     * The CSS style object defining any additional styling of the overview.
     */
    "svgStyle": object;
  }
  export interface ReferenceObjects {
    /**
     * The time value of this reference object. If not specified, no reference object will be shown. See Date and Time Formats for more details on the required string formats.
     */
    "value": string;
  }
  export interface Series {
    /**
     * The text of an empty timeline series.
     */
    "emptyText": string;
    /**
     * The identifier for the timeline series.
     */
    "id": string;
    /**
     * The direction in which items are laid out when in a horizontal orientation. This attribute is ignored when in a vertical orientation.
     */
    "itemLayout": string;
    /**
     * An array of items. If not specified, no data will be shown in this series.
     */
    "items": Array<object>;
    /**
     * The label displayed on the timeline series. In not specified, no label will be shown.
     */
    "label": string;
    /**
     * The CSS style object defining any additional styling of the series. If not specified, no additional styling will be applied. Only SVG CSS style properties are supported.
     */
    "svgStyle": object;
  }
  export interface StyleDefaults {
    /**
     * The duration of the animations in milliseconds.
     */
    "animationDuration": number;
    /**
     * The border color of the timeline.
     */
    "borderColor": string;
    /**
     * An object defining the default styling for this timeline's items.
     */
    "item": OjTimelineBase.StyleDefaultsItem;
    /**
     * An object defining the default styling for this timeline's major axis.
     */
    "majorAxis": OjTimelineBase.StyleDefaultsMajorAxis;
    /**
     * An object defining the default styling for this timeline's minor axis.
     */
    "minorAxis": OjTimelineBase.StyleDefaultsMinorAxis;
    /**
     * An object defining the default styling for this timeline's overview.
     */
    "overview": OjTimelineBase.StyleDefaultsOverview;
    /**
     * An object defining the default styling for this timeline's reference objects.
     */
    "referenceObject": OjTimelineBase.StyleDefaultsReferenceObject;
    /**
     * An object defining the default styling for this timeline's series.
     */
    "series": OjTimelineBase.StyleDefaultsSeries;
  }
  export interface StyleDefaultsItem {
    /**
     * The background color of the timeline items.
     */
    "backgroundColor": string;
    /**
     * The border color of the timeline items.
     */
    "borderColor": string;
    /**
     * The CSS style object defining the style of the timeline item description text.
     */
    "descriptionStyle": object;
    /**
     * The background color of the highlighted timeline items.
     */
    "hoverBackgroundColor": string;
    /**
     * The border color of the highlighted timeline items.
     */
    "hoverBorderColor": string;
    /**
     * The background color of the selected timeline items.
     */
    "selectedBackgroundColor": string;
    /**
     * The border color of the selected timeline items.
     */
    "selectedBorderColor": string;
    /**
     * The CSS style object defining the style of the timeline item title text.
     */
    "titleStyle": object;
  }
  export interface StyleDefaultsMajorAxis {
    /**
     * The CSS style object defining the style of the major axis label text.
     */
    "labelStyle": object;
    /**
     * The color of the major axis separators.
     */
    "separatorColor": string;
  }
  export interface StyleDefaultsMinorAxis {
    /**
     * The background color of the minor axis.
     */
    "backgroundColor": string;
    /**
     * The border color of the minor axis.
     */
    "borderColor": string;
    /**
     * The CSS style object defining the style of the minor axis label text.
     */
    "labelStyle": object;
    /**
     * The color of the minor axis separators.
     */
    "separatorColor": string;
  }
  export interface StyleDefaultsOverview {
    /**
     * The background color of the timeline overview.
     */
    "backgroundColor": string;
    /**
     * The CSS style object defining the style of the timeline overview label text.
     */
    "labelStyle": object;
    /**
     * An object defining the default styling for the timeline overview window.
     */
    "window": OjTimelineBase.StyleDefaultsOverviewWindow;
  }
  export interface StyleDefaultsOverviewWindow {
    /**
     * The background color of the timeline overview window.
     */
    "backgroundColor": string;
    /**
     * The border color of the timeline overview window.
     */
    "borderColor": string;
  }
  export interface StyleDefaultsReferenceObject {
    /**
     * The color of the reference objects.
     */
    "color": string;
  }
  export interface StyleDefaultsSeries {
    /**
     * The background color of the series.
     */
    "backgroundColor": string;
    /**
     * The array defining the default color ramp for the series items.
     */
    "colors": Array<string>;
    /**
     * The CSS style object defining the style of the series empty text.
     */
    "emptyTextStyle": object;
    /**
     * The CSS style object defining the style of the series label text.
     */
    "labelStyle": object;
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Provides properties to customize the screen reader text describing an item's description text.
     */
    "accessibleItemDesc": string;
    /**
     * Provides properties to customize the screen reader text describing an item's end time.
     */
    "accessibleItemEnd": string;
    /**
     * Provides properties to customize the screen reader text describing an item's start time.
     */
    "accessibleItemStart": string;
    /**
     * Provides properties to customize the screen reader text describing an item's title text.
     */
    "accessibleItemTitle": string;
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Used for the default tooltip label for instance item information.
     */
    "labelDate": string;
    /**
     * Used for the default tooltip label for item description information.
     */
    "labelDescription": string;
    /**
     * Used for the default tooltip label for end time information.
     */
    "labelEnd": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used for accessibility.
     */
    "labelSeries": string;
    /**
     * Used for the default tooltip label for start time information.
     */
    "labelStart": string;
    /**
     * Used for the default tooltip label for item title information.
     */
    "labelTitle": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
    /**
     * Used for the zoom in tooltip.
     */
    "tooltipZoomIn": string;
    /**
     * Used for the zoom out tooltip.
     */
    "tooltipZoomOut": string;
  }
  export interface ValueFormats {
    /**
     * Specifies tooltip behavior for the date value of an instance item.
     */
    "date": OjTimelineBase.ValueFormatsDate;
    /**
     * Specifies tooltip behavior for the description value.
     */
    "description": OjTimelineBase.ValueFormatsDescription;
    /**
     * Specifies tooltip behavior for the end value.
     */
    "end": OjTimelineBase.ValueFormatsEnd;
    /**
     * Specifies tooltip behavior for the series value.
     */
    "series": OjTimelineBase.ValueFormatsSeries;
    /**
     * Specifies tooltip behavior for the start value.
     */
    "start": OjTimelineBase.ValueFormatsStart;
    /**
     * Specifies tooltip behavior for the title value.
     */
    "title": OjTimelineBase.ValueFormatsTitle;
  }
  export interface ValueFormatsDate {
    /**
     * The converter used to format the label. See the Help documentation for more information.
     */
    "converter": object;
    /**
     * Specifies whether the date value is displayed in the tooltip.
     */
    "tooltipDisplay": string;
    /**
     * A string representing the label that is displayed before the date value in the tooltip. See the Help documentation for more information.
     */
    "tooltipLabel": string;
  }
  export interface ValueFormatsDescription {
    /**
     * Specifies whether the description value is displayed in the tooltip.
     */
    "tooltipDisplay": string;
    /**
     * A string representing the label that is displayed before the description value in the tooltip. See the Help documentation for more information.
     */
    "tooltipLabel": string;
  }
  export interface ValueFormatsEnd {
    /**
     * The converter used to format the label. See the Help documentation for more information.
     */
    "converter": object;
    /**
     * Specifies whether the end value is displayed in the tooltip.
     */
    "tooltipDisplay": string;
    /**
     * A string representing the label that is displayed before the end value in the tooltip. See the Help documentation for more information.
     */
    "tooltipLabel": string;
  }
  export interface ValueFormatsSeries {
    /**
     * Specifies whether the series value is displayed in the tooltip.
     */
    "tooltipDisplay": string;
    /**
     * A string representing the label that is displayed before the series value in the tooltip. See the Help documentation for more information.
     */
    "tooltipLabel": string;
  }
  export interface ValueFormatsStart {
    /**
     * The converter used to format the label. See the Help documentation for more information.
     */
    "converter": object;
    /**
     * Specifies whether the start value is displayed in the tooltip.
     */
    "tooltipDisplay": string;
    /**
     * A string representing the label that is displayed before the start value in the tooltip. See the Help documentation for more information.
     */
    "tooltipLabel": string;
  }
  export interface ValueFormatsTitle {
    /**
     * Specifies whether the title value is displayed in the tooltip.
     */
    "tooltipDisplay": string;
    /**
     * A string representing the label that is displayed before the title value in the tooltip. See the Help documentation for more information.
     */
    "tooltipLabel": string;
  }
}
