// @oracle/oraclejet-webdriver/elements/oj-timeline/OjTimelineBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTimelineBase = void 0;
const __1 = require("../");
class OjTimelineBase extends __1.OjWebElement {
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getData() {
        return this.getProperty("data");
    }
    getEnd() {
        return this.getProperty("end");
    }
    getMajorAxis() {
        return this.getProperty("majorAxis");
    }
    getMinorAxis() {
        return this.getProperty("minorAxis");
    }
    getOrientation() {
        return this.getProperty("orientation");
    }
    getOverview() {
        return this.getProperty("overview");
    }
    getReferenceObjects() {
        return this.getProperty("referenceObjects");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSeries() {
        return this.getProperty("series");
    }
    getStart() {
        return this.getProperty("start");
    }
    getStyleDefaults() {
        return this.getProperty("styleDefaults");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getValueFormats() {
        return this.getProperty("valueFormats");
    }
    getViewportEnd() {
        return this.getProperty("viewportEnd");
    }
    getViewportStart() {
        return this.getProperty("viewportStart");
    }
}
exports.OjTimelineBase = OjTimelineBase;
