// @oracle/oraclejet-webdriver/elements/oj-led-gauge/OjLedGaugeBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjLedGaugeBase = void 0;
const __1 = require("../");
class OjLedGaugeBase extends __1.OjWebElement {
    getBorderColor() {
        return this.getProperty("borderColor");
    }
    getColor() {
        return this.getProperty("color");
    }
    getLabel() {
        return this.getProperty("label");
    }
    getMax() {
        return this.getProperty("max");
    }
    getMetricLabel() {
        return this.getProperty("metricLabel");
    }
    getMin() {
        return this.getProperty("min");
    }
    getRotation() {
        return this.getProperty("rotation");
    }
    getSizeProperty() {
        return this.getProperty("size");
    }
    getSvgClassName() {
        return this.getProperty("svgClassName");
    }
    getSvgStyle() {
        return this.getProperty("svgStyle");
    }
    getThresholds() {
        return this.getProperty("thresholds");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getType() {
        return this.getProperty("type");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
    getVisualEffects() {
        return this.getProperty("visualEffects");
    }
}
exports.OjLedGaugeBase = OjLedGaugeBase;
