// @oracle/oraclejet-webdriver/elements/oj-led-gauge/OjLedGaugeBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjLedGauge, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjLedGaugeBase extends OjWebElement {
  /**
   * Gets the value of borderColor property.
   * The border color of the gauge. Only applies when thresholds are not defined. The default value varies based on theme.
   * @return The value of borderColor property.
   * 
   */
  public getBorderColor(): Promise<string> {
    return this.getProperty<string>("borderColor");
  }

  /**
   * Gets the value of color property.
   * The color of the gauge. Only applies when thresholds are not defined. The default value varies based on theme.
   * @return The value of color property.
   * 
   */
  public getColor(): Promise<string> {
    return this.getProperty<string>("color");
  }

  /**
   * Gets the value of label property.
   * An object defining the label.
   * @return The value of label property.
   * 
   */
  public getLabel(): Promise<OjLedGaugeBase.Label> {
    return this.getProperty<OjLedGaugeBase.Label>("label");
  }

  /**
   * Gets the value of max property.
   * The maximum value of the gauge.
   * @return The value of max property.
   * 
   */
  public getMax(): Promise<number> {
    return this.getProperty<number>("max");
  }

  /**
   * Gets the value of metricLabel property.
   * An object defining the value label.
   * @return The value of metricLabel property.
   * 
   */
  public getMetricLabel(): Promise<OjLedGaugeBase.MetricLabel> {
    return this.getProperty<OjLedGaugeBase.MetricLabel>("metricLabel");
  }

  /**
   * Gets the value of min property.
   * The minimum value of the gauge.
   * @return The value of min property.
   * 
   */
  public getMin(): Promise<number> {
    return this.getProperty<number>("min");
  }

  /**
   * Gets the value of rotation property.
   * The rotation angle for the gauge. Useful for changing the direction of triangle or arrow gauges.
   * @return The value of rotation property.
   * 
   */
  public getRotation(): Promise<number> {
    return this.getProperty<number>("rotation");
  }

  /**
   * Gets the value of size property.
   * Fraction of area to use. Values range from 0 to 1.
   * @return The value of size property.
   * 
   */
  public getSizeProperty(): Promise<number> {
    return this.getProperty<number>("size");
  }

  /**
   * Gets the value of svgClassName property.
   * The CSS style class to apply to the gauge. See the Help documentation for more information.
   * @return The value of svgClassName property.
   * 
   */
  public getSvgClassName(): Promise<string> {
    return this.getProperty<string>("svgClassName");
  }

  /**
   * Gets the value of svgStyle property.
   * The inline style to apply to the gauge. See the Help documentation for more information.
   * @return The value of svgStyle property.
   * 
   */
  public getSvgStyle(): Promise<object> {
    return this.getProperty<object>("svgStyle");
  }

  /**
   * Gets the value of thresholds property.
   * An array of objects specifying the gauge thresholds.
   * @return The value of thresholds property.
   * 
   */
  public getThresholds(): Promise<Array<OjLedGaugeBase.Thresholds>> {
    return this.getProperty<Array<OjLedGaugeBase.Thresholds>>("thresholds");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjLedGaugeBase.Tooltip> {
    return this.getProperty<OjLedGaugeBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjLedGaugeBase.Translations|null> {
    return this.getProperty<OjLedGaugeBase.Translations|null>("translations");
  }

  /**
   * Gets the value of type property.
   * The shape of the LED gauge. Can take the name of a built-in shape or the SVG path commands for a custom shape.
   * @return The value of type property.
   * 
   */
  public getType(): Promise<string> {
    return this.getProperty<string>("type");
  }

  /**
   * Sets the value of "value" property.
   * The metric value.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: number|null): Promise<void> {
    return this.setProperty<number|null>("value", value);
  }

  /**
   * Gets the value of value property.
   * The metric value.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<number|null> {
    return this.getProperty<number|null>("value");
  }

  /**
   * Gets the value of visualEffects property.
   * Defines whether visual effects such as overlays are applied to the gauge.
   * @return The value of visualEffects property.
   * 
   */
  public getVisualEffects(): Promise<string> {
    return this.getProperty<string>("visualEffects");
  }

}

namespace OjLedGaugeBase {
  export interface Label {
    /**
     * The CSS style object defining the style of the label.
     */
    "style": object;
    /**
     * The text for the label.
     */
    "text": string;
  }
  export interface MetricLabel {
    /**
     * The converter used to format the labels. See the Help documentation for more information.
     */
    "converter": object;
    /**
     * Defines if the label is rendered.
     */
    "rendered": string;
    /**
     * The scaling behavior of the labels. When using a converter, scaling should be set to none, as the formatted result may not be compatible with the scaling suffixes.
     */
    "scaling": string;
    /**
     * The CSS style object defining the style of the label.
     */
    "style": object;
    /**
     * The text for the label. If specified, text will overwrite the numeric value that is displayed by default. The converter, scaling, and textType attributes are ignored when text is specified.
     */
    "text": string;
    /**
     * Defines whether the label is a number or a percentage of the total value.
     */
    "textType": string;
  }
  export interface Thresholds {
    /**
     * The border color of the threshold.
     */
    "borderColor": string;
    /**
     * The color of the threshold.
     */
    "color": string;
    /**
     * The upper bound of the threshold. This value is ignored for the final threshold, which uses the maximum value of the gauge.
     */
    "max": number;
    /**
     * Specific description for the threshold and overwrites the shortDesc specified on gauge. This is used for accessibility and also for customizing the tooltip text.
     */
    "shortDesc": string;
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
  }
}
