// @oracle/oraclejet-webdriver/elements/oj-chart/OjChartBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjChartBase = void 0;
const __1 = require("../");
class OjChartBase extends __1.OjWebElement {
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAs() {
        return this.getProperty("as");
    }
    getCoordinateSystem() {
        return this.getProperty("coordinateSystem");
    }
    getData() {
        return this.getProperty("data");
    }
    getDataCursor() {
        return this.getProperty("dataCursor");
    }
    getDataCursorBehavior() {
        return this.getProperty("dataCursorBehavior");
    }
    changeDataCursorPosition(dataCursorPosition) {
        return this.setProperty("dataCursorPosition", dataCursorPosition);
    }
    getDataCursorPosition() {
        return this.getProperty("dataCursorPosition");
    }
    getDataLabel() {
        return this.getProperty("dataLabel");
    }
    getDnd() {
        return this.getProperty("dnd");
    }
    getDragMode() {
        return this.getProperty("dragMode");
    }
    getDrilling() {
        return this.getProperty("drilling");
    }
    getGroupComparator() {
        return this.getProperty("groupComparator");
    }
    getGroups() {
        return this.getProperty("groups");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHideAndShowBehavior() {
        return this.getProperty("hideAndShowBehavior");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getInitialZooming() {
        return this.getProperty("initialZooming");
    }
    getLegend() {
        return this.getProperty("legend");
    }
    getMultiSeriesDrilling() {
        return this.getProperty("multiSeriesDrilling");
    }
    getOrientation() {
        return this.getProperty("orientation");
    }
    getOtherThreshold() {
        return this.getProperty("otherThreshold");
    }
    getOverview() {
        return this.getProperty("overview");
    }
    getPieCenter() {
        return this.getProperty("pieCenter");
    }
    getPlotArea() {
        return this.getProperty("plotArea");
    }
    getPolarGridShape() {
        return this.getProperty("polarGridShape");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSeries() {
        return this.getProperty("series");
    }
    getSeriesComparator() {
        return this.getProperty("seriesComparator");
    }
    getSorting() {
        return this.getProperty("sorting");
    }
    getSplitDualY() {
        return this.getProperty("splitDualY");
    }
    getSplitterPosition() {
        return this.getProperty("splitterPosition");
    }
    getStack() {
        return this.getProperty("stack");
    }
    getStackLabel() {
        return this.getProperty("stackLabel");
    }
    getStackLabelProvider() {
        return this.getProperty("stackLabelProvider");
    }
    getStyleDefaults() {
        return this.getProperty("styleDefaults");
    }
    getTimeAxisType() {
        return this.getProperty("timeAxisType");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTouchResponse() {
        return this.getProperty("touchResponse");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getType() {
        return this.getProperty("type");
    }
    getValueFormats() {
        return this.getProperty("valueFormats");
    }
    getXAxis() {
        return this.getProperty("xAxis");
    }
    getY2Axis() {
        return this.getProperty("y2Axis");
    }
    getYAxis() {
        return this.getProperty("yAxis");
    }
    getZoomAndScroll() {
        return this.getProperty("zoomAndScroll");
    }
    getZoomDirection() {
        return this.getProperty("zoomDirection");
    }
}
exports.OjChartBase = OjChartBase;
