// @oracle/oraclejet-webdriver/elements/oj-input-text/OjInputTextBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjInputTextBase = void 0;
const __1 = require("../");
class OjInputTextBase extends __1.OjWebElement {
    getAsyncValidators() {
        return this.getProperty("asyncValidators");
    }
    getAutocomplete() {
        return this.getProperty("autocomplete");
    }
    getAutofocus() {
        return this.getProperty("autofocus");
    }
    getClearIcon() {
        return this.getProperty("clearIcon");
    }
    getConverter() {
        return this.getProperty("converter");
    }
    getDescribedBy() {
        return this.getProperty("describedBy");
    }
    getDisabled() {
        return this.getProperty("disabled");
    }
    getDisplayOptions() {
        return this.getProperty("displayOptions");
    }
    getHelp() {
        return this.getProperty("help");
    }
    getHelpHints() {
        return this.getProperty("helpHints");
    }
    getLabelEdge() {
        return this.getProperty("labelEdge");
    }
    getLabelHint() {
        return this.getProperty("labelHint");
    }
    getLabelledBy() {
        return this.getProperty("labelledBy");
    }
    getLength() {
        return this.getProperty("length");
    }
    getList() {
        return this.getProperty("list");
    }
    changeMessagesCustom(messagesCustom) {
        return this.setProperty("messagesCustom", messagesCustom);
    }
    getMessagesCustom() {
        return this.getProperty("messagesCustom");
    }
    getName() {
        return this.getProperty("name");
    }
    getPlaceholder() {
        return this.getProperty("placeholder");
    }
    getRawValue() {
        return this.getProperty("rawValue");
    }
    getReadonly() {
        return this.getProperty("readonly");
    }
    getRequired() {
        return this.getProperty("required");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getUserAssistanceDensity() {
        return this.getProperty("userAssistanceDensity");
    }
    getValid() {
        return this.getProperty("valid");
    }
    getValidators() {
        return this.getProperty("validators");
    }
    changeValue(value) {
        return this.setProperty("value", value);
    }
    getValue() {
        return this.getProperty("value");
    }
    getVirtualKeyboard() {
        return this.getProperty("virtualKeyboard");
    }
}
exports.OjInputTextBase = OjInputTextBase;
