// @oracle/oraclejet-webdriver/elements/oj-input-text/OjInputTextBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjInputText, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjInputTextBase extends OjWebElement {
  /**
   * Gets the value of asyncValidators property.
   * Specifies a list of asynchronous validators used by the component when performing validation. Use async-validators when you need to perform some validation work on the server. See the Help documentation for more information.
   * @return The value of asyncValidators property.
   * @deprecated Since 8.0.0. Use the validators property instead for either regular Validators or AsyncValidators.
   */
  public getAsyncValidators(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("asyncValidators");
  }

  /**
   * Gets the value of autocomplete property.
   * Specifies a component's autocomplete state. See the Help documentation for more information.
   * @return The value of autocomplete property.
   * 
   */
  public getAutocomplete(): Promise<string> {
    return this.getProperty<string>("autocomplete");
  }

  /**
   * Gets the value of autofocus property.
   * Specifies whether the component will get input focus when the page is loaded. See the Help documentation for more information.
   * @return The value of autofocus property.
   * 
   */
  public getAutofocus(): Promise<boolean> {
    return this.getProperty<boolean>("autofocus");
  }

  /**
   * Gets the value of clearIcon property.
   * Specifies if an icon to clear the input field should be visible.
   * @return The value of clearIcon property.
   * 
   */
  public getClearIcon(): Promise<string> {
    return this.getProperty<string>("clearIcon");
  }

  /**
   * Gets the value of converter property.
   * An object that converts the value. See the Help documentation for more information.
   * @return The value of converter property.
   * 
   */
  public getConverter(): Promise<object|null> {
    return this.getProperty<object|null>("converter");
  }

  /**
   * Gets the value of describedBy property.
   * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
   * @return The value of describedBy property.
   * 
   */
  public getDescribedBy(): Promise<string> {
    return this.getProperty<string>("describedBy");
  }

  /**
   * Gets the value of disabled property.
   * Specifies whether the component is disabled. The default is false.
   * @return The value of disabled property.
   * 
   */
  public getDisabled(): Promise<boolean> {
    return this.getProperty<boolean>("disabled");
  }

  /**
   * Gets the value of displayOptions property.
   * Display options for auxiliary content that determines whether or not it should be displayed.
   * @return The value of displayOptions property.
   * 
   */
  public getDisplayOptions(): Promise<OjInputTextBase.DisplayOptions> {
    return this.getProperty<OjInputTextBase.DisplayOptions>("displayOptions");
  }

  /**
   * Gets the value of help property.
   * Form component help information.
   * @return The value of help property.
   * 
   */
  public getHelp(): Promise<OjInputTextBase.Help> {
    return this.getProperty<OjInputTextBase.Help>("help");
  }

  /**
   * Gets the value of helpHints property.
   * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
   * @return The value of helpHints property.
   * 
   */
  public getHelpHints(): Promise<OjInputTextBase.HelpHints> {
    return this.getProperty<OjInputTextBase.HelpHints>("helpHints");
  }

  /**
   * Gets the value of labelEdge property.
   * Defines how the label of a component is created. See the Help documentation for more information.
   * @return The value of labelEdge property.
   * 
   */
  public getLabelEdge(): Promise<string> {
    return this.getProperty<string>("labelEdge");
  }

  /**
   * Gets the value of labelHint property.
   * Represents a hint for oj-form-layout element to render a label on the editable component.
   * @return The value of labelHint property.
   * 
   */
  public getLabelHint(): Promise<string> {
    return this.getProperty<string>("labelHint");
  }

  /**
   * Gets the value of labelledBy property.
   * The oj-label sets the labelledBy property programmatically on the form component. See the Help documentation for more information.
   * @return The value of labelledBy property.
   * 
   */
  public getLabelledBy(): Promise<string|null> {
    return this.getProperty<string|null>("labelledBy");
  }

  /**
   * Gets the value of length property.
   * An object whose properties describe the maximum length attributes. // * @ojtsignore tsdefonly
   * @return The value of length property.
   * 
   */
  public getLength(): Promise<OjInputTextBase.Length> {
    return this.getProperty<OjInputTextBase.Length>("length");
  }

  /**
   * Gets the value of list property.
   * Specifies a list of pre-defined options to present to the user. See the Help documentation for more information.
   * @return The value of list property.
   * 
   */
  public getList(): Promise<string> {
    return this.getProperty<string>("list");
  }

  /**
   * Sets the value of "messagesCustom" property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @param messagesCustom The value to set for "messagesCustom"
   * 
   */
  public changeMessagesCustom(messagesCustom: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("messagesCustom", messagesCustom);
  }

  /**
   * Gets the value of messagesCustom property.
   * A list of messages added by an application to the component. See the Help documentation for more information.
   * @return The value of messagesCustom property.
   * 
   */
  public getMessagesCustom(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("messagesCustom");
  }

  /**
   * Gets the value of name property.
   * Specifies the name of the component.
   * @return The value of name property.
   * @deprecated Since 6.0.0. JET does not use form submit, so this is not needed.
   */
  public getName(): Promise<string> {
    return this.getProperty<string>("name");
  }

  /**
   * Gets the value of placeholder property.
   * The placeholder text to set on the element.
   * @return The value of placeholder property.
   * 
   */
  public getPlaceholder(): Promise<string> {
    return this.getProperty<string>("placeholder");
  }

  /**
   * Gets the value of rawValue property.
   * Read-only property used for retrieving the current value from the input field in string form. See the Help documentation for more information.
   * @return The value of rawValue property.
   * 
   */
  public getRawValue(): Promise<string> {
    return this.getProperty<string>("rawValue");
  }

  /**
   * Gets the value of readonly property.
   * Specifies whether the component is read-only. A read-only element cannot be modified, but user interaction is allowed. See the Help documentation for more information.
   * @return The value of readonly property.
   * 
   */
  public getReadonly(): Promise<boolean> {
    return this.getProperty<boolean>("readonly");
  }

  /**
   * Gets the value of required property.
   * Specifies whether the component is required or optional. See the Help documentation for more information.
   * @return The value of required property.
   * 
   */
  public getRequired(): Promise<boolean> {
    return this.getProperty<boolean>("required");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjInputTextBase.Translations|null> {
    return this.getProperty<OjInputTextBase.Translations|null>("translations");
  }

  /**
   * Gets the value of userAssistanceDensity property.
   * Specifies the density of the form component's user assistance presentation.
   * @return The value of userAssistanceDensity property.
   * 
   */
  public getUserAssistanceDensity(): Promise<string> {
    return this.getProperty<string>("userAssistanceDensity");
  }

  /**
   * Gets the value of valid property.
   * The validity state of the component
   * @return The value of valid property.
   * 
   */
  public getValid(): Promise<string> {
    return this.getProperty<string>("valid");
  }

  /**
   * Gets the value of validators property.
   * Specifies a list of synchronous validators for performing validation by the element. See the Help documentation for more information.
   * @return The value of validators property.
   * 
   */
  public getValidators(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("validators");
  }

  /**
   * Sets the value of "value" property.
   * The value of the component.
   * @param value The value to set for "value"
   * 
   */
  public changeValue(value: any): Promise<void> {
    return this.setProperty<any>("value", value);
  }

  /**
   * Gets the value of value property.
   * The value of the component.
   * @return The value of value property.
   * 
   */
  public getValue(): Promise<any> {
    return this.getProperty<any>("value");
  }

  /**
   * Gets the value of virtualKeyboard property.
   * The type of virtual keyboard to display for entering a value on mobile browsers. See the Help documentation for more information.
   * @return The value of virtualKeyboard property.
   * 
   */
  public getVirtualKeyboard(): Promise<string> {
    return this.getProperty<string>("virtualKeyboard");
  }

}

namespace OjInputTextBase {
  export interface DisplayOptions {
    /**
     * Display options for auxiliary converter hint text that determines whether it should be displayed.
     */
    "converterHint": Array<string>|string;
    /**
     * Display options for auxiliary help instruction text that determines whether it should be displayed.
     */
    "helpInstruction": Array<string>|string;
    /**
     * Display options for auxiliary message text that determines whether it should be displayed.
     */
    "messages": Array<string>|string;
    /**
     * Display options for auxiliary validator hint text that determines whether it should be displayed.
     */
    "validatorHint": Array<string>|string;
  }
  export interface Help {
    /**
     * Represents advisory information for the component, such as would be appropriate for a tooltip.
     */
    "instruction": string;
  }
  export interface HelpHints {
    /**
     * Hint for help definition text associated with the label.
     */
    "definition": string;
    /**
     * Help source URL associated with the component.
     */
    "source": string;
  }
  export interface Length {
    /**
     * Specifies the manner in which the input text characters has to be counted.
     */
    "countBy": string;
    /**
     * Specifies the maximum number of characters to be entered in the input text.
     */
    "max": number|null;
  }
  export interface Translations {
    /**
     * Message to announce to AT via aria-live when the length.max is exceeded.
     */
    "accessibleMaxLengthExceeded": string;
    /**
     * Message to annnouce to AT via aria-live for the remaining character count.
     */
    "accessibleMaxLengthRemaining": string;
    /**
     * Provides properties to customize the message text used by the implicit regexp validator associated to the InputText and TextArea components.
     */
    "regexp": OjInputTextBase.TranslationsRegexp;
    /**
     * Provides properties to customize the summary, detail and hint text used by the implicit required validator associated to any editable component that supports the required option.
     */
    "required": OjInputTextBase.TranslationsRequired;
  }
  export interface TranslationsRegexp {
    /**
     * Provides properties to customize the error message detail used by the implicit regexp validator associated to the InputText and TextArea components.
     */
    "messageDetail": string;
    /**
     * Provides properties to customize the error message summary used by the implicit regexp validator associated to the InputText and TextArea components.
     */
    "messageSummary": string;
  }
  export interface TranslationsRequired {
    /**
     * Hint text used by required validation error.
     */
    "hint": string;
    /**
     * Message text that describes the details of the required validation error.
     */
    "messageDetail": string;
    /**
     * Message text for summarizing a required validation error.
     */
    "messageSummary": string;
  }
}
