// @oracle/oraclejet-webdriver/elements/oj-gantt/OjGanttBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjGanttBase = void 0;
const __1 = require("../");
class OjGanttBase extends __1.OjWebElement {
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAs() {
        return this.getProperty("as");
    }
    getAxisPosition() {
        return this.getProperty("axisPosition");
    }
    getDependencies() {
        return this.getProperty("dependencies");
    }
    getDependencyData() {
        return this.getProperty("dependencyData");
    }
    getDnd() {
        return this.getProperty("dnd");
    }
    getDragMode() {
        return this.getProperty("dragMode");
    }
    getEnd() {
        return this.getProperty("end");
    }
    getGridlines() {
        return this.getProperty("gridlines");
    }
    getMajorAxis() {
        return this.getProperty("majorAxis");
    }
    getMinorAxis() {
        return this.getProperty("minorAxis");
    }
    getReferenceObjects() {
        return this.getProperty("referenceObjects");
    }
    getRowAxis() {
        return this.getProperty("rowAxis");
    }
    getRowDefaults() {
        return this.getProperty("rowDefaults");
    }
    getRows() {
        return this.getProperty("rows");
    }
    changeScrollPosition(scrollPosition) {
        return this.setProperty("scrollPosition", scrollPosition);
    }
    getScrollPosition() {
        return this.getProperty("scrollPosition");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getStart() {
        return this.getProperty("start");
    }
    getTaskData() {
        return this.getProperty("taskData");
    }
    getTaskDefaults() {
        return this.getProperty("taskDefaults");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getValueFormats() {
        return this.getProperty("valueFormats");
    }
    getViewportEnd() {
        return this.getProperty("viewportEnd");
    }
    getViewportStart() {
        return this.getProperty("viewportStart");
    }
}
exports.OjGanttBase = OjGanttBase;
