// @oracle/oraclejet-webdriver/elements/oj-table/OjTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTable = void 0;
const OjTableBase_1 = require("./OjTableBase");
const __1 = require("../../");
class OjTable extends OjTableBase_1.OjTableBase {
    async changeSelected(selected) {
        await this.whenReady();
        await this.getDriver().executeScript(`
        const ele = arguments[0];
        const selected = arguments[1];
        require([\'ojs/ojkeyset\'], function(keySet) {
          ele.selected = {
            row: new keySet.KeySetImpl(selected.row),
            column: new keySet.KeySetImpl(selected.column)
          };
        });
      `, this, selected);
    }
    async getSelected() {
        await this.whenReady();
        const selected = await this.getDriver().executeScript(`
      const ele = arguments[0];
      const selected = {
        row: Array.from(ele.selected.row.values()),
        column: Array.from(ele.selected.column.values())
      }
      return selected;
    `, this);
        return selected;
    }
    async findCell(itemLocator) {
        await this.changeScrollPosition({
            rowIndex: itemLocator.rowIndex,
            columnIndex: undefined,
            offsetX: undefined,
            offsetY: undefined,
            rowKey: undefined,
            columnKey: undefined,
            x: undefined,
            y: undefined
        });
        const el = await this.getDriver().executeScript(`
      const element = arguments[0];
      const columnData = element.columns[${itemLocator.columnIndex}];
      const defaultCellTemplate = element.querySelectorAll('[slot="cellTemplate"]')[0];
      if (columnData.template == null && columnData.renderer == null &&
          defaultCellTemplate == null) {
        return null;
      } else {
        return element.getNodeBySubId({ subId: 'oj-table-cell', rowIndex: ${itemLocator.rowIndex}, columnIndex: ${itemLocator.columnIndex} });
      }
    `, this);
        if (el) {
            return __1.slotProxy(el, this);
        }
        else {
            throw Error(`No corresponding cell template or renderer found for the itemLocator`);
        }
    }
    async findRow(itemLocator) {
        await this.changeScrollPosition({
            rowIndex: itemLocator.rowIndex,
            columnIndex: undefined,
            offsetX: undefined,
            offsetY: undefined,
            rowKey: undefined,
            columnKey: undefined,
            x: undefined,
            y: undefined
        });
        const el = await this.getDriver().executeScript(`
      const element = arguments[0];
      const rowTemplate = element.querySelectorAll('[slot="rowTemplate"]')[0];
      const rowRenderer = element.rowRenderer;
      if (rowTemplate == null && rowRenderer == null) {
        return null;
      }
      const rowIndex = arguments[1];
      const rows = element.querySelectorAll('.oj-table-body-row');
      return rows[rowIndex];
    `, this, itemLocator.rowIndex);
        if (el) {
            return __1.slotProxy(el, this);
        }
        else {
            throw Error(`No corresponding row template or renderer found for the itemLocator`);
        }
    }
    async findHeader(itemLocator) {
        const el = await this.getDriver().executeScript(`
      const element = arguments[0];
      const columnData = element.columns[${itemLocator.columnIndex}];
      const defaultHeaderTemplate = element.querySelectorAll('[slot="headerTemplate"]')[0];
      if (columnData.headerTemplate == null && 
          columnData.headerRenderer == null && defaultHeaderTemplate == null) {
        return null;
      }
      return element.getNodeBySubId({ subId: 'oj-table-header', index: ${itemLocator.columnIndex} })
    `, this);
        if (el) {
            return __1.slotProxy(el, this);
        }
        else {
            throw Error(`No corresponding header template or renderer found for the itemLocator`);
        }
    }
    async findFooter(itemLocator) {
        const el = await this.getDriver().executeScript(`
      const element = arguments[0];
      const columnData = element.columns[${itemLocator.columnIndex}];
      const defaultFooterTemplate = element.querySelectorAll('[slot="footerTemplate"]')[0];
      if (columnData.footerTemplate == null &&
          columnData.footerRenderer == null && defaultFooterTemplate == null) {
        return null;
      }
      return element.getNodeBySubId({ subId: 'oj-table-footer', index: ${itemLocator.columnIndex} })
    `, this);
        if (el) {
            return __1.slotProxy(el, this);
        }
        else {
            throw Error(`No corresponding footer template or renderer found for the itemLocator`);
        }
    }
}
exports.OjTable = OjTable;
