// @oracle/oraclejet-webdriver/elements/oj-table/OjTable.ts
import {By, WebElement} from "selenium-webdriver";
import {OjTableBase} from "./OjTableBase";
import { slotProxy, SlotProxy } from "../../";

/**
 * This is the component WebElement for oj-table. Retrieve an instance of the class
 * using its convenience function:
 * ```javascript
 * await ojTable(driver, By.css('oj-table'))
 * ```
 */
export class OjTable extends OjTableBase {
  /**
   * Sets the value of "selected" property.
   * Specifies the current selected rows and/or columns in the table. See the Help documentation for more information.
   * @param selected The value to set for "selected"
   * @override
   * @typeparam K Type of keys
   */
  async changeSelected<K>(selected: { row: Array<K>, column: Array<K>}) {
    await this.whenReady();
    await this.getDriver().executeScript(`
        const ele = arguments[0];
        const selected = arguments[1];
        require([\'ojs/ojkeyset\'], function(keySet) {
          ele.selected = {
            row: new keySet.KeySetImpl(selected.row),
            column: new keySet.KeySetImpl(selected.column)
          };
        });
      `, this, selected);
  }

  /**
   * Gets the value of "selected" property.
   * Specifies the current selected rows and/or columns in the table. See the Help documentation for more information.
   * @override
   * @typeparam K Type of keys
   * @return The value of "selected" property.
   */
  async getSelected<K>(): Promise<{row: Array<K>, column: Array<K>}> {
    await this.whenReady();
    const selected: {row: Array<K>, column: Array<K>}  = await this.getDriver().executeScript(`
      const ele = arguments[0];
      const selected = {
        row: Array.from(ele.selected.row.values()),
        column: Array.from(ele.selected.column.values())
      }
      return selected;
    `, this);
    return selected;
  }

  /**
   * Retrieve a SlotProxy which represents a single cell.
   * Should be used only if cell contents are rendered using cell template/renderer/slot. Otherwise method will throw an exception.
   * @param itemLocator The row and column indexes of the cell.
   */
  async findCell(itemLocator: { rowIndex: number, columnIndex: number }): Promise<SlotProxy> {
    // Call changeScrollPosition to scroll the row into view and ensure it's
    // rendered in DOM
    await this.changeScrollPosition({
      rowIndex: itemLocator.rowIndex,
      columnIndex: undefined,
      offsetX: undefined,
      offsetY: undefined,
      rowKey: undefined,
      columnKey: undefined,
      x: undefined,
      y: undefined
    });
    const el = await this.getDriver().executeScript<WebElement>(`
      const element = arguments[0];
      const columnData = element.columns[${itemLocator.columnIndex}];
      const defaultCellTemplate = element.querySelectorAll('[slot="cellTemplate"]')[0];
      if (columnData.template == null && columnData.renderer == null &&
          defaultCellTemplate == null) {
        return null;
      } else {
        return element.getNodeBySubId({ subId: 'oj-table-cell', rowIndex: ${itemLocator.rowIndex}, columnIndex: ${itemLocator.columnIndex} });
      }
    `, this);
    if (el) {
      return slotProxy(el, this);
    } else {
      throw Error(`No corresponding cell template or renderer found for the itemLocator`);
    }
  }

  /**
   * Retrieve a SlotProxy which represents a single row.
   * Should be used only if row is rendered using rowTemplate slot or renderer. Otherwise method will throw an exception.
   * @param itemLocator The index of the row.
   */
  async findRow(itemLocator: { rowIndex: number}): Promise<SlotProxy> {
    // Call changeScrollPosition to scroll the row into view and ensure it's
    // rendered in DOM
    await this.changeScrollPosition({
      rowIndex: itemLocator.rowIndex,
      columnIndex: undefined,
      offsetX: undefined,
      offsetY: undefined,
      rowKey: undefined,
      columnKey: undefined,
      x: undefined,
      y: undefined
    });
    const el = await this.getDriver().executeScript<WebElement>(`
      const element = arguments[0];
      const rowTemplate = element.querySelectorAll('[slot="rowTemplate"]')[0];
      const rowRenderer = element.rowRenderer;
      if (rowTemplate == null && rowRenderer == null) {
        return null;
      }
      const rowIndex = arguments[1];
      const rows = element.querySelectorAll('.oj-table-body-row');
      return rows[rowIndex];
    `, this, itemLocator.rowIndex);
    if (el) {
      return slotProxy(el, this);
    } else {
      throw Error(`No corresponding row template or renderer found for the itemLocator`);
    }
  }

  /**
   * Retrieve a SlotProxy which represents the given header cell.
   * Should be used only if header is rendered using template/renderer/slot. Otherwise method will throw an exception.
   * @param itemLocator The column index of the header cell.
   */
  async findHeader(itemLocator: { columnIndex: number }): Promise<SlotProxy> {
    const el = await this.getDriver().executeScript<WebElement>(`
      const element = arguments[0];
      const columnData = element.columns[${itemLocator.columnIndex}];
      const defaultHeaderTemplate = element.querySelectorAll('[slot="headerTemplate"]')[0];
      if (columnData.headerTemplate == null && 
          columnData.headerRenderer == null && defaultHeaderTemplate == null) {
        return null;
      }
      return element.getNodeBySubId({ subId: 'oj-table-header', index: ${itemLocator.columnIndex} })
    `, this);
    if (el) {
      return slotProxy(el, this);
    } else {
      throw Error(`No corresponding header template or renderer found for the itemLocator`);
    }
  }

  /**
   * Retrieve a SlotProxy which represents the given footer cell.
   * Should be used only if footer is rendered using template/renderer/slot. Otherwise method will throw an exception.
   * @param itemLocator The column index of the footer cell.
   */
  async findFooter(itemLocator: { columnIndex: number }): Promise<SlotProxy> {
    const el = await this.getDriver().executeScript<WebElement>(`
      const element = arguments[0];
      const columnData = element.columns[${itemLocator.columnIndex}];
      const defaultFooterTemplate = element.querySelectorAll('[slot="footerTemplate"]')[0];
      if (columnData.footerTemplate == null &&
          columnData.footerRenderer == null && defaultFooterTemplate == null) {
        return null;
      }
      return element.getNodeBySubId({ subId: 'oj-table-footer', index: ${itemLocator.columnIndex} })
    `, this);
    if (el) {
      return slotProxy(el, this);
    } else {
      throw Error(`No corresponding footer template or renderer found for the itemLocator`);
    }
  }
}
