// @oracle/oraclejet-webdriver/elements/oj-table/OjTableBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjTable, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjTableBase extends OjWebElement {
  /**
   * Gets the value of accessibility property.
   * Accessibility attributes.
   * @return The value of accessibility property.
   * 
   */
  public getAccessibility(): Promise<OjTableBase.Accessibility|null> {
    return this.getProperty<OjTableBase.Accessibility|null>("accessibility");
  }

  /**
   * Gets the value of as property.
   * An alias for the '$current' context variable passed to the content of the cell template.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Sets the value of "columns" property.
   * An array of column definitions. See the Help documentation for more information.
   * @param columns The value to set for "columns"
   * 
   */
  public changeColumns(columns: Array<OjTableBase.Columns>|null): Promise<void> {
    return this.setProperty<Array<OjTableBase.Columns>|null>("columns", columns);
  }

  /**
   * Gets the value of columns property.
   * An array of column definitions. See the Help documentation for more information.
   * @return The value of columns property.
   * 
   */
  public getColumns(): Promise<Array<OjTableBase.Columns>|null> {
    return this.getProperty<Array<OjTableBase.Columns>|null>("columns");
  }

  /**
   * Gets the value of columnsDefault property.
   * Default values to apply to all column objects.
   * @return The value of columnsDefault property.
   * 
   */
  public getColumnsDefault(): Promise<OjTableBase.ColumnsDefault|null> {
    return this.getProperty<OjTableBase.ColumnsDefault|null>("columnsDefault");
  }

  /**
   * Sets the value of "currentRow" property.
   * Specifies the row that currently has keyboard focus. See the Help documentation for more information.
   * @param currentRow The value to set for "currentRow"
   * 
   */
  public changeCurrentRow(currentRow: object): Promise<void> {
    return this.setProperty<object>("currentRow", currentRow);
  }

  /**
   * Gets the value of currentRow property.
   * Specifies the row that currently has keyboard focus. See the Help documentation for more information.
   * @return The value of currentRow property.
   * 
   */
  public getCurrentRow(): Promise<object> {
    return this.getProperty<object>("currentRow");
  }

  /**
   * Gets the value of data property.
   * Specifies the data for the table. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of display property.
   * Specifies whether to display this table in list or grid mode. See the Help documentation for more information.
   * @return The value of display property.
   * 
   */
  public getDisplay(): Promise<string> {
    return this.getProperty<string>("display");
  }

  /**
   * Gets the value of dnd property.
   * Specifies drag and drop features. See the Help documentation for more information.
   * @return The value of dnd property.
   * 
   */
  public getDnd(): Promise<OjTableBase.Dnd> {
    return this.getProperty<OjTableBase.Dnd>("dnd");
  }

  /**
   * Gets the value of editMode property.
   * Specifies if the table is read-only or editable. See the Help documentation for more information.
   * @return The value of editMode property.
   * 
   */
  public getEditMode(): Promise<string> {
    return this.getProperty<string>("editMode");
  }

  /**
   * Sets the value of "editRow" property.
   * Specifies the row that is currently being edited. See the Help documentation for more information.
   * @param editRow The value to set for "editRow"
   * 
   */
  public changeEditRow(editRow: object): Promise<void> {
    return this.setProperty<object>("editRow", editRow);
  }

  /**
   * Gets the value of editRow property.
   * Specifies the row that is currently being edited. See the Help documentation for more information.
   * @return The value of editRow property.
   * 
   */
  public getEditRow(): Promise<object> {
    return this.getProperty<object>("editRow");
  }

  /**
   * Gets the value of firstSelectedRow property.
   * Read-only property used for retrieving the key and data of the first selected row. See the Help documentation for more information.
   * @return The value of firstSelectedRow property.
   * 
   */
  public getFirstSelectedRow(): Promise<OjTableBase.FirstSelectedRow> {
    return this.getProperty<OjTableBase.FirstSelectedRow>("firstSelectedRow");
  }

  /**
   * Gets the value of horizontalGridVisible property.
   * Specifies the visibility of the horizontal gridlines. See the Help documentation for more information.
   * @return The value of horizontalGridVisible property.
   * 
   */
  public getHorizontalGridVisible(): Promise<string> {
    return this.getProperty<string>("horizontalGridVisible");
  }

  /**
   * Gets the value of layout property.
   * The column sizing method used for the Table and its columns.
   * @return The value of layout property.
   * 
   */
  public getLayout(): Promise<string> {
    return this.getProperty<string>("layout");
  }

  /**
   * Gets the value of rowRenderer property.
   * A function that returns row content. The function takes a context argument, provided by the table. See the Help documentation for more information.
   * @return The value of rowRenderer property.
   * 
   */
  public getRowRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("rowRenderer");
  }

  /**
   * Gets the value of scrollPolicy property.
   * Specifies how data are fetched as user scrolls down the table.
   * @return The value of scrollPolicy property.
   * 
   */
  public getScrollPolicy(): Promise<string> {
    return this.getProperty<string>("scrollPolicy");
  }

  /**
   * Gets the value of scrollPolicyOptions property.
   * Specifies fetch options for scrolling behaviors that trigger data fetches.
   * @return The value of scrollPolicyOptions property.
   * 
   */
  public getScrollPolicyOptions(): Promise<OjTableBase.ScrollPolicyOptions|null> {
    return this.getProperty<OjTableBase.ScrollPolicyOptions|null>("scrollPolicyOptions");
  }

  /**
   * Sets the value of "scrollPosition" property.
   * Specifies the scroll position of the table. See the Help documentation for more information.
   * @param scrollPosition The value to set for "scrollPosition"
   * 
   */
  public changeScrollPosition(scrollPosition: OjTableBase.ScrollPosition): Promise<void> {
    return this.setProperty<OjTableBase.ScrollPosition>("scrollPosition", scrollPosition);
  }

  /**
   * Gets the value of scrollPosition property.
   * Specifies the scroll position of the table. See the Help documentation for more information.
   * @return The value of scrollPosition property.
   * 
   */
  public getScrollPosition(): Promise<OjTableBase.ScrollPosition> {
    return this.getProperty<OjTableBase.ScrollPosition>("scrollPosition");
  }

  /**
   * Gets the value of scrollToKey property.
   * Specifies the behavior when Table needs to scroll to a position based on an item key.
   * @return The value of scrollToKey property.
   * 
   */
  public getScrollToKey(): Promise<string|null> {
    return this.getProperty<string|null>("scrollToKey");
  }

  /**
   * Sets the value of "selected" property.
   * Specifies the current selected rows and/or columns in the table. See the Help documentation for more information.
   * @param selected The value to set for "selected"
   * 
   */
  public changeSelected(selected: OjTableBase.Selected): Promise<void> {
    return this.setProperty<OjTableBase.Selected>("selected", selected);
  }

  /**
   * Gets the value of selected property.
   * Specifies the current selected rows and/or columns in the table. See the Help documentation for more information.
   * @return The value of selected property.
   * 
   */
  public getSelected(): Promise<OjTableBase.Selected> {
    return this.getProperty<OjTableBase.Selected>("selected");
  }

  /**
   * Sets the value of "selection" property.
   * Specifies the current selections in the table. See the Help documentation for more information.
   * @param selection The value to set for "selection"
   * @deprecated Since 7.0.0. Use selected attribute instead.
   */
  public changeSelection(selection: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * Specifies the current selections in the table. See the Help documentation for more information.
   * @return The value of selection property.
   * @deprecated Since 7.0.0. Use selected attribute instead.
   */
  public getSelection(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the row and column selection modes.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<unknown|null> {
    return this.getProperty<unknown|null>("selectionMode");
  }

  /**
   * Gets the value of selectionRequired property.
   * Specifies whether selection is required on the Table.
   * @return The value of selectionRequired property.
   * 
   */
  public getSelectionRequired(): Promise<boolean> {
    return this.getProperty<boolean>("selectionRequired");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjTableBase.Translations|null> {
    return this.getProperty<OjTableBase.Translations|null>("translations");
  }

  /**
   * Gets the value of verticalGridVisible property.
   * Specifies the visibility of the vertical gridlines. See the Help documentation for more information.
   * @return The value of verticalGridVisible property.
   * 
   */
  public getVerticalGridVisible(): Promise<string> {
    return this.getProperty<string>("verticalGridVisible");
  }

}

namespace OjTableBase {
  export interface Accessibility {
    /**
     * Specifies the column ids to be used as the row headers by screen readers. See the Help documentation for more information.
     */
    "rowHeader": string|Array<string>;
  }
  export interface Columns {
    /**
     * The CSS class to apply to the column cells.
     */
    "className": string|null;
    /**
     * The data field that this column refers to.
     */
    "field": string|null;
    /**
     * The CSS class to apply to the footer cell.
     */
    "footerClassName": string|null;
    /**
     * A function that renders the content of the footer cell. The function takes a context argument, provided by the Table. See the Help documentation for more information.
     */
    "footerRenderer": null|null;
    /**
     * The CSS styling to apply to the footer cell.
     */
    "footerStyle": string|null;
    /**
     * The slot name used to specify the template for rendering the footer cell. See the Help documentation for more information.
     */
    "footerTemplate": string|null;
    /**
     * The edge that this column is frozen against. If present, the column will not be allowed to scroll past the specified edge of the Table.
     */
    "frozenEdge": string|null;
    /**
     * The CSS class to apply to the column header text.
     */
    "headerClassName": string|null;
    /**
     * A function that renders the content of the header. The function takes a context argument, provided by the Table. See the Help documentation for more information.
     */
    "headerRenderer": null|null;
    /**
     * The CSS styling to apply to the column header text.
     */
    "headerStyle": string|null;
    /**
     * The slot name used to specify the template for rendering the header cell. See the Help documentation for more information.
     */
    "headerTemplate": string|null;
    /**
     * Text to display in the header of the column.
     */
    "headerText": string|null;
    /**
     * The identifier for the column.
     */
    "id": string|null;
    /**
     * The maximum width style string of the column. See the Help documentation for more information.
     */
    "maxWidth": string|number|null;
    /**
     * The minimum width style string of the column. See the Help documentation for more information.
     */
    "minWidth": string|number|null;
    /**
     * A function that renders the content of the cell. The function takes a context argument, provided by the Table. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Enable or disable width resizing along the column end headers.
     */
    "resizable": string;
    /**
     * Specifies the row attribute used for sorting when a sort is invoked on this column. See the Help documentation for more information.
     */
    "sortProperty": string|null;
    /**
     * Specifies whether a column is sortable. See the Help documentation for more information.
     */
    "sortable": string;
    /**
     * The CSS styling to apply to the column cells.
     */
    "style": string|null;
    /**
     * The slot name used to specify the template for rendering the cell. See the Help documentation for more information.
     */
    "template": string|null;
    /**
     * The sizing weight of the column in a fixed layout. See the Help documentation for more information.
     */
    "weight": number|null;
    /**
     * The width style string of the column. See the Help documentation for more information.
     */
    "width": string|number|null;
  }
  export interface ColumnsDefault {
    /**
     * The CSS class to apply to the column cells.
     */
    "className": string|null;
    /**
     * The data field that this column refers to.
     */
    "field": string|null;
    /**
     * The CSS class to apply to the footer cell.
     */
    "footerClassName": string|null;
    /**
     * A function that renders the content of the footer cell. The function takes a context argument, provided by the Table. See the Help documentation for more information.
     */
    "footerRenderer": null|null;
    /**
     * The CSS styling to apply to the footer cell.
     */
    "footerStyle": string|null;
    /**
     * The slot name used to specify the template for rendering the footer cell. See the Help documentation for more information.
     */
    "footerTemplate": string|null;
    /**
     * The CSS class to apply to the column header text.
     */
    "headerClassName": string|null;
    /**
     * A function that renders the content of the header. The function takes a context argument, provided by the Table. See the Help documentation for more information.
     */
    "headerRenderer": null|null;
    /**
     * The CSS styling to apply to the column header text.
     */
    "headerStyle": string|null;
    /**
     * The slot name used to specify the template for rendering the header cell. See the Help documentation for more information.
     */
    "headerTemplate": string|null;
    /**
     * Text to display in the header of the column.
     */
    "headerText": string|null;
    /**
     * The maximum width style string of the column. See the Help documentation for more information.
     */
    "maxWidth": string|number|null;
    /**
     * The minimum width style string of the column. See the Help documentation for more information.
     */
    "minWidth": string|number|null;
    /**
     * A function that renders the content of the cell. The function takes a context argument, provided by the Table. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Enable or disable width resizing along the column end headers.
     */
    "resizable": string;
    /**
     * Specifies the row attribute used for sorting when a sort is invoked on this column. See the Help documentation for more information.
     */
    "sortProperty": string|null;
    /**
     * Specifies whether a column is sortable. See the Help documentation for more information.
     */
    "sortable": string;
    /**
     * The CSS styling to apply to the column cells.
     */
    "style": string|null;
    /**
     * The slot name used to specify the template for rendering the cell. See the Help documentation for more information.
     */
    "template": string|null;
    /**
     * The sizing weight of the column in a fixed layout. See the Help documentation for more information.
     */
    "weight": number|null;
    /**
     * The width style string of the column. See the Help documentation for more information.
     */
    "width": string|number|null;
  }
  export interface Dnd {
    /**
     * An object that describes drag functionality.
     */
    "drag": OjTableBase.DndDrag;
    /**
     * An object that describes drop functionality.
     */
    "drop": OjTableBase.DndDrop;
    /**
     * An object that describes reorder functionality.
     */
    "reorder": OjTableBase.DndReorder;
  }
  export interface DndDrag {
    /**
     * An object that describes drag functionality for a selected set of rows. See the Help documentation for more information.
     */
    "rows": OjTableBase.DndDragRows;
  }
  export interface DndDragRows {
    /**
     * The MIME types to use for the dragged data in the dataTransfer object. This can be a string if there is only one type, or an array of strings if multiple types are needed.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDrop {
    /**
     * An object that describes drop functionality for a selected set of columns. See the Help documentation for more information.
     */
    "columns": OjTableBase.DndDropColumns;
    /**
     * An object that describes drop functionality for a selected set of rows. See the Help documentation for more information.
     */
    "rows": OjTableBase.DndDropRows;
  }
  export interface DndDropColumns {
    /**
     * A data type or an array of data types this element can accept.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDropRows {
    /**
     * A data type or an array of data types this element can accept.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndReorder {
    /**
     * An object that describes reorder functionality for a selected column. See the Help documentation for more information.
     */
    "columns": string;
  }
  export interface FirstSelectedRow {
    /**
     * The data for first selected row.
     */
    "data": any;
    /**
     * The key of first selected row.
     */
    "key": any;
  }
  export interface ScrollPolicyOptions {
    /**
     * The number of data rows to fetch in each block.
     */
    "fetchSize": number;
    /**
     * The maximum number of rows to display before fetching more data rows will be stopped.
     */
    "maxCount": number;
    /**
     * The CSS selector string to an element used to determine the scroll position as well as the maximum scroll position. See the Help documentation for more information.
     */
    "scroller": string|null;
    /**
     * The bottom offset value (in pixels) used for the Table's external scroller. This value is used to specify the location where the Table's footer region becomes 'sticky' when the 'scroller' attribute is specified.
     */
    "scrollerOffsetBottom": number|null;
    /**
     * The end offset value (in pixels) used for the Table's external scroller. This value is used to specify the location where the Table's frozen 'end' columns (or frozen 'start' columns in RTL) become 'sticky' when the 'scroller' attribute is specified.
     */
    "scrollerOffsetEnd": number|null;
    /**
     * The start offset value (in pixels) used for the Table's external scroller. This value is used to specify the location where the Table's frozen 'start' columns (or frozen 'end' columns in RTL) become 'sticky' when the 'scroller' attribute is specified.
     */
    "scrollerOffsetStart": number|null;
    /**
     * The top offset value (in pixels) used for the Table's external scroller. This value is used to specify the location where the Table's header region becomes 'sticky' when the 'scroller' attribute is specified.
     */
    "scrollerOffsetTop": number|null;
  }
  export interface ScrollPosition {
    /**
     * The zero-based index of the cell at the origin of the table.
     */
    "columnIndex": number;
    /**
     * The key of the column. This corresponds to the identifier of the column specified in columns. If the column does not exists then the value is ignored.
     */
    "columnKey": any;
    /**
     * The horizontal offset in pixels relative to the column identified by columnKey/columnIndex.
     */
    "offsetX": number;
    /**
     * The vertical offset in pixels relative to the row identified by rowKey/rowIndex.
     */
    "offsetY": number;
    /**
     * The zero-based index of the cell at the origin of the table. If scrollPolicy is set to 'loadMoreOnScroll and the row index is greater than maxCount set in scrollPolicyOptions, then it will scroll and fetch until the end of the table is reached and there are no more rows to fetch.
     */
    "rowIndex": number;
    /**
     * The key of the row. If DataProvider is used for data and the key does not exists in the DataProvider, then the value is ignored. If DataProvider is not used then Table will fetch and scroll until the item is found or the end of the table is reached and there's no more items to fetch.
     */
    "rowKey": any;
    /**
     * The horizontal position in pixels.
     */
    "x": number;
    /**
     * The vertical position in pixels.
     */
    "y": number;
  }
  export interface Selected {
  }
  export interface Translations {
    /**
     * Provides properties to customize the accessible context for the column index.
     */
    "accessibleColumnContext": string;
    /**
     * Provides properties to customize the accessible context for the column footer index.
     */
    "accessibleColumnFooterContext": string;
    /**
     * Provides properties to customize the accessible context for the column header index.
     */
    "accessibleColumnHeaderContext": string;
    /**
     * Provides properties to customize the accessible context for the row index.
     */
    "accessibleRowContext": string;
    /**
     * Provides properties to customize the accesible text when a header is sorted ascending.
     */
    "accessibleSortAscending": string;
    /**
     * Provides properties to customize the accesible text when a header is sorted descending.
     */
    "accessibleSortDescending": string;
    /**
     * Provides properties to customize the accesible text when a header is able to be sorted.
     */
    "accessibleSortable": string;
    /**
     * Provides properties to customize the accessible context when an item has been selected.
     */
    "accessibleStateSelected": string;
    /**
     * Label for the bottom selection affordance on touch devices.
     */
    "labelAccSelectionAffordanceBottom": string;
    /**
     * Label for the top selection affordance on touch devices.
     */
    "labelAccSelectionAffordanceTop": string;
    /**
     * Provides properties to customize the context menu label for exiting non-contiguous selection.
     */
    "labelDisableNonContiguousSelection": string;
    /**
     * Edit row label.
     */
    "labelEditRow": string;
    /**
     * Provides properties to customize the context menu label for entering non-contiguous selection.
     */
    "labelEnableNonContiguousSelection": string;
    /**
     * Provides properties to customize the context menu label for column resize.
     */
    "labelResize": string;
    /**
     * Provides properties to customize the resize popup cancel button.
     */
    "labelResizePopupCancel": string;
    /**
     * Provides properties to customize the resize popup spinner label.
     */
    "labelResizePopupSpinner": string;
    /**
     * Provides properties to customize the resize popup submit button.
     */
    "labelResizePopupSubmit": string;
    /**
     * Select and Edit row label.
     */
    "labelSelectAndEditRow": string;
    /**
     * Select column label.
     */
    "labelSelectColum": string;
    /**
     * Select row label.
     */
    "labelSelectRow": string;
    /**
     * Context menu label for sort.
     */
    "labelSort": string;
    /**
     * Context menu label for sort ascending.
     */
    "labelSortAsc": string;
    /**
     * Context menu label for sort descending.
     */
    "labelSortDsc": string;
    /**
     * Column resize input width validation.
     */
    "msgColumnResizeWidthValidation": string;
    /**
     * Fetching data message.
     */
    "msgFetchingData": string;
    /**
     * Table initializing message.
     */
    "msgInitializing": string;
    /**
     * No data to display message.
     */
    "msgNoData": string;
    /**
     * Warning message detail that maxCount has been reached for scrollPolicy=loadMoreOnScroll.
     */
    "msgScrollPolicyMaxCountDetail": string;
    /**
     * Warning message summary that maxCount has been reached for scrollPolicy=loadMoreOnScroll.
     */
    "msgScrollPolicyMaxCountSummary": string;
    /**
     * Sort ascending status notification for accessibility.
     */
    "msgStatusSortAscending": string;
    /**
     * Sort descending status notification for accessibility.
     */
    "msgStatusSortDescending": string;
  }
}
