// @oracle/oraclejet-webdriver/elements/oj-table/OjTableBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTableBase = void 0;
const __1 = require("../");
class OjTableBase extends __1.OjWebElement {
    getAccessibility() {
        return this.getProperty("accessibility");
    }
    getAs() {
        return this.getProperty("as");
    }
    changeColumns(columns) {
        return this.setProperty("columns", columns);
    }
    getColumns() {
        return this.getProperty("columns");
    }
    getColumnsDefault() {
        return this.getProperty("columnsDefault");
    }
    changeCurrentRow(currentRow) {
        return this.setProperty("currentRow", currentRow);
    }
    getCurrentRow() {
        return this.getProperty("currentRow");
    }
    getData() {
        return this.getProperty("data");
    }
    getDisplay() {
        return this.getProperty("display");
    }
    getDnd() {
        return this.getProperty("dnd");
    }
    getEditMode() {
        return this.getProperty("editMode");
    }
    changeEditRow(editRow) {
        return this.setProperty("editRow", editRow);
    }
    getEditRow() {
        return this.getProperty("editRow");
    }
    getFirstSelectedRow() {
        return this.getProperty("firstSelectedRow");
    }
    getHorizontalGridVisible() {
        return this.getProperty("horizontalGridVisible");
    }
    getLayout() {
        return this.getProperty("layout");
    }
    getRowRenderer() {
        return this.getProperty("rowRenderer");
    }
    getScrollPolicy() {
        return this.getProperty("scrollPolicy");
    }
    getScrollPolicyOptions() {
        return this.getProperty("scrollPolicyOptions");
    }
    changeScrollPosition(scrollPosition) {
        return this.setProperty("scrollPosition", scrollPosition);
    }
    getScrollPosition() {
        return this.getProperty("scrollPosition");
    }
    getScrollToKey() {
        return this.getProperty("scrollToKey");
    }
    changeSelected(selected) {
        return this.setProperty("selected", selected);
    }
    getSelected() {
        return this.getProperty("selected");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSelectionRequired() {
        return this.getProperty("selectionRequired");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getVerticalGridVisible() {
        return this.getProperty("verticalGridVisible");
    }
}
exports.OjTableBase = OjTableBase;
