// @oracle/oraclejet-webdriver/elements/oj-refresher/OjRefresherBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjRefresher, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjRefresherBase extends OjWebElement {
  /**
   * Gets the value of refreshContent property.
   * Specifies the function to invoke when pull-to-refresh operation is triggered. See the Help documentation for more information.
   * @return The value of refreshContent property.
   * 
   */
  public getRefreshContent(): Promise<null> {
    return this.getProperty<null>("refreshContent");
  }

  /**
   * Gets the value of text property.
   * Specifies a text message to display in the pull-to-refresh panel after a pull down gesture.
   * @return The value of text property.
   * 
   */
  public getTextProperty(): Promise<string> {
    return this.getProperty<string>("text");
  }

  /**
   * Gets the value of threshold property.
   * The minimum distance in pixels that the user needs to pull down in order to trigger a refresh. See the Help documentation for more information.
   * @return The value of threshold property.
   * 
   */
  public getThreshold(): Promise<number> {
    return this.getProperty<number>("threshold");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjRefresherBase.Translations|null> {
    return this.getProperty<OjRefresherBase.Translations|null>("translations");
  }

}

namespace OjRefresherBase {
  export interface Translations {
    /**
     * Label assigned to the refresh message to denote refresh complete that appears after triggering refresh content.
     */
    "ariaRefreshCompleteLink": string;
    /**
     * Label assigned to the refresh link used for triggering refresh content when using keyboard or accessibility agents such as VoiceOver.
     */
    "ariaRefreshLink": string;
    /**
     * Label assigned to the refresh message to denote that the refresh is occurring that appears after triggering refresh content.
     */
    "ariaRefreshingLink": string;
  }
}
