// @oracle/oraclejet-webdriver/elements/oj-navigation-list/OjNavigationList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjNavigationList = void 0;
const selenium_webdriver_1 = require("selenium-webdriver");
const __1 = require("../../");
const OjNavigationListBase_1 = require("./OjNavigationListBase");
class OjNavigationList extends OjNavigationListBase_1.OjNavigationListBase {
    async changeExpanded(expanded) {
        await this.whenReady();
        await this.getDriver().executeScript([
            'var ele = arguments[0];',
            'require([\'ojs/ojkeyset\'], function(keySet) {',
            '  var expandedKeySet = new keySet.ExpandedKeySet(' + JSON.stringify(expanded) + ');',
            '  ele.expanded = expandedKeySet;',
            '});',
        ].join('\n'), this);
        await this.whenBusyContextReady();
    }
    async changeExpandedByIndex(indexes) {
        this.whenReady();
        let expansionTree = this._getExpansionTree(indexes);
        await this.getDriver().executeScript([
            'var ele = arguments[0]',
            'let expansionTree = arguments[1];',
            '_traverseListToExpand(null, expansionTree);',
            ' function _isCollapsed(item) {',
            '  return _hasClass(item,"oj-collapsed");',
            ' }',
            ' function _isExpanded(item) {',
            '  return _hasClass(item,"oj-expanded");',
            ' }',
            ' function  expandItem(item) {',
            '  let isExpanded = _isExpanded(item);',
            '  if (!isExpanded) {',
            '   var collapsedIcon = item.querySelector(".oj-navigationlist-collapse-icon");',
            '   collapsedIcon.click();',
            '  }',
            ' }',
            ' function collapseItem(item) {',
            '  let isCollapsed = _isCollapsed(item);',
            '  if (!isCollapsed) {',
            '   var expandedIcon = item.querySelector(".oj-navigationlist-expand-icon");',
            '   expandedIcon.click();',
            '  }',
            ' }',
            ' function _hasClass(element, className) {',
            ' let classes =  element.getAttribute("class");',
            ' let hasClass = false;',
            ' classes.split(" ").forEach(c => {',
            '  if (c === className) {',
            '   hasClass = true;',
            '  }',
            ' });',
            ' return hasClass;',
            ' }',
            ' function getItemByIndex(index, parent){',
            ' let item;',
            '  if (!parent)',
            '   item = ele.querySelector(".oj-navigationlist-element > .oj-navigationlist-item-element:nth-child(" + index + ")")',
            '  else {',
            '   item = parent.querySelector(".oj-navigationlist-group > .oj-navigationlist-item-element:nth-child(" + index + ")")',
            '  }',
            ' return item;',
            ' }',
            ' function  shouldBeExpanded(expansionTree, indexPath){',
            '  if (indexPath.length === 0) {',
            '   return false;',
            '  }',
            '  for (let i=0; i < indexPath.length; i++) {',
            '  expansionTree = expansionTree[indexPath[i]];',
            '   if (!expansionTree) {',
            '    return false;',
            '   }',
            '  }',
            ' return true;',
            ' }',
            ' function _traverseListToExpand(parent, expansionTree, parentIndexPath) {',
            '  let item, indexPath, index = 1;',
            '  item =  getItemByIndex(index, parent);',
            '  parentIndexPath = parentIndexPath && parentIndexPath.length > 0 ? parentIndexPath : []',
            '  while (item) {',
            '   let isExpanded =  _isExpanded(item);',
            '   let isCollapsed = _isCollapsed(item);',
            '     if (isCollapsed || isExpanded) {',
            '       indexPath = parentIndexPath.concat([index])',
            '       if (isCollapsed && shouldBeExpanded(expansionTree, indexPath)) {',
            '           expandItem(item);',
            '       }',
            '     _traverseListToExpand(item, expansionTree, indexPath);',
            '     if (isExpanded && !shouldBeExpanded(expansionTree, indexPath)) {',
            '         collapseItem(item);',
            '     }',
            '   }',
            '   index = index + 1;',
            '   item =  getItemByIndex(index, parent);',
            '  }',
            ' }',
        ].join('\n'), this, expansionTree);
    }
    _getExpansionTree(pathsToBeExpanded = []) {
        let children, expansionTree = {};
        for (let i = 0; i < pathsToBeExpanded.length; i++) {
            children = expansionTree;
            for (let j = 0; j < pathsToBeExpanded[i].length; j++) {
                let index = pathsToBeExpanded[i][j];
                if (!children[index]) {
                    children[index] = {};
                }
                children = children[index];
            }
        }
        return expansionTree;
    }
    async changeSelectionByIndex(index) {
        await this.whenReady();
        let indexArray, parent;
        if (!Array.isArray(index)) {
            indexArray = [index];
        }
        else {
            indexArray = index;
        }
        for (let i = 0; i < indexArray.length; i++) {
            parent = await this._getItemByIndex(indexArray[i], parent);
            if (i < indexArray.length - 1) {
                await this._expandItem(parent);
            }
            else {
                await parent.click();
            }
        }
    }
    async getExpanded() {
        await this.whenReady();
        let expanded = await this.getDriver().executeScript([
            'return Array.from(arguments[0].expanded.values());'
        ].join('\n'), this);
        return expanded;
    }
    async getItemByIndex(indexPath) {
        let item;
        for (let i = 0; i < indexPath.length; i++) {
            item = item ? await this._getItemByIndex(indexPath[i], item) : await this._getItemByIndex(indexPath[i]);
        }
        return item;
    }
    async _hasClass(element, className) {
        let classes = await element.getAttribute("class");
        let hasClass = false;
        classes.split(" ").forEach(c => {
            if (c === className) {
                hasClass = true;
            }
        });
        return hasClass;
    }
    async _getItemByIndex(index, parent) {
        let item;
        try {
            if (!parent)
                item = await this.findElement(selenium_webdriver_1.By.css('.oj-navigationlist-element > .oj-navigationlist-item-element:nth-child(' + index + ')'));
            else {
                item = await parent.findElement(selenium_webdriver_1.By.css('.oj-navigationlist-group > .oj-navigationlist-item-element:nth-child(' + index + ')'));
            }
        }
        catch (e) {
            item = null;
        }
        return item;
    }
    async _isCollapsed(item) {
        return await this._hasClass(item, 'oj-collapsed');
    }
    async _isExpanded(item) {
        return await this._hasClass(item, 'oj-expanded');
    }
    async _expandItem(item) {
        let isExpanded = await this._isExpanded(item);
        if (!isExpanded) {
            var collapsedIcon = await item.findElement(selenium_webdriver_1.By.css('.oj-navigationlist-collapse-icon'));
            await collapsedIcon.click();
            await this.whenBusyContextReady();
        }
    }
    async findItem(itemLocator) {
        const el = await this.getDriver().executeScript(`
    return arguments[0].getNodeBySubId({ subId: 'oj-navigationlist-item', key: ${JSON.stringify(itemLocator.key)} })
    `, this);
        return __1.slotProxy(el, this);
    }
}
exports.OjNavigationList = OjNavigationList;
