// @oracle/oraclejet-webdriver/elements/oj-navigation-list/OjNavigationListBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjNavigationListBase = void 0;
const __1 = require("../");
class OjNavigationListBase extends __1.OjWebElement {
    getAs() {
        return this.getProperty("as");
    }
    changeCurrentItem(currentItem) {
        return this.setProperty("currentItem", currentItem);
    }
    getCurrentItem() {
        return this.getProperty("currentItem");
    }
    getData() {
        return this.getProperty("data");
    }
    getDisplay() {
        return this.getProperty("display");
    }
    getDrillMode() {
        return this.getProperty("drillMode");
    }
    getEdge() {
        return this.getProperty("edge");
    }
    getHierarchyMenuThreshold() {
        return this.getProperty("hierarchyMenuThreshold");
    }
    getItem() {
        return this.getProperty("item");
    }
    getOverflow() {
        return this.getProperty("overflow");
    }
    getRootLabel() {
        return this.getProperty("rootLabel");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjNavigationListBase = OjNavigationListBase;
