// @oracle/oraclejet-webdriver/elements/oj-navigation-list/OjNavigationListBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjNavigationList, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjNavigationListBase extends OjWebElement {
  /**
   * Gets the value of as property.
   * Specifies the alias for the current item when referenced inside the item template.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Sets the value of "currentItem" property.
   * Specifies the key of the item that should have keyboard focus. See the Help documentation for more information.
   * @param currentItem The value to set for "currentItem"
   * 
   */
  public changeCurrentItem(currentItem: any): Promise<void> {
    return this.setProperty<any>("currentItem", currentItem);
  }

  /**
   * Gets the value of currentItem property.
   * Specifies the key of the item that should have keyboard focus. See the Help documentation for more information.
   * @return The value of currentItem property.
   * 
   */
  public getCurrentItem(): Promise<any> {
    return this.getProperty<any>("currentItem");
  }

  /**
   * Gets the value of data property.
   * Specifies the data provider for the Navigation List. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of display property.
   * Specifies what needs to be displayed. See the Help documentation for more information.
   * @return The value of display property.
   * 
   */
  public getDisplay(): Promise<string> {
    return this.getProperty<string>("display");
  }

  /**
   * Gets the value of drillMode property.
   * Specifies whether expand/collapse or sliding operations are allowed.
   * @return The value of drillMode property.
   * 
   */
  public getDrillMode(): Promise<string> {
    return this.getProperty<string>("drillMode");
  }

  /**
   * Gets the value of edge property.
   * Specifies the edge position of the Navigation List.
   * @return The value of edge property.
   * 
   */
  public getEdge(): Promise<string> {
    return this.getProperty<string>("edge");
  }

  /**
   * Gets the value of hierarchyMenuThreshold property.
   * Specifies the level at which the user can see hierarchical menu button. See the Help documentation for more information.
   * @return The value of hierarchyMenuThreshold property.
   * 
   */
  public getHierarchyMenuThreshold(): Promise<number> {
    return this.getProperty<number>("hierarchyMenuThreshold");
  }

  /**
   * Gets the value of item property.
   * Customize the functionality of each item in Navigation List.
   * @return The value of item property.
   * 
   */
  public getItem(): Promise<OjNavigationListBase.Item> {
    return this.getProperty<OjNavigationListBase.Item>("item");
  }

  /**
   * Gets the value of overflow property.
   * Specifies overflow behaviour for the Navigation List.
   * @return The value of overflow property.
   * 
   */
  public getOverflow(): Promise<string> {
    return this.getProperty<string>("overflow");
  }

  /**
   * Gets the value of rootLabel property.
   * Specifies the label for top level list items.
   * @return The value of rootLabel property.
   * 
   */
  public getRootLabel(): Promise<string> {
    return this.getProperty<string>("rootLabel");
  }

  /**
   * Sets the value of "selection" property.
   * Specifies the key of the selected item. See the Help documentation for more information.
   * @param selection The value to set for "selection"
   * 
   */
  public changeSelection(selection: any): Promise<void> {
    return this.setProperty<any>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * Specifies the key of the selected item. See the Help documentation for more information.
   * @return The value of selection property.
   * 
   */
  public getSelection(): Promise<any> {
    return this.getProperty<any>("selection");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjNavigationListBase.Translations|null> {
    return this.getProperty<OjNavigationListBase.Translations|null>("translations");
  }

}

namespace OjNavigationListBase {
  export interface Item {
    /**
     * Specifies the renderer for the item. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies whether the item can be selected. See the Help documentation for more information.
     */
    "selectable": null|boolean;
  }
  export interface Translations {
    /**
     * Provides properties to customize the title text used by Sliding Navigation List for root node.
     */
    "defaultRootLabel": string;
    /**
     * Provides properties to customize the hover message text used by Sliding Navigation List as label for Hierarchical Menu button.
     */
    "hierMenuBtnLabel": string;
    /**
     * Provides properties to customize the message text used by Navigation List when waiting for data.
     */
    "msgFetchingData": string;
    /**
     * Provides properties to customize the message text used by Navigation List when there are no items.
     */
    "msgNoData": string;
    /**
     * Label for overflow menu button.
     */
    "overflowItemLabel": string;
    /**
     * Provides properties to customize the hover message text used by Sliding Navigation List for previous button.
     */
    "previousIcon": string;
    /**
     * Provides text to read to screen reader when an item is selected.
     */
    "selectedLabel": string;
  }
}
