// @oracle/oraclejet-webdriver/elements/OjWebElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ojWebElement = exports.OjWebElement = void 0;
const selenium_webdriver_1 = require("selenium-webdriver");
const __1 = require("../");
const stale_element_1 = require("../lib/stale-element");
class OjWebElement extends selenium_webdriver_1.WebElement {
    constructor(el, locators) {
        super(el.getDriver(), el.getId());
        this._locators = locators;
    }
    async clear() {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_clear");
        await this.whenReady();
        await OjWebElement.trapStaleElement(this, super.clear, undefined);
        return this.getDriver().executeScript("arguments[0].focus()", this);
    }
    async click() {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_click");
        await this.whenReady();
        await this.getDriver().executeScript('arguments[0].focus()', this);
        return OjWebElement.trapStaleElement(this, super.click, undefined);
    }
    async getAttribute(attributeName) {
        await this.whenBusyContextReady();
        return OjWebElement.trapStaleElement(this, super.getAttribute, attributeName);
    }
    async getCssValue(cssStyleProperty) {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_getCssValue");
        await this.whenBusyContextReady();
        return OjWebElement.trapStaleElement(this, super.getCssValue, cssStyleProperty);
    }
    async getLocation() {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_getLocation");
        await this.whenBusyContextReady();
        return OjWebElement.trapStaleElement(this, super.getLocation, undefined);
    }
    async getRect() {
        if (super.getRect) {
            const screenshotManager = __1.ScreenshotManager.get();
            await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_getRect");
            await this.whenBusyContextReady();
            return OjWebElement.trapStaleElement(this, super.getRect, undefined);
        }
        throw new selenium_webdriver_1.error.UnknownMethodError('WebElement.getRect() is not supported for this version of selenium-webdriver');
    }
    async getSize() {
        if (super.getSize) {
            const screenshotManager = __1.ScreenshotManager.get();
            await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_getSize");
            await this.whenBusyContextReady();
            return OjWebElement.trapStaleElement(this, super.getSize, undefined);
        }
        else {
            return this.getRect();
        }
    }
    async getText() {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_getText");
        await this.whenBusyContextReady();
        return OjWebElement.trapStaleElement(this, super.getText, undefined);
    }
    async isDisplayed() {
        await this.whenBusyContextReady();
        return OjWebElement.trapStaleElement(this, super.isDisplayed, undefined);
    }
    async isEnabled() {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_isEnabled");
        await this.whenBusyContextReady();
        let disabled = await this.getProperty("disabled");
        if (typeof disabled === "boolean") {
            return !disabled;
        }
        return await OjWebElement.trapStaleElement(this, super.isEnabled, undefined);
    }
    async isSelected() {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_isSelected");
        await this.whenBusyContextReady();
        return OjWebElement.trapStaleElement(this, super.isSelected, undefined);
    }
    async sendKeys(...varArgs) {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_sendKeys");
        await this.whenReady();
        return OjWebElement.trapStaleElement(this, super.sendKeys, ...varArgs);
    }
    async submit() {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_submit");
        await this.whenReady();
        return OjWebElement.trapStaleElement(this, super.submit, undefined);
    }
    findElement(locator) {
        const p = this.whenReady().then(() => OjWebElement.trapStaleElement(this, super.findElement, locator));
        return new selenium_webdriver_1.WebElementPromise(this.getDriver(), p);
    }
    async findElements(locator) {
        const screenshotManager = __1.ScreenshotManager.get();
        await screenshotManager.takeAndSaveScreenShot(this.getDriver(), "before_findElements");
        await this.whenReady();
        const el = OjWebElement.trapStaleElement(this, super.findElements, locator);
        return el;
    }
    async takeScreenshot(optScroll) {
        const ss = OjWebElement.trapStaleElement(this, super.takeScreenshot, optScroll);
        return ss;
    }
    async whenReady() {
        const driver = this.getDriver();
        const timeout = 15000;
        const tag = `<${this._locators.tagName}>`;
        await this.whenBusyContextReady();
        this._skipBusyContext = true;
        try {
            await driver.wait(() => this.isDisplayed(), timeout, `waiting for ${tag} to be visible`);
            return driver.wait(async () => {
                const size = await this.getSize();
                return size.height > 0 && size.width > 0;
            }, timeout, `waiting for ${tag} to have non-zero size`);
        }
        finally {
            this._skipBusyContext = false;
        }
    }
    async whenBusyContextReady() {
        if (this._skipBusyContext) {
            return;
        }
        const remoteScript = `
      const done = arguments[0];
      if (typeof require === 'function') {
        require('ojs/ojcontext').getPageContext().getBusyContext()
        .whenReady().then(() => done())
        .catch(ex => done(ex.message))
      } else {
        done('RequireJS not found')
      }
      `;
        const ret = await this.getDriver()
            .executeAsyncScript(remoteScript)
            .catch(async (ex) => {
            const openStates = await this.getDriver().executeScript(`
        return require('ojs/ojcontext')
          .getPageContext().getBusyContext().getBusyStates()
          .map((state) => state.description);
        `).catch(ex => ex.message);
            console.warn(`${ex.message}\nopen busy states: ${openStates}`);
            throw Error(`BusyContext.whenReady() timed out: ${ex.message}`);
        });
        if (ret) {
            console.warn(`Cannot wait on BusyContext: ${ret}`);
        }
    }
    async getProperty(propertyName) {
        await this.whenBusyContextReady();
        return OjWebElement.trapStaleElement(this, this._getProperty, propertyName);
    }
    async setProperty(propertyName, value) {
        await this.whenReady();
        await OjWebElement.trapStaleElement(this, this._setProperty, propertyName, value);
    }
    hasSize() {
        return new selenium_webdriver_1.Condition(`<${this._locators.tagName}> has a non-zero size`, async (driver) => {
            const rect = await this.getSize();
            return rect.height > 0 && rect.width > 0;
        });
    }
    hasVisibility() {
        return new selenium_webdriver_1.Condition(`<${this._locators.tagName}> is displayed and scrolled into view`, async (driver) => {
            const displayed = await this.isDisplayed();
            if (displayed) {
                await driver.executeScript("arguments[0].scrollIntoView()", this);
                return true;
            }
            return displayed;
        });
    }
    async _getProperty(propertyName) {
        if (!OjWebElement._BROWSER_NAME) {
            const caps = await this.getDriver().getCapabilities();
            OjWebElement._BROWSER_NAME = caps.get(selenium_webdriver_1.Capability.BROWSER_NAME);
        }
        if (OjWebElement._BROWSER_NAME === selenium_webdriver_1.Browser.FIREFOX) {
            return this._getPropertyFirefox(propertyName);
        }
        return this.getDriver().executeAsyncScript(`let el = arguments[0], prop = arguments[1], done = arguments[2];
      require(['ojs/ojcustomelement-utils'], function(utils) {
        done(utils.CustomElementUtils.getElementProperty(el, prop));
      }, done);
      `, this, propertyName);
    }
    async _getPropertyFirefox(propertyName) {
        const driver = this.getDriver();
        await driver.executeAsyncScript(`require(['ojs/ojcustomelement-utils'], arguments[0])`);
        return driver.executeScript(`let el = arguments[0], prop = arguments[1];
      return require('ojs/ojcustomelement-utils').CustomElementUtils.getElementProperty(el, prop);`, this, propertyName);
    }
    async _setProperty(propertyName, value) {
        await this.getDriver().executeScript(`return arguments[0].setProperty ?
        arguments[0].setProperty(arguments[1], arguments[2])
        : arguments[0][arguments[1]] = arguments[2]`, this, propertyName, value);
    }
    static async trapStaleElement(el, cmd, ...cmdParams) {
        let result;
        try {
            result = await cmd.apply(el, cmdParams);
        }
        catch (ex) {
            if (ex instanceof selenium_webdriver_1.error.StaleElementReferenceError && el._locators.elLocator) {
                console.warn(`Trapped stale element calling '${cmd.name}' on ${el._locators.elLocator}`);
                const newEl = await stale_element_1.requeryElement(el);
                result = await cmd.apply(newEl, cmdParams);
            }
            else {
                throw ex;
            }
        }
        return result;
    }
}
exports.OjWebElement = OjWebElement;
async function ojWebElement(driver, by) {
    const el = await driver.findElement(by);
    if (!(el instanceof OjWebElement)) {
        return new OjWebElement(el, null);
    }
    return el;
}
exports.ojWebElement = ojWebElement;
