// @oracle/oraclejet-webdriver/elements/oj-spark-chart/OjSparkChartBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjSparkChartBase = void 0;
const __1 = require("../");
class OjSparkChartBase extends __1.OjWebElement {
    getAnimationDuration() {
        return this.getProperty("animationDuration");
    }
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAreaColor() {
        return this.getProperty("areaColor");
    }
    getAreaSvgClassName() {
        return this.getProperty("areaSvgClassName");
    }
    getAreaSvgStyle() {
        return this.getProperty("areaSvgStyle");
    }
    getAs() {
        return this.getProperty("as");
    }
    getBarGapRatio() {
        return this.getProperty("barGapRatio");
    }
    getBaselineScaling() {
        return this.getProperty("baselineScaling");
    }
    getColor() {
        return this.getProperty("color");
    }
    getData() {
        return this.getProperty("data");
    }
    getFirstColor() {
        return this.getProperty("firstColor");
    }
    getHighColor() {
        return this.getProperty("highColor");
    }
    getItems() {
        return this.getProperty("items");
    }
    getLastColor() {
        return this.getProperty("lastColor");
    }
    getLineStyle() {
        return this.getProperty("lineStyle");
    }
    getLineType() {
        return this.getProperty("lineType");
    }
    getLineWidth() {
        return this.getProperty("lineWidth");
    }
    getLowColor() {
        return this.getProperty("lowColor");
    }
    getMarkerShape() {
        return this.getProperty("markerShape");
    }
    getMarkerSize() {
        return this.getProperty("markerSize");
    }
    getReferenceObjects() {
        return this.getProperty("referenceObjects");
    }
    getSvgClassName() {
        return this.getProperty("svgClassName");
    }
    getSvgStyle() {
        return this.getProperty("svgStyle");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getType() {
        return this.getProperty("type");
    }
    getVisualEffects() {
        return this.getProperty("visualEffects");
    }
}
exports.OjSparkChartBase = OjSparkChartBase;
