// @oracle/oraclejet-webdriver/elements/oj-spark-chart/OjSparkChartBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjSparkChart, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjSparkChartBase extends OjWebElement {
  /**
   * Gets the value of animationDuration property.
   * The duration of the animations in milliseconds.
   * @return The value of animationDuration property.
   * 
   */
  public getAnimationDuration(): Promise<number> {
    return this.getProperty<number>("animationDuration");
  }

  /**
   * Gets the value of animationOnDataChange property.
   * Defines the animation that is applied on data changes.
   * @return The value of animationOnDataChange property.
   * 
   */
  public getAnimationOnDataChange(): Promise<string> {
    return this.getProperty<string>("animationOnDataChange");
  }

  /**
   * Gets the value of animationOnDisplay property.
   * Defines the animation that is shown on initial display.
   * @return The value of animationOnDisplay property.
   * 
   */
  public getAnimationOnDisplay(): Promise<string> {
    return this.getProperty<string>("animationOnDisplay");
  }

  /**
   * Gets the value of areaColor property.
   * The color of the area. Only applies if type is "area" or "lineWithArea".
   * @return The value of areaColor property.
   * 
   */
  public getAreaColor(): Promise<string> {
    return this.getProperty<string>("areaColor");
  }

  /**
   * Gets the value of areaSvgClassName property.
   * The CSS style class to apply if type is "area" or "lineWithArea".
   * @return The value of areaSvgClassName property.
   * 
   */
  public getAreaSvgClassName(): Promise<string> {
    return this.getProperty<string>("areaSvgClassName");
  }

  /**
   * Gets the value of areaSvgStyle property.
   * The inline style to apply if type is "area" or "lineWithArea".
   * @return The value of areaSvgStyle property.
   * 
   */
  public getAreaSvgStyle(): Promise<object> {
    return this.getProperty<object>("areaSvgStyle");
  }

  /**
   * Gets the value of as property.
   * An alias for the '$current' context variable passed to slot content for the itemTemplate slot.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Gets the value of barGapRatio property.
   * Specifies the width of the bar gap as a ratio of the item width. The valid value is a number from 0 to 1.
   * @return The value of barGapRatio property.
   * 
   */
  public getBarGapRatio(): Promise<number> {
    return this.getProperty<number>("barGapRatio");
  }

  /**
   * Gets the value of baselineScaling property.
   * Defines whether the axis baseline starts at the minimum value of the data or at zero.
   * @return The value of baselineScaling property.
   * 
   */
  public getBaselineScaling(): Promise<string> {
    return this.getProperty<string>("baselineScaling");
  }

  /**
   * Gets the value of color property.
   * The color of the data items.
   * @return The value of color property.
   * 
   */
  public getColor(): Promise<string> {
    return this.getProperty<string>("color");
  }

  /**
   * Gets the value of data property.
   * Specifies the data for the spark chart. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of firstColor property.
   * The color of the first data item.
   * @return The value of firstColor property.
   * 
   */
  public getFirstColor(): Promise<string> {
    return this.getProperty<string>("firstColor");
  }

  /**
   * Gets the value of highColor property.
   * The color of the data item with the greatest value.
   * @return The value of highColor property.
   * 
   */
  public getHighColor(): Promise<string> {
    return this.getProperty<string>("highColor");
  }

  /**
   * Gets the value of items property.
   * An array of objects that define the data for the spark chart. See the Help documentation for more information.
   * @return The value of items property.
   * 
   */
  public getItems(): Promise<Array<OjSparkChartBase.Items>|Array<number>|null|null> {
    return this.getProperty<Array<OjSparkChartBase.Items>|Array<number>|null|null>("items");
  }

  /**
   * Gets the value of lastColor property.
   * The color of the last data item.
   * @return The value of lastColor property.
   * 
   */
  public getLastColor(): Promise<string> {
    return this.getProperty<string>("lastColor");
  }

  /**
   * Gets the value of lineStyle property.
   * The line style of the data line. Only applies to line spark charts.
   * @return The value of lineStyle property.
   * 
   */
  public getLineStyle(): Promise<string> {
    return this.getProperty<string>("lineStyle");
  }

  /**
   * Gets the value of lineType property.
   * The line type of the data line or area. Only applies to line and area spark charts.
   * @return The value of lineType property.
   * 
   */
  public getLineType(): Promise<string> {
    return this.getProperty<string>("lineType");
  }

  /**
   * Gets the value of lineWidth property.
   * The width of the data line in pixels. Only applies to line spark charts.
   * @return The value of lineWidth property.
   * 
   */
  public getLineWidth(): Promise<number> {
    return this.getProperty<number>("lineWidth");
  }

  /**
   * Gets the value of lowColor property.
   * The color of the data item with the lowest value.
   * @return The value of lowColor property.
   * 
   */
  public getLowColor(): Promise<string> {
    return this.getProperty<string>("lowColor");
  }

  /**
   * Gets the value of markerShape property.
   * The shape of the data markers. See the Help documentation for more information.
   * @return The value of markerShape property.
   * 
   */
  public getMarkerShape(): Promise<string> {
    return this.getProperty<string>("markerShape");
  }

  /**
   * Gets the value of markerSize property.
   * The size of the data markers in pixels. Only applies to line and area spark charts.
   * @return The value of markerSize property.
   * 
   */
  public getMarkerSize(): Promise<number> {
    return this.getProperty<number>("markerSize");
  }

  /**
   * Gets the value of referenceObjects property.
   * An array of reference objects associated with the y axis of the spark chart.
   * @return The value of referenceObjects property.
   * 
   */
  public getReferenceObjects(): Promise<Array<OjSparkChartBase.ReferenceObjects>> {
    return this.getProperty<Array<OjSparkChartBase.ReferenceObjects>>("referenceObjects");
  }

  /**
   * Gets the value of svgClassName property.
   * The CSS style class to apply to the data items. If type is "lineWithArea", this style will only be applied to the line if areaSvgClassName is also specified. See the Help documentation for more information.
   * @return The value of svgClassName property.
   * 
   */
  public getSvgClassName(): Promise<string> {
    return this.getProperty<string>("svgClassName");
  }

  /**
   * Gets the value of svgStyle property.
   * The inline style to apply to the data items. If type is "lineWithArea", this style will only be applied to the line if areaSvgStyle is also specified. See the Help documentation for more information.
   * @return The value of svgStyle property.
   * 
   */
  public getSvgStyle(): Promise<object> {
    return this.getProperty<object>("svgStyle");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjSparkChartBase.Tooltip> {
    return this.getProperty<OjSparkChartBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjSparkChartBase.Translations|null> {
    return this.getProperty<OjSparkChartBase.Translations|null>("translations");
  }

  /**
   * Gets the value of type property.
   * The chart type.
   * @return The value of type property.
   * 
   */
  public getType(): Promise<string> {
    return this.getProperty<string>("type");
  }

  /**
   * Gets the value of visualEffects property.
   * Defines whether visual effects such as overlays are applied to the spark chart.
   * @return The value of visualEffects property.
   * 
   */
  public getVisualEffects(): Promise<string> {
    return this.getProperty<string>("visualEffects");
  }

}

namespace OjSparkChartBase {
  export interface Items {
    /**
     * The default border color for the data items.
     */
    "borderColor": string;
    /**
     * The color of the bar or marker for the data item. This override can be used to highlight important values or thresholds.
     */
    "color": string;
    /**
     * The high value for range bar/area. Define 'low' and 'high' instead of 'value' to create a range bar/area spark chart.
     */
    "high": number;
    /**
     * The low value for range bar/area. Define 'low' and 'high' instead of 'value' to create a range bar/area spark chart.
     */
    "low": number;
    /**
     * Defines whether a marker should be displayed for the data item. Only applies to line and area spark charts.
     */
    "markerDisplayed": string;
    /**
     * The shape of the data markers. Can take the name of a built-in shape or the SVG path commands for a custom shape. Only applies to line and area spark charts.
     */
    "markerShape": string;
    /**
     * The size of the data markers in pixels. Only applies to line and area spark charts.
     */
    "markerSize": number;
    /**
     * The CSS style class to apply to the data item. The style class and inline style will override any other styling specified through the properties. For tooltips and hover interactivity, it's recommended to also pass a representative color to the item color attribute.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the data item. The style class and inline style will override any other styling specified through the properties. For tooltips and hover interactivity, it's recommended to also pass a representative color to the item color attribute. Only SVG CSS style properties are supported.
     */
    "svgStyle": object;
    /**
     * The value of the data item.
     */
    "value": number;
  }
  export interface ReferenceObjects {
    /**
     * The color of the reference object.
     */
    "color": string;
    /**
     * The high value of a reference area.
     */
    "high": number;
    /**
     * The line style of a reference line.
     */
    "lineStyle": string;
    /**
     * The width of a reference line.
     */
    "lineWidth": number;
    /**
     * The location of the reference object relative to the data items.
     */
    "location": string;
    /**
     * The low value of a reference area.
     */
    "low": number;
    /**
     * The CSS style class to apply to the reference object. The style class and inline style will override any other styling specified through the properties.
     */
    "svgClassName": string;
    /**
     * The inline style to apply to the reference object. The style class and inline style will override any other styling specified through the properties. Only SVG CSS style properties are supported.
     */
    "svgStyle": object;
    /**
     * The type of reference object being shown.
     */
    "type": string;
    /**
     * The value of a reference line.
     */
    "value": number;
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
  }
}
