// @oracle/oraclejet-webdriver/elements/oj-stream-list/OjStreamListBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjStreamList, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjStreamListBase extends OjWebElement {
  /**
   * Gets the value of data property.
   * Specifies the data for the component. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Sets the value of "expanded" property.
   * Specifies the key set containing the keys of the items that should be expanded.
   * @param expanded The value to set for "expanded"
   * 
   */
  public changeExpanded(expanded: any): Promise<void> {
    return this.setProperty<any>("expanded", expanded);
  }

  /**
   * Gets the value of expanded property.
   * Specifies the key set containing the keys of the items that should be expanded.
   * @return The value of expanded property.
   * 
   */
  public getExpanded(): Promise<any> {
    return this.getProperty<any>("expanded");
  }

  /**
   * Gets the value of scrollPolicy property.
   * Specifies how data are fetched as user scrolls towards the bottom of the list.
   * @return The value of scrollPolicy property.
   * 
   */
  public getScrollPolicy(): Promise<string> {
    return this.getProperty<string>("scrollPolicy");
  }

  /**
   * Gets the value of scrollPolicyOptions property.
   * Specifies fetch options for scrolling behaviors that trigger data fetches. See the Help documentation for more information.
   * @return The value of scrollPolicyOptions property.
   * 
   */
  public getScrollPolicyOptions(): Promise<OjStreamListBase.ScrollPolicyOptions> {
    return this.getProperty<OjStreamListBase.ScrollPolicyOptions>("scrollPolicyOptions");
  }

  /**
   * Sets the value of "scrollPosition" property.
   * Specifies the current scroll position of the StreamList. See the Help documentation for more information.
   * @param scrollPosition The value to set for "scrollPosition"
   * 
   */
  public changeScrollPosition(scrollPosition: OjStreamListBase.ScrollPosition): Promise<void> {
    return this.setProperty<OjStreamListBase.ScrollPosition>("scrollPosition", scrollPosition);
  }

  /**
   * Gets the value of scrollPosition property.
   * Specifies the current scroll position of the StreamList. See the Help documentation for more information.
   * @return The value of scrollPosition property.
   * 
   */
  public getScrollPosition(): Promise<OjStreamListBase.ScrollPosition> {
    return this.getProperty<OjStreamListBase.ScrollPosition>("scrollPosition");
  }

}

namespace OjStreamListBase {
  export interface ScrollPolicyOptions {
    /**
     * The number of items to fetch in each block.
     */
    "fetchSize": number;
    /**
     * The maximum total number of items to fetch.
     */
    "maxCount": number;
    /**
     * The element or a CSS selector string to an element used to determine the scroll position as well as the maximum scroll position. See the Help documentation for more information.
     */
    "scroller": string|null;
  }
  export interface ScrollPosition {
    /**
     * The vertical position in pixels.
     */
    "y": number;
    /**
     * The key of the item. If DataProvider is used for data and the key does not exist in the DataProvider or if the item has not been fetched yet, then the value is ignored.
     */
    "key": any;
    /**
     * The vertical offset in pixels relative to the item identified by key.
     */
    "offsetY": number;
    /**
     * The key of the parent if tree data.
     */
    "parentKey": any;
  }
}
