// @oracle/oraclejet-webdriver/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.slotProxy = exports.Expectation = exports.ScreenshotManager = exports.DriverManager = exports.register = void 0;
const selenium_webdriver_1 = require("selenium-webdriver");
var driver_override_1 = require("./lib/driver-override");
Object.defineProperty(exports, "register", { enumerable: true, get: function () { return driver_override_1.register; } });
var driver_manager_1 = require("./lib/driver-manager");
Object.defineProperty(exports, "DriverManager", { enumerable: true, get: function () { return driver_manager_1.DriverManager; } });
const OjFetch_1 = require("./lib/OjFetch");
const sm = require("./lib/ScreenshotManager");
exports.ScreenshotManager = sm;
const Expectation_1 = require("./lib/Expectation");
Object.defineProperty(exports, "Expectation", { enumerable: true, get: function () { return Expectation_1.Expectation; } });
var slot_proxy_1 = require("./lib/slot-proxy");
Object.defineProperty(exports, "slotProxy", { enumerable: true, get: function () { return slot_proxy_1.slotProxy; } });
function pageReady() {
    return new selenium_webdriver_1.Condition("for JET page to indicate ready", async (driver) => {
        const size = await driver.executeScript(`
      const html = document.querySelector('html');
      return { width: html.clientWidth, height: html.clientHeight }
    `);
        if (!(size.width && size.height)) {
            return Promise.resolve(false);
        }
        const remoteScript = `
    const done = arguments[0];
    try {
      require('ojs/ojcontext').getPageContext().getBusyContext()
      .whenReady().then(done)
      .catch(ex => done(false))
    } catch (ex) {
      done(false);
    }
    `;
        return driver.executeAsyncScript(remoteScript);
    });
}
const ASSERT_SCRIPT = `
var viewmodelEl = arguments[0];
var assertion = arguments[1];
var callback = arguments[arguments.length - 1];
require([
  'chai',
  'knockout',
  'ojs/ojcspexpressionevaluator'
], function(chai, ko, ExpressionEval) {
  if (chai) {
    try {
      var viewmodel = ko.dataFor(viewmodelEl);
      var evaluator = new ExpressionEval();
      // Use ExpressionEval to get the value from the viewmodel
      var value = evaluator.createEvaluator(assertion.expression).evaluate([viewmodel]);

      function compare(value) {
        var expect = chai.expect(value);
        // For each chained assertion, call the assertion method with the arguments
        assertion.chain.forEach((desc) => {
          if (typeof desc === 'string') {
            // Simple assertions, "is", "to", "be", etc
            expect = expect[desc];
          } else {
            // Complex assertions (chained functions)
            const name = desc.name;
            const args = desc.args;
            let evaluatedArgs = [];

            if (name === 'match') {
              args[0] = new RegExp(args[0]);
            }

            if (args) {
              evaluatedArgs = args;
            }

            expect = expect[name](...evaluatedArgs);
          }
        });
        callback(true);
      }

      // If "value" is a Promise
      if (value && typeof value.then === 'function') {
        value.then(compare).catch(function(ex) { callback(ex.message) });
      } else {
        compare(value);
      }
    } catch (ex) {
      callback(ex.message);
    }
  } else {
    callback('Chai not found');
  }
}, function(ex) { callback(ex.message); });
`;
async function checkVersions(driver) {
    const pjson = require('./package.json');
    const version = pjson.version;
    const rev = pjson.config.revision;
    const jet = await driver.executeScript(`
    return { version: oj.version, revision: oj.revision }
  `);
    const localRegex = /^#@@oj\./;
    if (!(version === '1.0.0' && localRegex.test(rev)) &&
        (version !== jet.version || rev !== jet.revision)) {
        const mismatch = `version mismatch: oraclejet ${JSON.stringify(jet)}, oraclejet-webdriver {"version": "${version}", "revision": "${rev}"}`;
        console.warn(`
Your oraclejet UI version/revision does not match oraclejet-webdriver
${mismatch}
    `);
        throw Error(mismatch);
    }
}
const ojwd = {
    async get(driver, url) {
        await driver.get(url);
        const timeouts = await driver.manage().getTimeouts();
        try {
            await driver.wait(pageReady(), timeouts.pageLoad);
            await checkVersions(driver);
        }
        catch (ex) {
            if (ex instanceof selenium_webdriver_1.error.TimeoutError || ex instanceof selenium_webdriver_1.error.ScriptTimeoutError) {
                const openStates = await driver.executeAsyncScript(`
        let done = arguments[0];
        require(['ojs/ojcontext'], function(Context) {
          done(Context.getPageContext().getBusyContext().getBusyStates()
            .map((state) => state.description));
        }, done);
        `).catch(ex => ex.message);
                ex.message += ` - open busy states: ${openStates}`;
            }
            throw ex;
        }
    },
    pageReady() {
        console.warn("pageReady() is deprecated and will be removed in the next release. " +
            "Use ojwd.get() to open and wait for the page to be ready for testing");
        return pageReady();
    },
    async fetch(driver, requestInfo, responseInfo, init) {
        let ojFetch = new OjFetch_1.OjFetch();
        return ojFetch.fetch(driver, requestInfo, responseInfo, init);
    },
    async waitAndFindElement(driver, by) {
        return driver.findElement(by);
    },
    async assertViewModelValue(driver, vmLocator, expectation) {
        const viewmodelEl = await driver.findElement(vmLocator);
        await viewmodelEl.whenBusyContextReady();
        const result = await driver.executeAsyncScript(ASSERT_SCRIPT, viewmodelEl, expectation.assertion);
        if (result !== true) {
            throw Error(String(result));
        }
        return result;
    }
};
exports.default = ojwd;
