// @oracle/oraclejet-webdriver/elements/oj-picto-chart/OjPictoChartBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjPictoChartBase = void 0;
const __1 = require("../");
class OjPictoChartBase extends __1.OjWebElement {
    getAnimationDuration() {
        return this.getProperty("animationDuration");
    }
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAs() {
        return this.getProperty("as");
    }
    getColumnCount() {
        return this.getProperty("columnCount");
    }
    getColumnWidth() {
        return this.getProperty("columnWidth");
    }
    getData() {
        return this.getProperty("data");
    }
    getDrilling() {
        return this.getProperty("drilling");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getHoverBehaviorDelay() {
        return this.getProperty("hoverBehaviorDelay");
    }
    getItems() {
        return this.getProperty("items");
    }
    getLayout() {
        return this.getProperty("layout");
    }
    getLayoutOrigin() {
        return this.getProperty("layoutOrigin");
    }
    getRowCount() {
        return this.getProperty("rowCount");
    }
    getRowHeight() {
        return this.getProperty("rowHeight");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjPictoChartBase = OjPictoChartBase;
