// @oracle/oraclejet-webdriver/elements/oj-selector/OjSelectorBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjSelector, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjSelectorBase extends OjWebElement {
  /**
   * Gets the value of rowKey property.
   * Specifies the row key of each selector. If the selectionMode property is 'all', rowKey is ignored.
   * @return The value of rowKey property.
   * 
   */
  public getRowKey(): Promise<any> {
    return this.getProperty<any>("rowKey");
  }

  /**
   * Sets the value of "indeterminate" property.
   * Visual only state to indicate partial selection
   * @param indeterminate The value to set for "indeterminate"
   * 
   */
  public changeIndeterminate(indeterminate: boolean): Promise<void> {
    return this.setProperty<boolean>("indeterminate", indeterminate);
  }

  /**
   * Gets the value of indeterminate property.
   * Visual only state to indicate partial selection
   * @return The value of indeterminate property.
   * 
   */
  public getIndeterminate(): Promise<boolean> {
    return this.getProperty<boolean>("indeterminate");
  }

  /**
   * Sets the value of "selectedKeys" property.
   * Specifies the selectedKeys, should be hooked into the collection component.
   * @param selectedKeys The value to set for "selectedKeys"
   * 
   */
  public changeSelectedKeys(selectedKeys: any): Promise<void> {
    return this.setProperty<any>("selectedKeys", selectedKeys);
  }

  /**
   * Gets the value of selectedKeys property.
   * Specifies the selectedKeys, should be hooked into the collection component.
   * @return The value of selectedKeys property.
   * 
   */
  public getSelectedKeys(): Promise<any> {
    return this.getProperty<any>("selectedKeys");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the selection mode.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

}
