// @oracle/oraclejet-webdriver/elements/oj-film-strip/OjFilmStripBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjFilmStrip, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjFilmStripBase extends OjWebElement {
  /**
   * Gets the value of arrowPlacement property.
   * Specifies the placement of the navigation arrows.
   * @return The value of arrowPlacement property.
   * 
   */
  public getArrowPlacement(): Promise<string> {
    return this.getProperty<string>("arrowPlacement");
  }

  /**
   * Gets the value of arrowVisibility property.
   * Specifies the visibility of the navigation arrows.
   * @return The value of arrowVisibility property.
   * 
   */
  public getArrowVisibility(): Promise<string> {
    return this.getProperty<string>("arrowVisibility");
  }

  /**
   * Sets the value of "currentItem" property.
   * Specifies the child item whose logical page should be displayed.
   * @param currentItem The value to set for "currentItem"
   * 
   */
  public changeCurrentItem(currentItem: OjFilmStripBase.CurrentItem): Promise<void> {
    return this.setProperty<OjFilmStripBase.CurrentItem>("currentItem", currentItem);
  }

  /**
   * Gets the value of currentItem property.
   * Specifies the child item whose logical page should be displayed.
   * @return The value of currentItem property.
   * 
   */
  public getCurrentItem(): Promise<OjFilmStripBase.CurrentItem> {
    return this.getProperty<OjFilmStripBase.CurrentItem>("currentItem");
  }

  /**
   * Gets the value of looping property.
   * Specifies the navigation looping behavior
   * @return The value of looping property.
   * 
   */
  public getLooping(): Promise<string> {
    return this.getProperty<string>("looping");
  }

  /**
   * Gets the value of maxItemsPerPage property.
   * Specifies the maximum number of child items shown in a logical page.
   * @return The value of maxItemsPerPage property.
   * 
   */
  public getMaxItemsPerPage(): Promise<number> {
    return this.getProperty<number>("maxItemsPerPage");
  }

  /**
   * Gets the value of orientation property.
   * Specifies the orientation of the FilmStrip.
   * @return The value of orientation property.
   * 
   */
  public getOrientation(): Promise<string> {
    return this.getProperty<string>("orientation");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjFilmStripBase.Translations|null> {
    return this.getProperty<OjFilmStripBase.Translations|null>("translations");
  }

}

namespace OjFilmStripBase {
  export interface CurrentItem {
    /**
     * string id of the item
     */
    "id": string;
    /**
     * numeric 0-based index of the item
     */
    "index": number;
  }
  export interface Translations {
    /**
     * Accessible label for the next page navigation arrow.
     */
    "labelAccArrowNextPage": string;
    /**
     * Accessible label for the previous page navigation arrow.
     */
    "labelAccArrowPreviousPage": string;
    /**
     * Accessible page information label for the filmstrip element.
     */
    "labelAccFilmStrip": string;
    /**
     * Tooltip for the next page navigation arrow.
     */
    "tipArrowNextPage": string;
    /**
     * Tooltip for the previous page navigation arrow.
     */
    "tipArrowPreviousPage": string;
  }
}
