// @oracle/oraclejet-webdriver/elements/oj-diagram/OjDiagramBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjDiagramBase = void 0;
const __1 = require("../");
class OjDiagramBase extends __1.OjWebElement {
    getAnimationOnDataChange() {
        return this.getProperty("animationOnDataChange");
    }
    getAnimationOnDisplay() {
        return this.getProperty("animationOnDisplay");
    }
    getAs() {
        return this.getProperty("as");
    }
    getData() {
        return this.getProperty("data");
    }
    getDnd() {
        return this.getProperty("dnd");
    }
    getFocusRenderer() {
        return this.getProperty("focusRenderer");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHighlightMatch() {
        return this.getProperty("highlightMatch");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getHoverRenderer() {
        return this.getProperty("hoverRenderer");
    }
    getLayout() {
        return this.getProperty("layout");
    }
    getLinkContent() {
        return this.getProperty("linkContent");
    }
    getLinkData() {
        return this.getProperty("linkData");
    }
    getLinkHighlightMode() {
        return this.getProperty("linkHighlightMode");
    }
    getLinkProperties() {
        return this.getProperty("linkProperties");
    }
    getMaxZoom() {
        return this.getProperty("maxZoom");
    }
    getMinZoom() {
        return this.getProperty("minZoom");
    }
    getNodeContent() {
        return this.getProperty("nodeContent");
    }
    getNodeData() {
        return this.getProperty("nodeData");
    }
    getNodeHighlightMode() {
        return this.getProperty("nodeHighlightMode");
    }
    getNodeProperties() {
        return this.getProperty("nodeProperties");
    }
    getOverview() {
        return this.getProperty("overview");
    }
    getPanDirection() {
        return this.getProperty("panDirection");
    }
    getPanning() {
        return this.getProperty("panning");
    }
    getPromotedLinkBehavior() {
        return this.getProperty("promotedLinkBehavior");
    }
    getRenderer() {
        return this.getProperty("renderer");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getSelectionRenderer() {
        return this.getProperty("selectionRenderer");
    }
    getStyleDefaults() {
        return this.getProperty("styleDefaults");
    }
    getTooltip() {
        return this.getProperty("tooltip");
    }
    getTouchResponse() {
        return this.getProperty("touchResponse");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getZoomRenderer() {
        return this.getProperty("zoomRenderer");
    }
    getZooming() {
        return this.getProperty("zooming");
    }
}
exports.OjDiagramBase = OjDiagramBase;
