// @oracle/oraclejet-webdriver/elements/oj-diagram/OjDiagramBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjDiagram, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjDiagramBase extends OjWebElement {
  /**
   * Gets the value of animationOnDataChange property.
   * Specifies the animation that is applied on data changes.
   * @return The value of animationOnDataChange property.
   * 
   */
  public getAnimationOnDataChange(): Promise<string> {
    return this.getProperty<string>("animationOnDataChange");
  }

  /**
   * Gets the value of animationOnDisplay property.
   * Specifies the animation that is shown on initial display.
   * @return The value of animationOnDisplay property.
   * 
   */
  public getAnimationOnDisplay(): Promise<string> {
    return this.getProperty<string>("animationOnDisplay");
  }

  /**
   * Gets the value of as property.
   * An alias for the '$current' context variable passed to slot content for the nodeTemplate or linkTemplate slots.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Gets the value of data property.
   * Specifies the data for the component.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object> {
    return this.getProperty<object>("data");
  }

  /**
   * Gets the value of dnd property.
   * Used to customize the drag and drop features.
   * @return The value of dnd property.
   * 
   */
  public getDnd(): Promise<OjDiagramBase.Dnd> {
    return this.getProperty<OjDiagramBase.Dnd>("dnd");
  }

  /**
   * Gets the value of focusRenderer property.
   * Specifies custom renderer for the diagram nodes used for focus updates.
   * @return The value of focusRenderer property.
   * @deprecated Since 8.0.0. Use nodeContent.focusRenderer instead.
   */
  public getFocusRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("focusRenderer");
  }

  /**
   * Sets the value of "hiddenCategories" property.
   * Specifies categories used for filtering.
   * @param hiddenCategories The value to set for "hiddenCategories"
   * 
   */
  public changeHiddenCategories(hiddenCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("hiddenCategories", hiddenCategories);
  }

  /**
   * Gets the value of hiddenCategories property.
   * Specifies categories used for filtering.
   * @return The value of hiddenCategories property.
   * 
   */
  public getHiddenCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("hiddenCategories");
  }

  /**
   * Gets the value of highlightMatch property.
   * Specifies matching condition used for category highlighting.
   * @return The value of highlightMatch property.
   * 
   */
  public getHighlightMatch(): Promise<string> {
    return this.getProperty<string>("highlightMatch");
  }

  /**
   * Sets the value of "highlightedCategories" property.
   * Specifies categories used for highlighting.
   * @param highlightedCategories The value to set for "highlightedCategories"
   * 
   */
  public changeHighlightedCategories(highlightedCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("highlightedCategories", highlightedCategories);
  }

  /**
   * Gets the value of highlightedCategories property.
   * Specifies categories used for highlighting.
   * @return The value of highlightedCategories property.
   * 
   */
  public getHighlightedCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("highlightedCategories");
  }

  /**
   * Gets the value of hoverBehavior property.
   * Defines the behavior applied when hovering over diagram nodes and links.
   * @return The value of hoverBehavior property.
   * 
   */
  public getHoverBehavior(): Promise<string> {
    return this.getProperty<string>("hoverBehavior");
  }

  /**
   * Gets the value of hoverRenderer property.
   * Specifies custom renderer for the diagram nodes used for hover updates.
   * @return The value of hoverRenderer property.
   * @deprecated Since 8.0.0. Use nodeContent.hoverRenderer instead.
   */
  public getHoverRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("hoverRenderer");
  }

  /**
   * Gets the value of layout property.
   * Specifies layout callback used to position nodes and links.
   * @return The value of layout property.
   * 
   */
  public getLayout(): Promise<null> {
    return this.getProperty<null>("layout");
  }

  /**
   * Gets the value of linkContent property.
   * An object containing an optional callbacks function for link customization.
   * @return The value of linkContent property.
   * 
   */
  public getLinkContent(): Promise<OjDiagramBase.LinkContent> {
    return this.getProperty<OjDiagramBase.LinkContent>("linkContent");
  }

  /**
   * Gets the value of linkData property.
   * Specifies the DataProvider for the diagram links. See the Help documentation for more information.
   * @return The value of linkData property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getLinkData(): Promise<object|null> {
    return this.getProperty<object|null>("linkData");
  }

  /**
   * Gets the value of linkHighlightMode property.
   * Defines link highlighting mode.
   * @return The value of linkHighlightMode property.
   * 
   */
  public getLinkHighlightMode(): Promise<string> {
    return this.getProperty<string>("linkHighlightMode");
  }

  /**
   * Gets the value of linkProperties property.
   * Optional callback for customizing link appearance based on link data.
   * @return The value of linkProperties property.
   * @deprecated Since 6.0.0. See nodeData and linkData usage.
   */
  public getLinkProperties(): Promise<null|null> {
    return this.getProperty<null|null>("linkProperties");
  }

  /**
   * Gets the value of maxZoom property.
   * Specifies the maximum zoom level for this diagram.
   * @return The value of maxZoom property.
   * 
   */
  public getMaxZoom(): Promise<number> {
    return this.getProperty<number>("maxZoom");
  }

  /**
   * Gets the value of minZoom property.
   * Specifies the minimum zoom level for this diagram.
   * @return The value of minZoom property.
   * 
   */
  public getMinZoom(): Promise<number> {
    return this.getProperty<number>("minZoom");
  }

  /**
   * Gets the value of nodeContent property.
   * An object containing an optional callbacks function for node customization.
   * @return The value of nodeContent property.
   * 
   */
  public getNodeContent(): Promise<OjDiagramBase.NodeContent> {
    return this.getProperty<OjDiagramBase.NodeContent>("nodeContent");
  }

  /**
   * Gets the value of nodeData property.
   * Specifies the DataProvider for the diagram nodes. See the Help documentation for more information.
   * @return The value of nodeData property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getNodeData(): Promise<object|null> {
    return this.getProperty<object|null>("nodeData");
  }

  /**
   * Gets the value of nodeHighlightMode property.
   * Defines node highlighting mode.
   * @return The value of nodeHighlightMode property.
   * 
   */
  public getNodeHighlightMode(): Promise<string> {
    return this.getProperty<string>("nodeHighlightMode");
  }

  /**
   * Gets the value of nodeProperties property.
   * Optional callback for customizing node appearance based on node data.
   * @return The value of nodeProperties property.
   * @deprecated Since 6.0.0. See nodeData and linkData usage.
   */
  public getNodeProperties(): Promise<null|null> {
    return this.getProperty<null|null>("nodeProperties");
  }

  /**
   * Gets the value of overview property.
   * An object, used to define a diagram overview. If not specified, no overview will be shown.
   * @return The value of overview property.
   * 
   */
  public getOverview(): Promise<OjDiagramBase.Overview> {
    return this.getProperty<OjDiagramBase.Overview>("overview");
  }

  /**
   * Gets the value of panDirection property.
   * Specifies if panning allowed in horizontal and vertical directions.
   * @return The value of panDirection property.
   * 
   */
  public getPanDirection(): Promise<string> {
    return this.getProperty<string>("panDirection");
  }

  /**
   * Gets the value of panning property.
   * Specifies whether panning is allowed in Diagram.
   * @return The value of panning property.
   * 
   */
  public getPanning(): Promise<string> {
    return this.getProperty<string>("panning");
  }

  /**
   * Gets the value of promotedLinkBehavior property.
   * Defines promoted link behavior for the component.
   * @return The value of promotedLinkBehavior property.
   * 
   */
  public getPromotedLinkBehavior(): Promise<string> {
    return this.getProperty<string>("promotedLinkBehavior");
  }

  /**
   * Gets the value of renderer property.
   * Specifies custom renderer for the diagram nodes used for initial rendering.
   * @return The value of renderer property.
   * @deprecated Since 8.0.0. Use nodeContent.renderer instead.
   */
  public getRenderer(): Promise<null> {
    return this.getProperty<null>("renderer");
  }

  /**
   * Sets the value of "selection" property.
   * An array containing the ids of the selected nodes and links.
   * @param selection The value to set for "selection"
   * 
   */
  public changeSelection(selection: Array<any>): Promise<void> {
    return this.setProperty<Array<any>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * An array containing the ids of the selected nodes and links.
   * @return The value of selection property.
   * 
   */
  public getSelection(): Promise<Array<any>> {
    return this.getProperty<Array<any>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the selection behavior on the diagram. See the Help documentation for more information.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<string> {
    return this.getProperty<string>("selectionMode");
  }

  /**
   * Gets the value of selectionRenderer property.
   * Specifies custom renderer for the diagram nodes used for selection updates.
   * @return The value of selectionRenderer property.
   * @deprecated Since 8.0.0. Use nodeContent.selectionRenderer instead.
   */
  public getSelectionRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("selectionRenderer");
  }

  /**
   * Gets the value of styleDefaults property.
   * An object defining the style defaults for this diagram.
   * @return The value of styleDefaults property.
   * 
   */
  public getStyleDefaults(): Promise<OjDiagramBase.StyleDefaults> {
    return this.getProperty<OjDiagramBase.StyleDefaults>("styleDefaults");
  }

  /**
   * Gets the value of tooltip property.
   * An object containing an optional callback function for tooltip customization.
   * @return The value of tooltip property.
   * 
   */
  public getTooltip(): Promise<OjDiagramBase.Tooltip> {
    return this.getProperty<OjDiagramBase.Tooltip>("tooltip");
  }

  /**
   * Gets the value of touchResponse property.
   * Specifies touch response behavior.
   * @return The value of touchResponse property.
   * 
   */
  public getTouchResponse(): Promise<string> {
    return this.getProperty<string>("touchResponse");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjDiagramBase.Translations|null> {
    return this.getProperty<OjDiagramBase.Translations|null>("translations");
  }

  /**
   * Gets the value of zoomRenderer property.
   * Specifies custom renderer for the diagram nodes used for zoom updates.
   * @return The value of zoomRenderer property.
   * @deprecated Since 8.0.0. Use nodeContent.zoomRenderer instead.
   */
  public getZoomRenderer(): Promise<null|null> {
    return this.getProperty<null|null>("zoomRenderer");
  }

  /**
   * Gets the value of zooming property.
   * Specifies whether zooming is allowed in Diagram.
   * @return The value of zooming property.
   * 
   */
  public getZooming(): Promise<string> {
    return this.getProperty<string>("zooming");
  }

}

namespace OjDiagramBase {
  export interface Dnd {
    /**
     * Used to customize drag functionality.
     */
    "drag": OjDiagramBase.DndDrag;
    /**
     * An object that describes drop functionality.
     */
    "drop": OjDiagramBase.DndDrop;
  }
  export interface DndDrag {
    /**
     * Used to customize the drag features for nodes.
     */
    "nodes": OjDiagramBase.DndDragNodes;
    /**
     * Defines link creation functionality.
     */
    "ports": OjDiagramBase.DndDragPorts;
  }
  export interface DndDragNodes {
    /**
     * Specifies MIME type for dragged data.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDragPorts {
    /**
     * Specifies MIME type for dragged data.
     */
    "dataTypes": string|Array<string>;
    /**
     * Specifies DOM element selector used to start link creation.
     */
    "selector": string;
  }
  export interface DndDrop {
    /**
     * Allows dropping on the diagram background.
     */
    "background": OjDiagramBase.DndDropBackground;
    /**
     * Allows dropping on diagram links.
     */
    "links": OjDiagramBase.DndDropLinks;
    /**
     * Allows dropping on diagram nodes.
     */
    "nodes": OjDiagramBase.DndDropNodes;
    /**
     * Allows dropping a link end on a port.
     */
    "ports": OjDiagramBase.DndDropPorts;
  }
  export interface DndDropBackground {
    /**
     * Specifies MIME types of objects that can be dropped on the Diagram background.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDropLinks {
    /**
     * Specifies MIME types of objects that can be dropped on the Diagram links.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDropNodes {
    /**
     * Specifies MIME types of objects that can be dropped on the Diagram nodes.
     */
    "dataTypes": string|Array<string>;
  }
  export interface DndDropPorts {
    /**
     * Specifies MIME types of objects that can be dropped on the Diagram ports.
     */
    "dataTypes": string|Array<string>;
    /**
     * Specifies DOM element selector used for link creation end.
     */
    "selector": string;
  }
  export interface LinkContent {
    /**
     * Specifies custom renderer for the diagram links used for focus updates.
     */
    "focusRenderer": null|null;
    /**
     * Specifies custom renderer for the diagram links used for hover updates.
     */
    "hoverRenderer": null|null;
    /**
     * Specifies custom renderer for the diagram links used for selection updates.
     */
    "selectionRenderer": null|null;
  }
  export interface NodeContent {
    /**
     * Specifies custom renderer for the diagram nodes used for focus updates.
     */
    "focusRenderer": null|null;
    /**
     * Specifies custom renderer for the diagram nodes used for hover updates.
     */
    "hoverRenderer": null|null;
    /**
     * Specifies custom renderer for the diagram nodes used for selection updates.
     */
    "selectionRenderer": null|null;
    /**
     * Specifies custom renderer for the diagram nodes used for zoom updates.
     */
    "zoomRenderer": null|null;
  }
  export interface Overview {
    /**
     * Specifies the region that will be scaled to fit within the overview.
     */
    "fitArea": string;
    /**
     * Horizontal alignment for diagram overview window
     */
    "halign": string;
    /**
     * Overview window height.
     */
    "height": number;
    /**
     * Controls how the fit area is scaled within the overview.
     */
    "preserveAspectRatio": string;
    /**
     * Specifies whether the overview scrollbar is rendered.
     */
    "rendered": string;
    /**
     * Vertical alignment for diagram overview window
     */
    "valign": string;
    /**
     * Overview window width.
     */
    "width": number;
  }
  export interface StyleDefaults {
    /**
     * Defines animation duration in milliseconds.
     */
    "animationDuration": number;
    /**
     * Specifies initial hover delay in milliseconds for highlighting data items.
     */
    "hoverBehaviorDelay": number;
    /**
     * Default link styles
     */
    "linkDefaults": OjDiagramBase.StyleDefaultsLinkDefaults;
    /**
     * Default node styles
     */
    "nodeDefaults": OjDiagramBase.StyleDefaultsNodeDefaults;
    /**
     * Promoted link styles
     */
    "promotedLink": OjDiagramBase.StyleDefaultsPromotedLink;
  }
  export interface StyleDefaultsLinkDefaults {
    /**
     * Default link color.
     */
    "color": string;
    /**
     * Specifies the type of end connector on the link.
     */
    "endConnectorType": string;
    /**
     * Specifies CSS styles for the link label.
     */
    "labelStyle": object;
    /**
     * Specifies the type of start connector on the link.
     */
    "startConnectorType": string;
    /**
     * The default SVG CSS style class to apply to the link.
     */
    "svgClassName": string;
    /**
     * Specifies SVG CSS styles for the link.
     */
    "svgStyle": object;
    /**
     * Default link width in pixels.
     */
    "width": number;
  }
  export interface StyleDefaultsNodeDefaults {
    /**
     * Default style for the node icon.
     */
    "icon": OjDiagramBase.StyleDefaultsNodeDefaultsIcon;
    /**
     * Specifies CSS styles for the node label.
     */
    "labelStyle": object;
    /**
     * Determines when to display the expand/collapse button.
     */
    "showDisclosure": string;
  }
  export interface StyleDefaultsNodeDefaultsIcon {
    /**
     * Default border color of the icon.
     */
    "borderColor": string;
    /**
     * Specifies default border radius of the icon.
     */
    "borderRadius": string;
    /**
     * Default border width of the icon in pixels.
     */
    "borderWidth": number;
    /**
     * Default color of the icon.
     */
    "color": string;
    /**
     * Default icon height.
     */
    "height": number;
    /**
     * Default fill pattern of the icon.
     */
    "pattern": string;
    /**
     * Specifies default shape of the icon.
     */
    "shape": string;
    /**
     * The URI of the node image
     */
    "source": string;
    /**
     * The optional URI of the node hover image.
     */
    "sourceHover": string;
    /**
     * The optional URI of the selected image on hover.
     */
    "sourceHoverSelected": string;
    /**
     * The optional URI of the selected image.
     */
    "sourceSelected": string;
    /**
     * The SVG CSS style class to apply to the node icon.
     */
    "svgClassName": string;
    /**
     * Specifies CSS styles for the icon.
     */
    "svgStyle": object;
    /**
     * Default icon width.
     */
    "width": number;
  }
  export interface StyleDefaultsPromotedLink {
    /**
     * Default promoted link color. The default value varies based on theme.
     */
    "color": string;
    /**
     * Specifies the type of end connector on the promoted link.
     */
    "endConnectorType": string;
    /**
     * Specifies the type of start connector on the promoted link.
     */
    "startConnectorType": string;
    /**
     * The SVG CSS style class to apply to the promoted link.
     */
    "svgClassName": string;
    /**
     * Specifies CSS styles of the promoted link.
     */
    "svgStyle": object;
    /**
     * Default link width in pixels.
     */
    "width": number;
  }
  export interface Tooltip {
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used for the promoted link tooltip - promoted link consists of a single link.
     */
    "promotedLink": string;
    /**
     * Used to describe promoted link type for accessibility.
     */
    "promotedLinkAriaDesc": string;
    /**
     * Used for the promoted link tooltip - promoted link consists of multiple links.
     */
    "promotedLinks": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
  }
}
