// @oracle/oraclejet-webdriver/elements/oj-data-grid/OjDataGridBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjDataGrid, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjDataGridBase extends OjWebElement {
  /**
   * Gets the value of bandingInterval property.
   * Specifies the data body row and column banding intervals.
   * @return The value of bandingInterval property.
   * 
   */
  public getBandingInterval(): Promise<OjDataGridBase.BandingInterval> {
    return this.getProperty<OjDataGridBase.BandingInterval>("bandingInterval");
  }

  /**
   * Gets the value of cell property.
   * Specifies attributes on the cells.
   * @return The value of cell property.
   * 
   */
  public getCell(): Promise<OjDataGridBase.Cell> {
    return this.getProperty<OjDataGridBase.Cell>("cell");
  }

  /**
   * Sets the value of "currentCell" property.
   * Specifies the grid's currently focused cell.
   * @param currentCell The value to set for "currentCell"
   * 
   */
  public changeCurrentCell(currentCell: object): Promise<void> {
    return this.setProperty<object>("currentCell", currentCell);
  }

  /**
   * Gets the value of currentCell property.
   * Specifies the grid's currently focused cell.
   * @return The value of currentCell property.
   * 
   */
  public getCurrentCell(): Promise<object> {
    return this.getProperty<object>("currentCell");
  }

  /**
   * Gets the value of data property.
   * Specifies the data for the grid. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<null> {
    return this.getProperty<null>("data");
  }

  /**
   * Gets the value of dnd property.
   * Specifies drag and drop features.
   * @return The value of dnd property.
   * 
   */
  public getDnd(): Promise<OjDataGridBase.Dnd> {
    return this.getProperty<OjDataGridBase.Dnd>("dnd");
  }

  /**
   * Sets the value of "editMode" property.
   * Specifies if the grid is read-only or editable.
   * @param editMode The value to set for "editMode"
   * 
   */
  public changeEditMode(editMode: string): Promise<void> {
    return this.setProperty<string>("editMode", editMode);
  }

  /**
   * Gets the value of editMode property.
   * Specifies if the grid is read-only or editable.
   * @return The value of editMode property.
   * 
   */
  public getEditMode(): Promise<string> {
    return this.getProperty<string>("editMode");
  }

  /**
   * Gets the value of gridlines property.
   * Specifies the visibility of the grid body vertical and horizontal gridlines.
   * @return The value of gridlines property.
   * 
   */
  public getGridlines(): Promise<OjDataGridBase.Gridlines> {
    return this.getProperty<OjDataGridBase.Gridlines>("gridlines");
  }

  /**
   * Gets the value of header property.
   * Specifies attributes on the headers.
   * @return The value of header property.
   * 
   */
  public getHeader(): Promise<OjDataGridBase.Header> {
    return this.getProperty<OjDataGridBase.Header>("header");
  }

  /**
   * Gets the value of scrollPolicy property.
   * Specifies how data are fetched as user scrolls through the data grid.
   * @return The value of scrollPolicy property.
   * 
   */
  public getScrollPolicy(): Promise<string|null> {
    return this.getProperty<string|null>("scrollPolicy");
  }

  /**
   * Gets the value of scrollPolicyOptions property.
   * Specifies fetch options for scrolling behaviors that trigger data fetches.
   * @return The value of scrollPolicyOptions property.
   * 
   */
  public getScrollPolicyOptions(): Promise<OjDataGridBase.ScrollPolicyOptions|null> {
    return this.getProperty<OjDataGridBase.ScrollPolicyOptions|null>("scrollPolicyOptions");
  }

  /**
   * Sets the value of "scrollPosition" property.
   * Specifies the current scroll position of the data grid. See the Help documentation for more information.
   * @param scrollPosition The value to set for "scrollPosition"
   * 
   */
  public changeScrollPosition(scrollPosition: OjDataGridBase.ScrollPosition): Promise<void> {
    return this.setProperty<OjDataGridBase.ScrollPosition>("scrollPosition", scrollPosition);
  }

  /**
   * Gets the value of scrollPosition property.
   * Specifies the current scroll position of the data grid. See the Help documentation for more information.
   * @return The value of scrollPosition property.
   * 
   */
  public getScrollPosition(): Promise<OjDataGridBase.ScrollPosition> {
    return this.getProperty<OjDataGridBase.ScrollPosition>("scrollPosition");
  }

  /**
   * Gets the value of scrollToKey property.
   * Specifies the behavior when DataGrid needs to scroll to a position based on a row/column key.
   * @return The value of scrollToKey property.
   * 
   */
  public getScrollToKey(): Promise<string|null> {
    return this.getProperty<string|null>("scrollToKey");
  }

  /**
   * Sets the value of "selection" property.
   * Specifies the grid's current selection.
   * @param selection The value to set for "selection"
   * 
   */
  public changeSelection(selection: Array<object>): Promise<void> {
    return this.setProperty<Array<object>>("selection", selection);
  }

  /**
   * Gets the value of selection property.
   * Specifies the grid's current selection.
   * @return The value of selection property.
   * 
   */
  public getSelection(): Promise<Array<object>> {
    return this.getProperty<Array<object>>("selection");
  }

  /**
   * Gets the value of selectionMode property.
   * Specifies the row and cell selection modes. See the Help documentation for more information.
   * @return The value of selectionMode property.
   * 
   */
  public getSelectionMode(): Promise<OjDataGridBase.SelectionMode> {
    return this.getProperty<OjDataGridBase.SelectionMode>("selectionMode");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjDataGridBase.Translations|null> {
    return this.getProperty<OjDataGridBase.Translations|null>("translations");
  }

}

namespace OjDataGridBase {
  export interface BandingInterval {
    /**
     * Specifies the data body column banding intervals.
     */
    "column": number;
    /**
     * Specifies the data body row banding intervals.
     */
    "row": number;
  }
  export interface Cell {
    /**
     * Specifies CSS class names to apply to cells. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * A function that returns cell content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies inline style to add to the cells. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface Dnd {
    /**
     * Specifies drag and drop reordering features.
     */
    "reorder": OjDataGridBase.DndReorder;
  }
  export interface DndReorder {
    /**
     * Specifies whether drag and drop row reordering is enabled.
     */
    "row": string;
  }
  export interface Gridlines {
    /**
     * Specifies the visibility of the horizontal gridlines.
     */
    "horizontal": string;
    /**
     * Specifies the visibility of the vertical gridlines.
     */
    "vertical": string;
  }
  export interface Header {
    /**
     * Specifies attributes on the column headers.
     */
    "column": OjDataGridBase.HeaderColumn;
    /**
     * Specifies attributes on the column end headers.
     */
    "columnEnd": OjDataGridBase.HeaderColumnEnd;
    /**
     * Specifies attributes on the row headers.
     */
    "row": OjDataGridBase.HeaderRow;
    /**
     * Specifies attributes on the row end headers.
     */
    "rowEnd": OjDataGridBase.HeaderRowEnd;
  }
  export interface HeaderColumn {
    /**
     * Specifies CSS class names to apply to column headers. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * Specifies attributes on the column header labels.
     */
    "label": OjDataGridBase.HeaderColumnLabel;
    /**
     * A function that returns column header content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies resizing options along the column headers.
     */
    "resizable": OjDataGridBase.HeaderColumnResizable;
    /**
     * Specifies whether sorting is enabled on the column headers. See the Help documentation for more information.
     */
    "sortable": null|string;
    /**
     * Specifies inline style to add to the column headers. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderColumnLabel {
    /**
     * Specifies CSS class names to apply to column header labels. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * A function that returns column header label content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies inline style to apply to column header labels. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderColumnResizable {
    /**
     * Enable or disable height resizing along the column headers.
     */
    "height": string;
    /**
     * Enable or disable width resizing along the column headers.
     */
    "width": string|null;
  }
  export interface HeaderColumnEnd {
    /**
     * Specifies CSS class names to apply to column end headers. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * Specifies attributes on the column end header labels.
     */
    "label": OjDataGridBase.HeaderColumnEndLabel;
    /**
     * A function that returns column end header content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies resizing options along the column end headers.
     */
    "resizable": OjDataGridBase.HeaderColumnEndResizable;
    /**
     * Specifies inline style to add to the column end headers. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderColumnEndLabel {
    /**
     * Specifies CSS class names to apply to column end header labels. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * A function that returns column end header label content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies inline style to apply to column end header labels. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderColumnEndResizable {
    /**
     * Enable or disable height resizing along the column end headers.
     */
    "height": string;
    /**
     * Enable or disable width resizing along the column end headers.
     */
    "width": string|null;
  }
  export interface HeaderRow {
    /**
     * Specifies CSS class names to apply to row headers. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * Specifies attributes on the row header labels.
     */
    "label": OjDataGridBase.HeaderRowLabel;
    /**
     * A function that returns row header content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies resizing options on the row headers.
     */
    "resizable": OjDataGridBase.HeaderRowResizable;
    /**
     * Specifies whether sorting is enabled on the row headers. See the Help documentation for more information.
     */
    "sortable": null|string;
    /**
     * Specifies inline style to add to the row headers. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderRowLabel {
    /**
     * Specifies CSS class names to apply to row header labels. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * A function that returns row header label content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies inline style to apply to row header labels. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderRowResizable {
    /**
     * Enable or disable height resizing along the row headers.
     */
    "height": string|null;
    /**
     * Enable or disable width resizing along the row headers.
     */
    "width": string;
  }
  export interface HeaderRowEnd {
    /**
     * Specifies CSS class names to apply to row end headers. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * Specifies attributes on the row end header labels.
     */
    "label": OjDataGridBase.HeaderRowEndLabel;
    /**
     * A function that returns row end header content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies resizing options along the row end headers.
     */
    "resizable": OjDataGridBase.HeaderRowEndResizable;
    /**
     * Specifies inline style to add to the row end headers. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderRowEndLabel {
    /**
     * Specifies CSS class names to apply to row end header labels. See the Help documentation for more information.
     */
    "className": null|string|null;
    /**
     * A function that returns row end header label content. The function takes a context argument, provided by the datagrid. See the Help documentation for more information.
     */
    "renderer": null|null;
    /**
     * Specifies inline style to apply to row end header labels. See the Help documentation for more information.
     */
    "style": null|string|null;
  }
  export interface HeaderRowEndResizable {
    /**
     * Enable or disable height resizing along the row end headers.
     */
    "height": string|null;
    /**
     * Enable or disable width resizing along the row end headers.
     */
    "width": string;
  }
  export interface ScrollPolicyOptions {
    /**
     * The maximum number of columns to display before fetching more data rows will be stopped.
     */
    "maxColumnCount": number;
    /**
     * The maximum number of rows to display before fetching more data rows will be stopped.
     */
    "maxRowCount": number;
  }
  export interface ScrollPosition {
    /**
     * The zero-based column index of the cell at the origin of the grid. If scrollPolicy is set to 'loadMoreOnScroll' and the index is greater than maxCount set in scrollPolicyOptions, then it will scroll and fetch until the end of the list is reached and there's no more items to fetch.
     */
    "columnIndex": number;
    /**
     * The column key of the cell at the origin of the grid. If DataGridDataSource is used for data and the key does not exist in the DataGridDataSource, then the value is ignored. If it is unknown in the data source then the grid will fetch and scroll until the item is found or the end of the axis is reached and there's no more items to fetch.
     */
    "columnKey": any;
    /**
     * The horizontal offset in pixel relative to the cell identified by key/index.
     */
    "offsetX": number;
    /**
     * The vertical offset in pixel relative to the cell identified by key/index.
     */
    "offsetY": number;
    /**
     * The zero-based row index of the cell at the origin of the grid. If scrollPolicy is set to 'loadMoreOnScroll' and the index is greater than maxCount set in scrollPolicyOptions, then it will scroll and fetch until the end of the list is reached and there's no more items to fetch.
     */
    "rowIndex": number;
    /**
     * The row key of the cell at the origin of the grid. If DataGridDataSource is used for data and the key does not exist in the DataGridDataSource, then the value is ignored. If it is unknown in the data source then the grid will fetch and scroll until the item is found or the end of the axis is reached and there's no more items to fetch.
     */
    "rowKey": any;
    /**
     * The horizontal position in pixels.
     */
    "x": number;
    /**
     * The vertical position in pixels.
     */
    "y": number;
  }
  export interface SelectionMode {
    /**
     * Specifies the cell selection mode. By default, cell selection is disabled.
     */
    "cell": string;
    /**
     * Specifies the row selection mode. By default, row selection is disabled.
     */
    "row": string;
  }
  export interface Translations {
    /**
     * Provides properties to customize the accessible context to enter actionable mode.
     */
    "accessibleActionableMode": string;
    /**
     * Provides properties to customize the accessible context for the column index.
     */
    "accessibleColumnContext": string;
    /**
     * Provides properties to customize the accessible context for the column end header index.
     */
    "accessibleColumnEndHeaderContext": string;
    /**
     * Provides properties to customize the accessible context for the column end header label.
     */
    "accessibleColumnEndHeaderLabelContext": string;
    /**
     * Provides properties to customize the accessible context for the column header index.
     */
    "accessibleColumnHeaderContext": string;
    /**
     * Provides properties to customize the accessible context for the column header label.
     */
    "accessibleColumnHeaderLabelContext": string;
    /**
     * Provides properties to customize the accessible context when a column is selected.
     */
    "accessibleColumnSelected": string;
    /**
     * Provides properties to customize the accessible context for the cell column extent/span.
     */
    "accessibleColumnSpanContext": string;
    /**
     * Provides properties to customize the accessible text when a cell contains controls.
     */
    "accessibleContainsControls": string;
    /**
     * Provides properties to customize the accessible context when the first column is reached.
     */
    "accessibleFirstColumn": string;
    /**
     * Provides properties to customize the accessible context when the first row is reached.
     */
    "accessibleFirstRow": string;
    /**
     * Provides properties to customize the accessible context when the last column is reached.
     */
    "accessibleLastColumn": string;
    /**
     * Provides properties to customize the accessible context when the last row is reached.
     */
    "accessibleLastRow": string;
    /**
     * Provides properties to customize the accessible context for the header level.
     */
    "accessibleLevelContext": string;
    /**
     * Provides properties to customize the accessible context when multiple cells are selected.
     */
    "accessibleMultiCellSelected": string;
    /**
     * Provides properties to customize the accessible context to enter navigation mode.
     */
    "accessibleNavigationMode": string;
    /**
     * Provides properties to customize the accessible context for discontinuous selection off.
     */
    "accessibleRangeSelectModeOff": string;
    /**
     * Provides properties to customize the accessible context for discontinuous selection on.
     */
    "accessibleRangeSelectModeOn": string;
    /**
     * Provides properties to customize the accessible context when a row is collapsed.
     */
    "accessibleRowCollapsed": string;
    /**
     * Provides properties to customize the accessible context for the row index.
     */
    "accessibleRowContext": string;
    /**
     * Provides properties to customize the accessible context for the row end header index.
     */
    "accessibleRowEndHeaderContext": string;
    /**
     * Provides properties to customize the accessible context for the row end header label.
     */
    "accessibleRowEndHeaderLabelContext": string;
    /**
     * Provides properties to customize the accessible context when a row is expanded.
     */
    "accessibleRowExpanded": string;
    /**
     * Provides properties to customize the accessible context for the row header index.
     */
    "accessibleRowHeaderContext": string;
    /**
     * Provides properties to customize the accessible context for the row header label.
     */
    "accessibleRowHeaderLabelContext": string;
    /**
     * Provides properties to customize the accessible context when a row is selected.
     */
    "accessibleRowSelected": string;
    /**
     * Provides properties to customize the accessible context for the cell extent/span.
     */
    "accessibleRowSpanContext": string;
    /**
     * Provides properties to customize the accessible context for the bottom selection affordance on touch device.
     */
    "accessibleSelectionAffordanceBottom": string;
    /**
     * Provides properties to customize the accessible context for the top selection affordance on touch devices.
     */
    "accessibleSelectionAffordanceTop": string;
    /**
     * Provides properties to customize the accessible text when a header is sorted ascending.
     */
    "accessibleSortAscending": string;
    /**
     * Provides properties to customize the accessible text when a header is sorted descending.
     */
    "accessibleSortDescending": string;
    /**
     * Provides properties to customize the accessible context when an item has been selected.
     */
    "accessibleStateSelected": string;
    /**
     * Provides properties to customize the accessible context read when the exact row and column count are unknown.
     */
    "accessibleSummaryEstimate": string;
    /**
     * Provides properties to customize the accessible context read when the exact row and column count are known.
     */
    "accessibleSummaryExact": string;
    /**
     * Provides properties to customize the accessible context reading out the total number of rows expanded inside the data grid.
     */
    "accessibleSummaryExpanded": string;
    /**
     * Provides properties to customize the context menu cut label for row reordering.
     */
    "labelCut": string;
    /**
     * Provides properties to customize the context menu label for exiting non-contiguous selection.
     */
    "labelDisableNonContiguous": string;
    /**
     * Provides properties to customize the context menu label for entering non-contiguous selection.
     */
    "labelEnableNonContiguous": string;
    /**
     * Provides properties to customize the context menu paste label for row reordering.
     */
    "labelPaste": string;
    /**
     * Provides properties to customize the context menu resize label.
     */
    "labelResize": string;
    /**
     * Provides properties to customize the resize dialog cancel button.
     */
    "labelResizeDialogCancel": string;
    /**
     * Provides properties to customize the resize dialog submit button.
     */
    "labelResizeDialogSubmit": string;
    /**
     * Provides properties to customize the context menu resize height label.
     */
    "labelResizeHeight": string;
    /**
     * Provides properties to customize the context menu resize width label.
     */
    "labelResizeWidth": string;
    /**
     * Provides properties to customize the context menu sort column label.
     */
    "labelSortCol": string;
    /**
     * Provides properties to customize the context menu sort column ascending label.
     */
    "labelSortColAsc": string;
    /**
     * Provides properties to customize the context menu sort column descending label.
     */
    "labelSortColDsc": string;
    /**
     * Provides properties to customize the context menu sort row label.
     */
    "labelSortRow": string;
    /**
     * Provides properties to customize the context menu sort row ascending label.
     */
    "labelSortRowAsc": string;
    /**
     * Provides properties to customize the context menu sort row descending label.
     */
    "labelSortRowDsc": string;
    /**
     * Provides properties to customize the text when fetching data.
     */
    "msgFetchingData": string;
    /**
     * Provides properties to customize the empty data grid text.
     */
    "msgNoData": string;
  }
}
