// @oracle/oraclejet-webdriver/elements/oj-data-grid/OjDataGridBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjDataGridBase = void 0;
const __1 = require("../");
class OjDataGridBase extends __1.OjWebElement {
    getBandingInterval() {
        return this.getProperty("bandingInterval");
    }
    getCell() {
        return this.getProperty("cell");
    }
    changeCurrentCell(currentCell) {
        return this.setProperty("currentCell", currentCell);
    }
    getCurrentCell() {
        return this.getProperty("currentCell");
    }
    getData() {
        return this.getProperty("data");
    }
    getDnd() {
        return this.getProperty("dnd");
    }
    changeEditMode(editMode) {
        return this.setProperty("editMode", editMode);
    }
    getEditMode() {
        return this.getProperty("editMode");
    }
    getGridlines() {
        return this.getProperty("gridlines");
    }
    getHeader() {
        return this.getProperty("header");
    }
    getScrollPolicy() {
        return this.getProperty("scrollPolicy");
    }
    getScrollPolicyOptions() {
        return this.getProperty("scrollPolicyOptions");
    }
    changeScrollPosition(scrollPosition) {
        return this.setProperty("scrollPosition", scrollPosition);
    }
    getScrollPosition() {
        return this.getProperty("scrollPosition");
    }
    getScrollToKey() {
        return this.getProperty("scrollToKey");
    }
    changeSelection(selection) {
        return this.setProperty("selection", selection);
    }
    getSelection() {
        return this.getProperty("selection");
    }
    getSelectionMode() {
        return this.getProperty("selectionMode");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
}
exports.OjDataGridBase = OjDataGridBase;
