// @oracle/oraclejet-webdriver/elements/oj-legend/OjLegendBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjLegend, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjLegendBase extends OjWebElement {
  /**
   * Gets the value of as property.
   * An alias for the '$current' context variable passed to slot content for the nodeTemplate slot.
   * @return The value of as property.
   * @deprecated Since 6.2.0. Set the alias directly on the template element using the data-oj-as attribute instead.
   */
  public getAs(): Promise<string> {
    return this.getProperty<string>("as");
  }

  /**
   * Gets the value of data property.
   * Specifies the DataProvider for the sections and items of the legend. See the Help documentation for more information.
   * @return The value of data property.
   * @deprecated Since 11.0.0. Data sets from a DataProvider cannot be sent to WebDriverJS; use ViewModels or page variables instead.
   */
  public getData(): Promise<object|null> {
    return this.getProperty<object|null>("data");
  }

  /**
   * Gets the value of drilling property.
   * Specifies whether drilling is enabled. Drillable objects will show a pointer cursor on hover and fire an ojDrill event on click. See the Help documentation for more information.
   * @return The value of drilling property.
   * 
   */
  public getDrilling(): Promise<string> {
    return this.getProperty<string>("drilling");
  }

  /**
   * Sets the value of "expanded" property.
   * Specifies the key set containing the ids of sections that should be expanded on initial render. See the Help documentation for more information.
   * @param expanded The value to set for "expanded"
   * 
   */
  public changeExpanded(expanded: null): Promise<void> {
    return this.setProperty<null>("expanded", expanded);
  }

  /**
   * Gets the value of expanded property.
   * Specifies the key set containing the ids of sections that should be expanded on initial render. See the Help documentation for more information.
   * @return The value of expanded property.
   * 
   */
  public getExpanded(): Promise<null> {
    return this.getProperty<null>("expanded");
  }

  /**
   * Gets the value of halign property.
   * Defines the horizontal alignment of the legend contents.
   * @return The value of halign property.
   * 
   */
  public getHalign(): Promise<string> {
    return this.getProperty<string>("halign");
  }

  /**
   * Sets the value of "hiddenCategories" property.
   * An array of categories that will be hidden.
   * @param hiddenCategories The value to set for "hiddenCategories"
   * 
   */
  public changeHiddenCategories(hiddenCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("hiddenCategories", hiddenCategories);
  }

  /**
   * Gets the value of hiddenCategories property.
   * An array of categories that will be hidden.
   * @return The value of hiddenCategories property.
   * 
   */
  public getHiddenCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("hiddenCategories");
  }

  /**
   * Gets the value of hideAndShowBehavior property.
   * Defines whether the legend can be used to initiate hide and show behavior on referenced data items.
   * @return The value of hideAndShowBehavior property.
   * 
   */
  public getHideAndShowBehavior(): Promise<string> {
    return this.getProperty<string>("hideAndShowBehavior");
  }

  /**
   * Sets the value of "highlightedCategories" property.
   * An array of categories that will be highlighted.
   * @param highlightedCategories The value to set for "highlightedCategories"
   * 
   */
  public changeHighlightedCategories(highlightedCategories: Array<string>): Promise<void> {
    return this.setProperty<Array<string>>("highlightedCategories", highlightedCategories);
  }

  /**
   * Gets the value of highlightedCategories property.
   * An array of categories that will be highlighted.
   * @return The value of highlightedCategories property.
   * 
   */
  public getHighlightedCategories(): Promise<Array<string>> {
    return this.getProperty<Array<string>>("highlightedCategories");
  }

  /**
   * Gets the value of hoverBehavior property.
   * Defines the behavior applied when hovering over a legend item.
   * @return The value of hoverBehavior property.
   * 
   */
  public getHoverBehavior(): Promise<string> {
    return this.getProperty<string>("hoverBehavior");
  }

  /**
   * Gets the value of hoverBehaviorDelay property.
   * Specifies initial hover delay in milliseconds for highlighting items in legend.
   * @return The value of hoverBehaviorDelay property.
   * 
   */
  public getHoverBehaviorDelay(): Promise<number> {
    return this.getProperty<number>("hoverBehaviorDelay");
  }

  /**
   * Gets the value of orientation property.
   * Defines the orientation of the legend, which determines the direction in which the legend items are laid out.
   * @return The value of orientation property.
   * 
   */
  public getOrientation(): Promise<string> {
    return this.getProperty<string>("orientation");
  }

  /**
   * Gets the value of scrolling property.
   * Defines whether scrolling is enabled for the legend.
   * @return The value of scrolling property.
   * 
   */
  public getScrolling(): Promise<string> {
    return this.getProperty<string>("scrolling");
  }

  /**
   * Gets the value of sections property.
   * An array of objects specifying the legend sections.
   * @return The value of sections property.
   * 
   */
  public getSections(): Promise<Array<OjLegendBase.Sections>|null> {
    return this.getProperty<Array<OjLegendBase.Sections>|null>("sections");
  }

  /**
   * Gets the value of symbolHeight property.
   * The height of the legend symbol in pixels. See the Help documentation for more information.
   * @return The value of symbolHeight property.
   * 
   */
  public getSymbolHeight(): Promise<number> {
    return this.getProperty<number>("symbolHeight");
  }

  /**
   * Gets the value of symbolWidth property.
   * The width of the legend symbol in pixels. See the Help documentation for more information.
   * @return The value of symbolWidth property.
   * 
   */
  public getSymbolWidth(): Promise<number> {
    return this.getProperty<number>("symbolWidth");
  }

  /**
   * Gets the value of textStyle property.
   * The CSS style object defining the style of the legend item text.
   * @return The value of textStyle property.
   * 
   */
  public getTextStyle(): Promise<object> {
    return this.getProperty<object>("textStyle");
  }

  /**
   * Gets the value of trackResize property.
   * Defines whether the element will automatically render in response to changes in size. See the Help documentation for more information.
   * @return The value of trackResize property.
   * 
   */
  public getTrackResize(): Promise<string> {
    return this.getProperty<string>("trackResize");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjLegendBase.Translations|null> {
    return this.getProperty<OjLegendBase.Translations|null>("translations");
  }

  /**
   * Gets the value of valign property.
   * Defines the vertical alignment of the legend contents.
   * @return The value of valign property.
   * 
   */
  public getValign(): Promise<string> {
    return this.getProperty<string>("valign");
  }

}

namespace OjLegendBase {
  export interface Sections {
    /**
     * Whether the section is collapsible. Only applies if the legend orientation is vertical.
     */
    "collapsible": string;
    /**
     * Whether the section is initially expanded. Only applies if the section is collapsible.
     */
    "expanded": string;
    /**
     * The id of the legend section. For the DataProvider case, the key for the node will be used as the id.
     */
    "id": string;
    /**
     * An array of objects with the following properties defining the legend items. Also accepts a Promise for deferred data rendering. No data will be rendered if the Promise is rejected.
     */
    "items": Array<object>;
    /**
     * An array of nested legend sections.
     */
    "sections": Array<object>;
    /**
     * The title of the legend section.
     */
    "title": string;
    /**
     * The horizontal alignment of the section title. If the section is collapsible or nested, only start alignment is supported.
     */
    "titleHalign": string;
    /**
     * The CSS style object defining the style of the section title. The following style properties are supported: color, cursor, fontFamily, fontSize, fontStyle, fontWeight, textDecoration.
     */
    "titleStyle": object;
  }
  export interface Translations {
    /**
     * Used to describe the data visualization type for accessibility.
     */
    "componentName": string;
    /**
     * Used to display a label and its value.
     */
    "labelAndValue": string;
    /**
     * Text shown for clearing multiple selection on touch devices.
     */
    "labelClearSelection": string;
    /**
     * Used to display a count out of a total.
     */
    "labelCountWithTotal": string;
    /**
     * Label for data visualizations used for accessibility.
     */
    "labelDataVisualization": string;
    /**
     * Text shown when the component receives invalid data.
     */
    "labelInvalidData": string;
    /**
     * Text shown when the component receives no data.
     */
    "labelNoData": string;
    /**
     * Used to describe the collapsed state for accessibility.
     */
    "stateCollapsed": string;
    /**
     * Used to describe a drillable object for accessibility.
     */
    "stateDrillable": string;
    /**
     * Used to describe the expanded state for accessibility.
     */
    "stateExpanded": string;
    /**
     * Used to describe the hidden state for accessibility.
     */
    "stateHidden": string;
    /**
     * Used to describe the isolated state for accessibility.
     */
    "stateIsolated": string;
    /**
     * Used to describe the maximized state for accessibility.
     */
    "stateMaximized": string;
    /**
     * Used to describe the minimized state for accessibility.
     */
    "stateMinimized": string;
    /**
     * Used to describe the selected state for accessibility.
     */
    "stateSelected": string;
    /**
     * Used to describe the unselected state for accessibility.
     */
    "stateUnselected": string;
    /**
     * Used to describe the visible state for accessibility.
     */
    "stateVisible": string;
    /**
     * Used for the collapse tooltip.
     */
    "tooltipCollapse": string;
    /**
     * Used for the expand tooltip.
     */
    "tooltipExpand": string;
  }
}
