// @oracle/oraclejet-webdriver/elements/oj-legend/OjLegendBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjLegendBase = void 0;
const __1 = require("../");
class OjLegendBase extends __1.OjWebElement {
    getAs() {
        return this.getProperty("as");
    }
    getData() {
        return this.getProperty("data");
    }
    getDrilling() {
        return this.getProperty("drilling");
    }
    changeExpanded(expanded) {
        return this.setProperty("expanded", expanded);
    }
    getExpanded() {
        return this.getProperty("expanded");
    }
    getHalign() {
        return this.getProperty("halign");
    }
    changeHiddenCategories(hiddenCategories) {
        return this.setProperty("hiddenCategories", hiddenCategories);
    }
    getHiddenCategories() {
        return this.getProperty("hiddenCategories");
    }
    getHideAndShowBehavior() {
        return this.getProperty("hideAndShowBehavior");
    }
    changeHighlightedCategories(highlightedCategories) {
        return this.setProperty("highlightedCategories", highlightedCategories);
    }
    getHighlightedCategories() {
        return this.getProperty("highlightedCategories");
    }
    getHoverBehavior() {
        return this.getProperty("hoverBehavior");
    }
    getHoverBehaviorDelay() {
        return this.getProperty("hoverBehaviorDelay");
    }
    getOrientation() {
        return this.getProperty("orientation");
    }
    getScrolling() {
        return this.getProperty("scrolling");
    }
    getSections() {
        return this.getProperty("sections");
    }
    getSymbolHeight() {
        return this.getProperty("symbolHeight");
    }
    getSymbolWidth() {
        return this.getProperty("symbolWidth");
    }
    getTextStyle() {
        return this.getProperty("textStyle");
    }
    getTrackResize() {
        return this.getProperty("trackResize");
    }
    getTranslations() {
        return this.getProperty("translations");
    }
    getValign() {
        return this.getProperty("valign");
    }
}
exports.OjLegendBase = OjLegendBase;
