// @oracle/oraclejet-webdriver/elements/oj-row-expander/OjRowExpanderBase.ts
import {OjWebElement} from "../";

/**
 * This is the base class for OjRowExpander, and is generated from
 * the component metadata.  Do not put overrides in here as the
 * contents are meant to be replaced by the generated functions.
 * Put overrides into the component's default class, found in
 * index.ts.
 */
export class OjRowExpanderBase extends OjWebElement {
  /**
   * Gets the value of context property.
   * The context object obtained from the column renderer (Table) or cell renderer (DataGrid)
   * @return The value of context property.
   * 
   */
  public getContext(): Promise<object> {
    return this.getProperty<object>("context");
  }

  /**
   * Sets the value of "expanded" property.
   * Specifies if the RowExpander is expanded.
   * @param expanded The value to set for "expanded"
   * @deprecated Since 7.0.0. The source of truth for expansion is the data source key set.
   */
  public changeExpanded(expanded: boolean|null): Promise<void> {
    return this.setProperty<boolean|null>("expanded", expanded);
  }

  /**
   * Gets the value of expanded property.
   * Specifies if the RowExpander is expanded.
   * @return The value of expanded property.
   * @deprecated Since 7.0.0. The source of truth for expansion is the data source key set.
   */
  public getExpanded(): Promise<boolean|null> {
    return this.getProperty<boolean|null>("expanded");
  }

  /**
   * Gets the value of translations property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of translations property.
   * 
   */
  public getTranslations(): Promise<OjRowExpanderBase.Translations|null> {
    return this.getProperty<OjRowExpanderBase.Translations|null>("translations");
  }

}

namespace OjRowExpanderBase {
  export interface Translations {
    /**
     * Provides properties to customize the accessible context level of the expanded or collapsed row.
     */
    "accessibleLevelDescription": string;
    /**
     * Provides properties to customize the accessible context when a row is collapsed.
     */
    "accessibleRowCollapsed": string;
    /**
     * Provides properties to customize the accessible context level and index of the row.
     */
    "accessibleRowDescription": string;
    /**
     * Provides properties to customize the accessible context when a row is expanded.
     */
    "accessibleRowExpanded": string;
    /**
     * Provides properties to customize the accessible state of a collapsed row.
     */
    "accessibleStateCollapsed": string;
    /**
     * Provides properties to customize the accessible state of an expanded row.
     */
    "accessibleStateExpanded": string;
  }
}
